"""
DO NOT EDIT THIS FILE!!
This file is automatically generated by the process_schemas.py program
in the scripts directory. It is not intended to be edited directly. If
you need to update the GEL protocol classes, please run the script
on the appropriate schema version.
"""
from protocols.protocol import ProtocolElement
from protocols.protocol import SearchRequest
from protocols.protocol import SearchResponse
from protocols.protocol import avro_parse

import avro.schema

version = '2.1.0'


class ActionType(object):
    """
    this is the type of actionability for the reported event
    """
    Therapy = "Therapy"
    Prognosis = "Prognosis"
    Diagnosis = "Diagnosis"

    def __hash__(self):
        return str(self).__hash__()


class Actions(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "Actions", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"actionType", "type": {"type": "enum", "name": "ActionType", "doc": "", "symbols": ["Therapy",
"Prognosis", "Diagnosis"]}}, {"name": "tumorType", "type": "string", "doc": ""}, {"name":
"evidence", "type": ["null", {"type": "array", "items": "string"}], "doc": ""}, {"name": "drug",
"type": ["null", "string"], "doc": ""}, {"name": "status", "type": ["null", "string"], "doc": ""},
{"name": "comments", "type": ["null", {"type": "array", "items": "string"}], "doc": ""}, {"name":
"evidenceType", "type": ["null", "string"], "doc": ""}, {"name": "source", "type": "string", "doc":
""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "actionType",
        "comments",
        "drug",
        "evidence",
        "evidenceType",
        "source",
        "status",
        "tumorType",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'actionType', 'comments', 'drug', 'evidence', 'evidenceType',
        'source', 'status', 'tumorType'
    ]

    def __init__(self, **kwargs):
        self.actionType = kwargs.get(
            'actionType', None)
        self.comments = kwargs.get(
            'comments', None)
        self.drug = kwargs.get(
            'drug', None)
        self.evidence = kwargs.get(
            'evidence', None)
        self.evidenceType = kwargs.get(
            'evidenceType', None)
        self.source = kwargs.get(
            'source', None)
        self.status = kwargs.get(
            'status', None)
        self.tumorType = kwargs.get(
            'tumorType', None)


class AdoptedStatus(object):
    """
    adoptedin means adopted into the family adoptedout means child
    belonged to the family and was adopted out
    """
    not_adopted = "not_adopted"
    adoptedin = "adoptedin"
    adoptedout = "adoptedout"

    def __hash__(self):
        return str(self).__hash__()


class AffectionStatus(object):
    """
    Affection Status
    """
    unaffected = "unaffected"
    affected = "affected"
    unknown = "unknown"

    def __hash__(self):
        return str(self).__hash__()


class AnalysisPanel(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "AnalysisPanel", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"specificDisease", "type": "string"}, {"name": "panelName", "type": "string"}, {"name":
"panelVersion", "type": ["null", "string"]}, {"name": "review_outcome", "type": "string"}, {"name":
"multiple_genetic_origins", "type": "string"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "multiple_genetic_origins",
        "panelName",
        "panelVersion",
        "review_outcome",
        "specificDisease",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'multiple_genetic_origins', 'panelName', 'panelVersion',
        'review_outcome', 'specificDisease'
    ]

    def __init__(self, **kwargs):
        self.multiple_genetic_origins = kwargs.get(
            'multiple_genetic_origins', None)
        self.panelName = kwargs.get(
            'panelName', None)
        self.panelVersion = kwargs.get(
            'panelVersion', None)
        self.review_outcome = kwargs.get(
            'review_outcome', None)
        self.specificDisease = kwargs.get(
            'specificDisease', None)


class Ancestries(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "Ancestries", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"mothersEthnicOrigin", "type": ["null", {"type": "enum", "name": "EthnicCategory", "doc": "",
"symbols": ["D", "E", "F", "G", "A", "B", "C", "L", "M", "N", "H", "J", "K", "P", "S", "R", "Z"]}],
"doc": ""}, {"name": "mothersOtherRelevantAncestry", "type": ["null", "string"], "doc": ""},
{"name": "fathersEthnicOrigin", "type": ["null", "EthnicCategory"], "doc": ""}, {"name":
"fathersOtherRelevantAncestry", "type": ["null", "string"], "doc": ""}, {"name":
"chiSquare1KGenomesPhase3Pop", "type": ["null", {"type": "array", "items": {"type": "record",
"name": "ChiSquare1KGenomesPhase3Pop", "doc": "", "fields": [{"name": "kGSuperPopCategory", "type":
{"type": "enum", "name": "KGSuperPopCategory", "doc": "", "symbols": ["AFR", "AMR", "EAS", "EUR",
"SAS"]}, "doc": ""}, {"name": "kGPopCategory", "type": ["null", {"type": "enum", "name":
"KGPopCategory", "doc": "", "symbols": ["ACB", "ASW", "BEB", "CDX", "CEU", "CHB", "CHS", "CLM",
"ESN", "FIN", "GBR", "GIH", "GWD", "IBS", "ITU", "JPT", "KHV", "LWK", "MSL", "MXL", "PEL", "PJL",
"PUR", "STU", "TSI", "YRI"]}], "doc": ""}, {"name": "chiSquare", "type": "float", "doc": ""}]}}],
"doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "chiSquare1KGenomesPhase3Pop",
        "fathersEthnicOrigin",
        "fathersOtherRelevantAncestry",
        "mothersEthnicOrigin",
        "mothersOtherRelevantAncestry",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'chiSquare1KGenomesPhase3Pop': ChiSquare1KGenomesPhase3Pop,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'chiSquare1KGenomesPhase3Pop': ChiSquare1KGenomesPhase3Pop,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'chiSquare1KGenomesPhase3Pop', 'fathersEthnicOrigin',
        'fathersOtherRelevantAncestry', 'mothersEthnicOrigin',
        'mothersOtherRelevantAncestry'
    ]

    def __init__(self, **kwargs):
        self.chiSquare1KGenomesPhase3Pop = kwargs.get(
            'chiSquare1KGenomesPhase3Pop', None)
        self.fathersEthnicOrigin = kwargs.get(
            'fathersEthnicOrigin', None)
        self.fathersOtherRelevantAncestry = kwargs.get(
            'fathersOtherRelevantAncestry', None)
        self.mothersEthnicOrigin = kwargs.get(
            'mothersEthnicOrigin', None)
        self.mothersOtherRelevantAncestry = kwargs.get(
            'mothersOtherRelevantAncestry', None)


class ArrayConcordance(ProtocolElement):
    """
    Array concordance with WGS data  * `numberOfSites`: Number of
    sites considered * `numberOfDiscordantSites`: Number of sites
    discordant between WGS and Genotyping Array
    """
    _schemaSource = """
{"type": "record", "name": "ArrayConcordance", "namespace": "Gel_BioInf_Models", "doc": "",
"fields": [{"name": "numberOfSites", "type": "double"}, {"name": "numberOfDiscordantSites", "type":
"double"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "numberOfDiscordantSites",
        "numberOfSites",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'numberOfDiscordantSites', 'numberOfSites'
    ]

    def __init__(self, **kwargs):
        self.numberOfDiscordantSites = kwargs.get(
            'numberOfDiscordantSites', None)
        self.numberOfSites = kwargs.get(
            'numberOfSites', None)


class ArrayGenotypingRate(ProtocolElement):
    """
    Array genotyping rate for the sample * `IID` - Individual sample
    ID * `number_missing_genotypes`: The number of missing genotypes *
    `number_total_genotypes`: The number of total genotypes
    """
    _schemaSource = """
{"type": "record", "name": "ArrayGenotypingRate", "namespace": "Gel_BioInf_Models", "doc": "",
"fields": [{"name": "IID", "type": "string"}, {"name": "number_missing_genotypes", "type":
"double"}, {"name": "number_total_genotypes", "type": "double"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "IID",
        "number_missing_genotypes",
        "number_total_genotypes",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'IID', 'number_missing_genotypes', 'number_total_genotypes'
    ]

    def __init__(self, **kwargs):
        self.IID = kwargs.get(
            'IID', None)
        self.number_missing_genotypes = kwargs.get(
            'number_missing_genotypes', None)
        self.number_total_genotypes = kwargs.get(
            'number_total_genotypes', None)


class BamHeaderMachine(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "BamHeaderMachine", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"machines", "type": {"type": "array", "items": {"type": "record", "name": "Machine", "fields":
[{"name": "DATE", "type": "double"}, {"name": "MACHINE", "type": "string"}, {"name": "FLOWCELL",
"type": "string"}, {"name": "RUN", "type": "string"}]}}}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "machines",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'machines': Machine,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'machines': Machine,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'machines'
    ]

    def __init__(self, **kwargs):
        self.machines = kwargs.get(
            'machines', None)


class BamHeaderOther(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "BamHeaderOther", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"PIPELINE_ID", "type": "string"}, {"name": "PIPELINE_VERSION", "type": "string"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "PIPELINE_ID",
        "PIPELINE_VERSION",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'PIPELINE_ID', 'PIPELINE_VERSION'
    ]

    def __init__(self, **kwargs):
        self.PIPELINE_ID = kwargs.get(
            'PIPELINE_ID', None)
        self.PIPELINE_VERSION = kwargs.get(
            'PIPELINE_VERSION', None)


class CalledGenotype(ProtocolElement):
    """
    This is intended to hold the genotypes for the family members In
    principle it is a phased zygosity as in VCF spec and called by the
    analysis provider if further phasing is conducted
    """
    _schemaSource = """
{"type": "record", "name": "CalledGenotype", "namespace": "Gel_BioInf_Models", "doc": "", "fields":
[{"name": "gelId", "type": "string", "doc": ""}, {"name": "sampleId", "type": "string", "doc": ""},
{"name": "genotype", "type": "string", "doc": ""}, {"name": "phaseSet", "type": ["null", "int"],
"doc": ""}, {"name": "depthReference", "type": ["null", "int"], "doc": ""}, {"name":
"depthAlternate", "type": ["null", "int"], "doc": ""}, {"name": "copyNumber", "type": ["null",
"int"], "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "copyNumber",
        "depthAlternate",
        "depthReference",
        "gelId",
        "genotype",
        "phaseSet",
        "sampleId",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'copyNumber', 'depthAlternate', 'depthReference', 'gelId',
        'genotype', 'phaseSet', 'sampleId'
    ]

    def __init__(self, **kwargs):
        self.copyNumber = kwargs.get(
            'copyNumber', None)
        self.depthAlternate = kwargs.get(
            'depthAlternate', None)
        self.depthReference = kwargs.get(
            'depthReference', None)
        self.gelId = kwargs.get(
            'gelId', None)
        self.genotype = kwargs.get(
            'genotype', None)
        self.phaseSet = kwargs.get(
            'phaseSet', None)
        self.sampleId = kwargs.get(
            'sampleId', None)


class CancerDemographics(ProtocolElement):
    """
    This defines a Cancer Demographics
    """
    _schemaSource = """
{"type": "record", "name": "CancerDemographics", "namespace": "Gel_BioInf_Models", "doc": "",
"fields": [{"name": "gelId", "type": "string", "doc": ""}, {"name": "center", "type": "string",
"doc": ""}, {"name": "centerPatientId", "type": "string", "doc": ""}, {"name":
"labkeyParticipantId", "type": ["null", "string"], "doc": ""}, {"name": "primaryDiagnosis", "type":
["null", "string"], "doc": ""}, {"name": "dataModelVersion", "type": "string", "doc": "", "default":
"v2.4"}, {"name": "sex", "type": ["null", {"type": "enum", "name": "Sex", "symbols": ["M", "F"]}],
"doc": ""}, {"name": "consentStatus", "type": {"type": "record", "name": "ConsentStatus", "doc": "",
"fields": [{"name": "programmeConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"primaryFindingConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"secondaryFindingConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"carrierStatusConsent", "type": "boolean", "doc": "", "default": false}]}, "doc": ""}, {"name":
"additionalInformation", "type": ["null", {"type": "map", "values": "string"}], "doc": ""}, {"name":
"sampleId", "type": ["null", {"type": "array", "items": "string"}], "doc": ""}, {"name":
"assignedICD10", "type": ["null", "string"], "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "additionalInformation",
        "assignedICD10",
        "center",
        "centerPatientId",
        "consentStatus",
        "gelId",
        "labkeyParticipantId",
        "primaryDiagnosis",
        "sampleId",
        "sex",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'consentStatus': ConsentStatus,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'consentStatus': ConsentStatus,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'additionalInformation', 'assignedICD10', 'center',
        'centerPatientId', 'consentStatus', 'dataModelVersion',
        'gelId', 'labkeyParticipantId', 'primaryDiagnosis',
        'sampleId', 'sex'
    ]

    def __init__(self, **kwargs):
        self.additionalInformation = kwargs.get(
            'additionalInformation', None)
        self.assignedICD10 = kwargs.get(
            'assignedICD10', None)
        self.center = kwargs.get(
            'center', None)
        self.centerPatientId = kwargs.get(
            'centerPatientId', None)
        self.consentStatus = kwargs.get(
            'consentStatus', ConsentStatus())
        self.dataModelVersion = kwargs.get(
            'dataModelVersion', 'v2.4')
        self.gelId = kwargs.get(
            'gelId', None)
        self.labkeyParticipantId = kwargs.get(
            'labkeyParticipantId', None)
        self.primaryDiagnosis = kwargs.get(
            'primaryDiagnosis', None)
        self.sampleId = kwargs.get(
            'sampleId', None)
        self.sex = kwargs.get(
            'sex', None)


class CancerInterpretationRequest(ProtocolElement):
    """
    This record represents basic information for this report
    """
    _schemaSource = """
{"type": "record", "name": "CancerInterpretationRequest", "namespace": "Gel_BioInf_Models", "doc":
"", "fields": [{"name": "versionControl", "type": {"type": "record", "name": "VersionControl",
"fields": [{"name": "GitVersionControl", "type": "string", "doc": "", "default": "2.1.0"}]}, "doc":
""}, {"name": "reportRequestId", "type": "string", "doc": ""}, {"name": "reportVersion", "type":
"int", "doc": ""}, {"name": "interpretGenome", "type": "boolean", "doc": "", "default": false},
{"name": "BAMs", "type": {"type": "array", "items": {"type": "record", "name": "File", "fields":
[{"name": "SampleId", "type": ["null", "string", {"type": "array", "items": "string"}], "doc": ""},
{"name": "URIFile", "type": "string", "doc": ""}, {"name": "fileType", "type": {"type": "enum",
"name": "FileType", "doc": "", "symbols": ["BAM", "gVCF", "VCF_small", "VCF_CNV", "VCF_SV",
"VCF_SV_CNV", "SVG", "TIER", "ANN", "BigWig"]}}]}}, "doc": ""}, {"name": "VCFs", "type": {"type":
"array", "items": "File"}, "doc": ""}, {"name": "bigWigs", "type": {"type": "array", "items":
"File"}, "doc": ""}, {"name": "annotationFile", "type": ["null", "File"], "doc": ""}, {"name":
"cancerParticipant", "type": {"type": "record", "name": "CancerParticipant", "doc": "", "fields":
[{"name": "versionControl", "type": "VersionControl", "doc": ""}, {"name": "cancerDemographics",
"type": {"type": "record", "name": "CancerDemographics", "doc": "", "fields": [{"name": "gelId",
"type": "string", "doc": ""}, {"name": "center", "type": "string", "doc": ""}, {"name":
"centerPatientId", "type": "string", "doc": ""}, {"name": "labkeyParticipantId", "type": ["null",
"string"], "doc": ""}, {"name": "primaryDiagnosis", "type": ["null", "string"], "doc": ""}, {"name":
"dataModelVersion", "type": "string", "doc": "", "default": "v2.4"}, {"name": "sex", "type":
["null", {"type": "enum", "name": "Sex", "symbols": ["M", "F"]}], "doc": ""}, {"name":
"consentStatus", "type": {"type": "record", "name": "ConsentStatus", "doc": "", "fields": [{"name":
"programmeConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"primaryFindingConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"secondaryFindingConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"carrierStatusConsent", "type": "boolean", "doc": "", "default": false}]}, "doc": ""}, {"name":
"additionalInformation", "type": ["null", {"type": "map", "values": "string"}], "doc": ""}, {"name":
"sampleId", "type": ["null", {"type": "array", "items": "string"}], "doc": ""}, {"name":
"assignedICD10", "type": ["null", "string"], "doc": ""}]}, "doc": ""}, {"name": "cancerSamples",
"type": {"type": "array", "items": {"type": "record", "name": "CancerSample", "fields": [{"name":
"sampleId", "type": "string", "doc": ""}, {"name": "labId", "type": ["null", "string"], "doc": ""},
{"name": "gelPhase", "type": ["null", {"type": "enum", "name": "GelPhase", "symbols": ["CRUK",
"OXFORD", "CLL", "IIP", "MAIN", "EXPT"]}], "doc": ""}, {"name": "sampleType", "type": {"type":
"enum", "name": "SampleType", "symbols": ["germline", "tumor"]}, "doc": ""}, {"name":
"sampleDiagnosis", "type": ["null", "string"], "doc": ""}, {"name": "tumorType", "type": ["null",
"string"], "doc": ""}, {"name": "tumorSubType", "type": ["null", "string"], "doc": ""}, {"name":
"preservationMethod", "type": ["null", {"type": "enum", "name": "PreservationMethod", "symbols":
["FFPE", "FF", "UNKNOWN", "BLOOD", "GL", "SALIVA", "LEUK"]}], "doc": ""}, {"name": "phase", "type":
["null", {"type": "enum", "name": "Phase", "symbols": ["PRIMARY", "METASTATIC", "RECURRENCE"]}],
"doc": ""}, {"name": "method", "type": ["null", {"type": "enum", "name": "Method", "symbols":
["RESECTION", "BIOPSY", "BLOOD"]}], "doc": ""}, {"name": "cellularity", "type": ["null", "string"],
"doc": ""}, {"name": "tumorContent", "type": ["null", "string"], "doc": ""}, {"name": "grade",
"type": ["null", "string"], "doc": ""}, {"name": "tnm_stage_version", "type": ["null", "string"],
"doc": ""}, {"name": "tmn_stage_grouping", "type": ["null", "string"], "doc": ""}]}}, "doc": ""},
{"name": "matchedSamples", "type": {"type": "array", "items": {"type": "record", "name":
"MatchedSamples", "doc": "", "fields": [{"name": "germlineSampleId", "type": ["null", "string"],
"doc": ""}, {"name": "tumorSampleId", "type": ["null", "string"], "doc": ""}]}}, "doc": ""}]}},
{"name": "analysisURI", "type": "string", "doc": ""}, {"name": "analysisVersion", "type": "string",
"doc": ""}, {"name": "additionalInfo", "type": ["null", {"type": "map", "values": "string"}], "doc":
""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "BAMs",
        "VCFs",
        "additionalInfo",
        "analysisURI",
        "analysisVersion",
        "annotationFile",
        "bigWigs",
        "cancerParticipant",
        "reportRequestId",
        "reportVersion",
        "versionControl",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'BAMs': File,
            'VCFs': File,
            'annotationFile': File,
            'bigWigs': File,
            'cancerParticipant': CancerParticipant,
            'versionControl': VersionControl,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'BAMs': File,
            'VCFs': File,
            'annotationFile': File,
            'bigWigs': File,
            'cancerParticipant': CancerParticipant,
            'versionControl': VersionControl,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'BAMs', 'VCFs', 'additionalInfo', 'analysisURI',
        'analysisVersion', 'annotationFile', 'bigWigs',
        'cancerParticipant', 'interpretGenome', 'reportRequestId',
        'reportVersion', 'versionControl'
    ]

    def __init__(self, **kwargs):
        self.BAMs = kwargs.get(
            'BAMs', None)
        self.VCFs = kwargs.get(
            'VCFs', None)
        self.additionalInfo = kwargs.get(
            'additionalInfo', None)
        self.analysisURI = kwargs.get(
            'analysisURI', None)
        self.analysisVersion = kwargs.get(
            'analysisVersion', None)
        self.annotationFile = kwargs.get(
            'annotationFile', None)
        self.bigWigs = kwargs.get(
            'bigWigs', None)
        self.cancerParticipant = kwargs.get(
            'cancerParticipant', CancerParticipant())
        self.interpretGenome = kwargs.get(
            'interpretGenome', False)
        self.reportRequestId = kwargs.get(
            'reportRequestId', None)
        self.reportVersion = kwargs.get(
            'reportVersion', None)
        self.versionControl = kwargs.get(
            'versionControl', VersionControl())


class CancerInterpretedGenome(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "CancerInterpretedGenome", "namespace": "Gel_BioInf_Models", "fields":
[{"name": "versionControl", "type": {"type": "record", "name": "VersionControl", "fields": [{"name":
"GitVersionControl", "type": "string", "doc": "", "default": "2.1.0"}]}, "doc": ""}, {"name":
"reportRequestId", "type": "string", "doc": ""}, {"name": "analysisId", "type": "string", "doc":
""}, {"name": "reportURI", "type": "string", "doc": ""}, {"name": "reportedVariants", "type":
{"type": "array", "items": {"type": "record", "name": "ReportedSomaticVariants", "fields": [{"name":
"reportedVariant", "type": {"type": "record", "name": "ReportedVariant", "fields": [{"name":
"versionControl", "type": "VersionControl", "doc": ""}, {"name": "chromosome", "type": "string",
"doc": ""}, {"name": "dbSNPid", "type": ["null", "string"], "doc": ""}, {"name": "position", "type":
"int", "doc": ""}, {"name": "reference", "type": "string", "doc": ""}, {"name": "alternate", "type":
"string", "doc": ""}, {"name": "calledGenotypes", "type": {"type": "array", "items": {"type":
"record", "name": "CalledGenotype", "doc": "", "fields": [{"name": "gelId", "type": "string", "doc":
""}, {"name": "sampleId", "type": "string", "doc": ""}, {"name": "genotype", "type": "string",
"doc": ""}, {"name": "phaseSet", "type": ["null", "int"], "doc": ""}, {"name": "depthReference",
"type": ["null", "int"], "doc": ""}, {"name": "depthAlternate", "type": ["null", "int"], "doc": ""},
{"name": "copyNumber", "type": ["null", "int"], "doc": ""}]}}, "doc": ""}, {"name": "reportEvents",
"type": {"type": "array", "items": {"type": "record", "name": "ReportEvent", "fields": [{"name":
"reportEventId", "type": "string", "doc": ""}, {"name": "phenotype", "type": "string", "doc": ""},
{"name": "panelName", "type": ["null", "string"], "doc": ""}, {"name": "panelVersion", "type":
["null", "string"], "doc": ""}, {"name": "modeOfInheritance", "type": {"type": "enum", "name":
"ReportedModeOfInheritance", "doc": "", "symbols": ["monoallelic", "monoallelic_not_imprinted",
"monoallelic_maternally_imprinted", "monoallelic_paternally_imprinted", "biallelic",
"monoallelic_and_biallelic", "monoallelic_and_more_severe_biallelic", "xlinked_biallelic",
"xlinked_monoallelic", "mitochondrial", "unknown"]}, "doc": ""}, {"name": "genomicFeature", "type":
{"type": "record", "name": "GenomicFeature", "fields": [{"name": "featureType", "type": {"type":
"enum", "name": "FeatureTypes", "symbols": ["RegulatoryRegion", "Gene", "Transcript"]}, "doc": ""},
{"name": "ensemblId", "type": "string", "doc": ""}, {"name": "ids", "type": {"type": "map",
"values": "string"}, "doc": ""}]}, "doc": ""}, {"name": "penetrance", "type": {"type": "enum",
"name": "Penetrance", "doc": "", "symbols": ["complete", "incomplete"]}, "doc": ""}, {"name":
"score", "type": "float", "doc": ""}, {"name": "vendorSpecificScores", "type": ["null", {"type":
"map", "values": "float"}], "doc": ""}, {"name": "variantClassification", "type": ["null", {"type":
"enum", "name": "VariantClassification", "doc": "", "symbols": ["BENIGN", "LIKELY_BENIGN", "VUS",
"LIKELY_PATHOGENIC", "PATHOGENIC"]}], "doc": ""}, {"name": "fullyExplainsPhenotype", "type":
["null", "boolean"], "doc": ""}, {"name": "groupOfVariants", "type": ["null", "int"], "doc": ""},
{"name": "eventJustification", "type": "string", "doc": ""}, {"name": "tier", "type": ["null",
{"type": "enum", "name": "Tier", "doc": "", "symbols": ["NONE", "TIER1", "TIER2", "TIER3"]}], "doc":
""}]}}, "doc": ""}, {"name": "additionalTextualVariantAnnotations", "type": ["null", {"type": "map",
"values": "string"}], "doc": ""}, {"name": "additionalNumericVariantAnnotations", "type": ["null",
{"type": "map", "values": "float"}], "doc": ""}, {"name": "comments", "type": ["null", {"type":
"array", "items": "string"}], "doc": ""}]}, "doc": ""}, {"name": "somaticOrGermline", "type":
{"type": "enum", "name": "SomaticOrGermline", "doc": "", "symbols": ["unknown", "somatic",
"germline"]}, "doc": ""}]}}}, {"name": "reportedStructuralVariants", "type": {"type": "array",
"items": {"type": "record", "name": "ReportedSomaticStructuralVariants", "fields": [{"name":
"reportedStructuralVariant", "type": {"type": "record", "name": "ReportedStructuralVariant",
"fields": [{"name": "versionControl", "type": "VersionControl", "doc": ""}, {"name": "chromosome",
"type": "string", "doc": ""}, {"name": "start", "type": "int", "doc": ""}, {"name": "end", "type":
"int", "doc": ""}, {"name": "type", "type": "string", "doc": ""}, {"name": "reference", "type":
"string", "doc": ""}, {"name": "alternate", "type": "string", "doc": ""}, {"name":
"calledGenotypes", "type": {"type": "array", "items": "CalledGenotype"}}, {"name": "reportEvents",
"type": {"type": "array", "items": "ReportEvent"}, "doc": ""}, {"name":
"additionalTextualVariantAnnotations", "type": ["null", {"type": "map", "values": "string"}], "doc":
""}, {"name": "additionalNumericVariantAnnotations", "type": ["null", {"type": "map", "values":
"float"}], "doc": ""}, {"name": "comments", "type": ["null", {"type": "array", "items": "string"}],
"doc": ""}]}, "doc": ""}, {"name": "somaticOrGermline", "type": "SomaticOrGermline", "doc": ""}]}}},
{"name": "comments", "type": ["null", {"type": "array", "items": "string"}], "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "analysisId",
        "comments",
        "reportRequestId",
        "reportURI",
        "reportedStructuralVariants",
        "reportedVariants",
        "versionControl",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'reportedStructuralVariants': ReportedSomaticStructuralVariants,
            'reportedVariants': ReportedSomaticVariants,
            'versionControl': VersionControl,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'reportedStructuralVariants': ReportedSomaticStructuralVariants,
            'reportedVariants': ReportedSomaticVariants,
            'versionControl': VersionControl,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'analysisId', 'comments', 'reportRequestId', 'reportURI',
        'reportedStructuralVariants', 'reportedVariants',
        'versionControl'
    ]

    def __init__(self, **kwargs):
        self.analysisId = kwargs.get(
            'analysisId', None)
        self.comments = kwargs.get(
            'comments', None)
        self.reportRequestId = kwargs.get(
            'reportRequestId', None)
        self.reportURI = kwargs.get(
            'reportURI', None)
        self.reportedStructuralVariants = kwargs.get(
            'reportedStructuralVariants', None)
        self.reportedVariants = kwargs.get(
            'reportedVariants', None)
        self.versionControl = kwargs.get(
            'versionControl', VersionControl())


class CancerParticipant(ProtocolElement):
    """
    This defines a Cancer Participant (demographics and sample
    information)
    """
    _schemaSource = """
{"type": "record", "name": "CancerParticipant", "namespace": "Gel_BioInf_Models", "doc": "",
"fields": [{"name": "versionControl", "type": {"type": "record", "name": "VersionControl", "fields":
[{"name": "GitVersionControl", "type": "string", "doc": "", "default": "2.1.0"}]}, "doc": ""},
{"name": "cancerDemographics", "type": {"type": "record", "name": "CancerDemographics", "doc": "",
"fields": [{"name": "gelId", "type": "string", "doc": ""}, {"name": "center", "type": "string",
"doc": ""}, {"name": "centerPatientId", "type": "string", "doc": ""}, {"name":
"labkeyParticipantId", "type": ["null", "string"], "doc": ""}, {"name": "primaryDiagnosis", "type":
["null", "string"], "doc": ""}, {"name": "dataModelVersion", "type": "string", "doc": "", "default":
"v2.4"}, {"name": "sex", "type": ["null", {"type": "enum", "name": "Sex", "symbols": ["M", "F"]}],
"doc": ""}, {"name": "consentStatus", "type": {"type": "record", "name": "ConsentStatus", "doc": "",
"fields": [{"name": "programmeConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"primaryFindingConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"secondaryFindingConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"carrierStatusConsent", "type": "boolean", "doc": "", "default": false}]}, "doc": ""}, {"name":
"additionalInformation", "type": ["null", {"type": "map", "values": "string"}], "doc": ""}, {"name":
"sampleId", "type": ["null", {"type": "array", "items": "string"}], "doc": ""}, {"name":
"assignedICD10", "type": ["null", "string"], "doc": ""}]}, "doc": ""}, {"name": "cancerSamples",
"type": {"type": "array", "items": {"type": "record", "name": "CancerSample", "fields": [{"name":
"sampleId", "type": "string", "doc": ""}, {"name": "labId", "type": ["null", "string"], "doc": ""},
{"name": "gelPhase", "type": ["null", {"type": "enum", "name": "GelPhase", "symbols": ["CRUK",
"OXFORD", "CLL", "IIP", "MAIN", "EXPT"]}], "doc": ""}, {"name": "sampleType", "type": {"type":
"enum", "name": "SampleType", "symbols": ["germline", "tumor"]}, "doc": ""}, {"name":
"sampleDiagnosis", "type": ["null", "string"], "doc": ""}, {"name": "tumorType", "type": ["null",
"string"], "doc": ""}, {"name": "tumorSubType", "type": ["null", "string"], "doc": ""}, {"name":
"preservationMethod", "type": ["null", {"type": "enum", "name": "PreservationMethod", "symbols":
["FFPE", "FF", "UNKNOWN", "BLOOD", "GL", "SALIVA", "LEUK"]}], "doc": ""}, {"name": "phase", "type":
["null", {"type": "enum", "name": "Phase", "symbols": ["PRIMARY", "METASTATIC", "RECURRENCE"]}],
"doc": ""}, {"name": "method", "type": ["null", {"type": "enum", "name": "Method", "symbols":
["RESECTION", "BIOPSY", "BLOOD"]}], "doc": ""}, {"name": "cellularity", "type": ["null", "string"],
"doc": ""}, {"name": "tumorContent", "type": ["null", "string"], "doc": ""}, {"name": "grade",
"type": ["null", "string"], "doc": ""}, {"name": "tnm_stage_version", "type": ["null", "string"],
"doc": ""}, {"name": "tmn_stage_grouping", "type": ["null", "string"], "doc": ""}]}}, "doc": ""},
{"name": "matchedSamples", "type": {"type": "array", "items": {"type": "record", "name":
"MatchedSamples", "doc": "", "fields": [{"name": "germlineSampleId", "type": ["null", "string"],
"doc": ""}, {"name": "tumorSampleId", "type": ["null", "string"], "doc": ""}]}}, "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "cancerDemographics",
        "cancerSamples",
        "matchedSamples",
        "versionControl",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'cancerDemographics': CancerDemographics,
            'cancerSamples': CancerSample,
            'matchedSamples': MatchedSamples,
            'versionControl': VersionControl,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'cancerDemographics': CancerDemographics,
            'cancerSamples': CancerSample,
            'matchedSamples': MatchedSamples,
            'versionControl': VersionControl,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'cancerDemographics', 'cancerSamples', 'matchedSamples',
        'versionControl'
    ]

    def __init__(self, **kwargs):
        self.cancerDemographics = kwargs.get(
            'cancerDemographics', CancerDemographics())
        self.cancerSamples = kwargs.get(
            'cancerSamples', None)
        self.matchedSamples = kwargs.get(
            'matchedSamples', None)
        self.versionControl = kwargs.get(
            'versionControl', VersionControl())


class CancerSample(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "CancerSample", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"sampleId", "type": "string", "doc": ""}, {"name": "labId", "type": ["null", "string"], "doc": ""},
{"name": "gelPhase", "type": ["null", {"type": "enum", "name": "GelPhase", "symbols": ["CRUK",
"OXFORD", "CLL", "IIP", "MAIN", "EXPT"]}], "doc": ""}, {"name": "sampleType", "type": {"type":
"enum", "name": "SampleType", "symbols": ["germline", "tumor"]}, "doc": ""}, {"name":
"sampleDiagnosis", "type": ["null", "string"], "doc": ""}, {"name": "tumorType", "type": ["null",
"string"], "doc": ""}, {"name": "tumorSubType", "type": ["null", "string"], "doc": ""}, {"name":
"preservationMethod", "type": ["null", {"type": "enum", "name": "PreservationMethod", "symbols":
["FFPE", "FF", "UNKNOWN", "BLOOD", "GL", "SALIVA", "LEUK"]}], "doc": ""}, {"name": "phase", "type":
["null", {"type": "enum", "name": "Phase", "symbols": ["PRIMARY", "METASTATIC", "RECURRENCE"]}],
"doc": ""}, {"name": "method", "type": ["null", {"type": "enum", "name": "Method", "symbols":
["RESECTION", "BIOPSY", "BLOOD"]}], "doc": ""}, {"name": "cellularity", "type": ["null", "string"],
"doc": ""}, {"name": "tumorContent", "type": ["null", "string"], "doc": ""}, {"name": "grade",
"type": ["null", "string"], "doc": ""}, {"name": "tnm_stage_version", "type": ["null", "string"],
"doc": ""}, {"name": "tmn_stage_grouping", "type": ["null", "string"], "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "cellularity",
        "gelPhase",
        "grade",
        "labId",
        "method",
        "phase",
        "preservationMethod",
        "sampleDiagnosis",
        "sampleId",
        "sampleType",
        "tmn_stage_grouping",
        "tnm_stage_version",
        "tumorContent",
        "tumorSubType",
        "tumorType",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'cellularity', 'gelPhase', 'grade', 'labId', 'method',
        'phase', 'preservationMethod', 'sampleDiagnosis', 'sampleId',
        'sampleType', 'tmn_stage_grouping', 'tnm_stage_version',
        'tumorContent', 'tumorSubType', 'tumorType'
    ]

    def __init__(self, **kwargs):
        self.cellularity = kwargs.get(
            'cellularity', None)
        self.gelPhase = kwargs.get(
            'gelPhase', None)
        self.grade = kwargs.get(
            'grade', None)
        self.labId = kwargs.get(
            'labId', None)
        self.method = kwargs.get(
            'method', None)
        self.phase = kwargs.get(
            'phase', None)
        self.preservationMethod = kwargs.get(
            'preservationMethod', None)
        self.sampleDiagnosis = kwargs.get(
            'sampleDiagnosis', None)
        self.sampleId = kwargs.get(
            'sampleId', None)
        self.sampleType = kwargs.get(
            'sampleType', None)
        self.tmn_stage_grouping = kwargs.get(
            'tmn_stage_grouping', None)
        self.tnm_stage_version = kwargs.get(
            'tnm_stage_version', None)
        self.tumorContent = kwargs.get(
            'tumorContent', None)
        self.tumorSubType = kwargs.get(
            'tumorSubType', None)
        self.tumorType = kwargs.get(
            'tumorType', None)


class ChiSquare1KGenomesPhase3Pop(ProtocolElement):
    """
    Chi-square test for goodness of fit of this sample to 1000 Genomes
    Phase 3 populations
    """
    _schemaSource = """
{"type": "record", "name": "ChiSquare1KGenomesPhase3Pop", "namespace": "Gel_BioInf_Models", "doc":
"", "fields": [{"name": "kGSuperPopCategory", "type": {"type": "enum", "name": "KGSuperPopCategory",
"doc": "", "symbols": ["AFR", "AMR", "EAS", "EUR", "SAS"]}, "doc": ""}, {"name": "kGPopCategory",
"type": ["null", {"type": "enum", "name": "KGPopCategory", "doc": "", "symbols": ["ACB", "ASW",
"BEB", "CDX", "CEU", "CHB", "CHS", "CLM", "ESN", "FIN", "GBR", "GIH", "GWD", "IBS", "ITU", "JPT",
"KHV", "LWK", "MSL", "MXL", "PEL", "PJL", "PUR", "STU", "TSI", "YRI"]}], "doc": ""}, {"name":
"chiSquare", "type": "float", "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "chiSquare",
        "kGPopCategory",
        "kGSuperPopCategory",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'chiSquare', 'kGPopCategory', 'kGSuperPopCategory'
    ]

    def __init__(self, **kwargs):
        self.chiSquare = kwargs.get(
            'chiSquare', None)
        self.kGPopCategory = kwargs.get(
            'kGPopCategory', None)
        self.kGSuperPopCategory = kwargs.get(
            'kGSuperPopCategory', None)


class ClinicalReportCancer(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "ClinicalReportCancer", "namespace": "Gel_BioInf_Models", "fields":
[{"name": "interpretationRequestID", "type": "string", "doc": ""}, {"name":
"interpretationRequestVersion", "type": "string", "doc": ""}, {"name": "reportingDate", "type":
"string", "doc": ""}, {"name": "user", "type": "string", "doc": ""}, {"name": "candidateVariants",
"type": {"type": "array", "items": {"type": "record", "name": "ReportedVariant", "fields": [{"name":
"versionControl", "type": {"type": "record", "name": "VersionControl", "fields": [{"name":
"GitVersionControl", "type": "string", "doc": "", "default": "2.1.0"}]}, "doc": ""}, {"name":
"chromosome", "type": "string", "doc": ""}, {"name": "dbSNPid", "type": ["null", "string"], "doc":
""}, {"name": "position", "type": "int", "doc": ""}, {"name": "reference", "type": "string", "doc":
""}, {"name": "alternate", "type": "string", "doc": ""}, {"name": "calledGenotypes", "type":
{"type": "array", "items": {"type": "record", "name": "CalledGenotype", "doc": "", "fields":
[{"name": "gelId", "type": "string", "doc": ""}, {"name": "sampleId", "type": "string", "doc": ""},
{"name": "genotype", "type": "string", "doc": ""}, {"name": "phaseSet", "type": ["null", "int"],
"doc": ""}, {"name": "depthReference", "type": ["null", "int"], "doc": ""}, {"name":
"depthAlternate", "type": ["null", "int"], "doc": ""}, {"name": "copyNumber", "type": ["null",
"int"], "doc": ""}]}}, "doc": ""}, {"name": "reportEvents", "type": {"type": "array", "items":
{"type": "record", "name": "ReportEvent", "fields": [{"name": "reportEventId", "type": "string",
"doc": ""}, {"name": "phenotype", "type": "string", "doc": ""}, {"name": "panelName", "type":
["null", "string"], "doc": ""}, {"name": "panelVersion", "type": ["null", "string"], "doc": ""},
{"name": "modeOfInheritance", "type": {"type": "enum", "name": "ReportedModeOfInheritance", "doc":
"", "symbols": ["monoallelic", "monoallelic_not_imprinted", "monoallelic_maternally_imprinted",
"monoallelic_paternally_imprinted", "biallelic", "monoallelic_and_biallelic",
"monoallelic_and_more_severe_biallelic", "xlinked_biallelic", "xlinked_monoallelic",
"mitochondrial", "unknown"]}, "doc": ""}, {"name": "genomicFeature", "type": {"type": "record",
"name": "GenomicFeature", "fields": [{"name": "featureType", "type": {"type": "enum", "name":
"FeatureTypes", "symbols": ["RegulatoryRegion", "Gene", "Transcript"]}, "doc": ""}, {"name":
"ensemblId", "type": "string", "doc": ""}, {"name": "ids", "type": {"type": "map", "values":
"string"}, "doc": ""}]}, "doc": ""}, {"name": "penetrance", "type": {"type": "enum", "name":
"Penetrance", "doc": "", "symbols": ["complete", "incomplete"]}, "doc": ""}, {"name": "score",
"type": "float", "doc": ""}, {"name": "vendorSpecificScores", "type": ["null", {"type": "map",
"values": "float"}], "doc": ""}, {"name": "variantClassification", "type": ["null", {"type": "enum",
"name": "VariantClassification", "doc": "", "symbols": ["BENIGN", "LIKELY_BENIGN", "VUS",
"LIKELY_PATHOGENIC", "PATHOGENIC"]}], "doc": ""}, {"name": "fullyExplainsPhenotype", "type":
["null", "boolean"], "doc": ""}, {"name": "groupOfVariants", "type": ["null", "int"], "doc": ""},
{"name": "eventJustification", "type": "string", "doc": ""}, {"name": "tier", "type": ["null",
{"type": "enum", "name": "Tier", "doc": "", "symbols": ["NONE", "TIER1", "TIER2", "TIER3"]}], "doc":
""}]}}, "doc": ""}, {"name": "additionalTextualVariantAnnotations", "type": ["null", {"type": "map",
"values": "string"}], "doc": ""}, {"name": "additionalNumericVariantAnnotations", "type": ["null",
{"type": "map", "values": "float"}], "doc": ""}, {"name": "comments", "type": ["null", {"type":
"array", "items": "string"}], "doc": ""}]}}, "doc": ""}, {"name": "candidateStructuralVariants",
"type": {"type": "array", "items": {"type": "record", "name": "ReportedStructuralVariant", "fields":
[{"name": "versionControl", "type": "VersionControl", "doc": ""}, {"name": "chromosome", "type":
"string", "doc": ""}, {"name": "start", "type": "int", "doc": ""}, {"name": "end", "type": "int",
"doc": ""}, {"name": "type", "type": "string", "doc": ""}, {"name": "reference", "type": "string",
"doc": ""}, {"name": "alternate", "type": "string", "doc": ""}, {"name": "calledGenotypes", "type":
{"type": "array", "items": "CalledGenotype"}}, {"name": "reportEvents", "type": {"type": "array",
"items": "ReportEvent"}, "doc": ""}, {"name": "additionalTextualVariantAnnotations", "type":
["null", {"type": "map", "values": "string"}], "doc": ""}, {"name":
"additionalNumericVariantAnnotations", "type": ["null", {"type": "map", "values": "float"}], "doc":
""}, {"name": "comments", "type": ["null", {"type": "array", "items": "string"}], "doc": ""}]}},
"doc": ""}, {"name": "genomicInterpretation", "type": "string", "doc": ""}, {"name": "references",
"type": ["null", {"type": "array", "items": "string"}], "doc": ""}, {"name":
"referenceDatabasesVersions", "type": {"type": "map", "values": "string"}, "doc": ""}, {"name":
"softwareVersions", "type": {"type": "map", "values": "string"}, "doc": ""}, {"name":
"genePanelsCoverage", "type": {"type": "map", "values": {"type": "array", "items": {"type": "map",
"values": "string"}}}, "doc": ""}, {"name": "cancerParticipant", "type": {"type": "record", "name":
"CancerParticipant", "doc": "", "fields": [{"name": "versionControl", "type": "VersionControl",
"doc": ""}, {"name": "cancerDemographics", "type": {"type": "record", "name": "CancerDemographics",
"doc": "", "fields": [{"name": "gelId", "type": "string", "doc": ""}, {"name": "center", "type":
"string", "doc": ""}, {"name": "centerPatientId", "type": "string", "doc": ""}, {"name":
"labkeyParticipantId", "type": ["null", "string"], "doc": ""}, {"name": "primaryDiagnosis", "type":
["null", "string"], "doc": ""}, {"name": "dataModelVersion", "type": "string", "doc": "", "default":
"v2.4"}, {"name": "sex", "type": ["null", {"type": "enum", "name": "Sex", "symbols": ["M", "F"]}],
"doc": ""}, {"name": "consentStatus", "type": {"type": "record", "name": "ConsentStatus", "doc": "",
"fields": [{"name": "programmeConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"primaryFindingConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"secondaryFindingConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"carrierStatusConsent", "type": "boolean", "doc": "", "default": false}]}, "doc": ""}, {"name":
"additionalInformation", "type": ["null", {"type": "map", "values": "string"}], "doc": ""}, {"name":
"sampleId", "type": ["null", {"type": "array", "items": "string"}], "doc": ""}, {"name":
"assignedICD10", "type": ["null", "string"], "doc": ""}]}, "doc": ""}, {"name": "cancerSamples",
"type": {"type": "array", "items": {"type": "record", "name": "CancerSample", "fields": [{"name":
"sampleId", "type": "string", "doc": ""}, {"name": "labId", "type": ["null", "string"], "doc": ""},
{"name": "gelPhase", "type": ["null", {"type": "enum", "name": "GelPhase", "symbols": ["CRUK",
"OXFORD", "CLL", "IIP", "MAIN", "EXPT"]}], "doc": ""}, {"name": "sampleType", "type": {"type":
"enum", "name": "SampleType", "symbols": ["germline", "tumor"]}, "doc": ""}, {"name":
"sampleDiagnosis", "type": ["null", "string"], "doc": ""}, {"name": "tumorType", "type": ["null",
"string"], "doc": ""}, {"name": "tumorSubType", "type": ["null", "string"], "doc": ""}, {"name":
"preservationMethod", "type": ["null", {"type": "enum", "name": "PreservationMethod", "symbols":
["FFPE", "FF", "UNKNOWN", "BLOOD", "GL", "SALIVA", "LEUK"]}], "doc": ""}, {"name": "phase", "type":
["null", {"type": "enum", "name": "Phase", "symbols": ["PRIMARY", "METASTATIC", "RECURRENCE"]}],
"doc": ""}, {"name": "method", "type": ["null", {"type": "enum", "name": "Method", "symbols":
["RESECTION", "BIOPSY", "BLOOD"]}], "doc": ""}, {"name": "cellularity", "type": ["null", "string"],
"doc": ""}, {"name": "tumorContent", "type": ["null", "string"], "doc": ""}, {"name": "grade",
"type": ["null", "string"], "doc": ""}, {"name": "tnm_stage_version", "type": ["null", "string"],
"doc": ""}, {"name": "tmn_stage_grouping", "type": ["null", "string"], "doc": ""}]}}, "doc": ""},
{"name": "matchedSamples", "type": {"type": "array", "items": {"type": "record", "name":
"MatchedSamples", "doc": "", "fields": [{"name": "germlineSampleId", "type": ["null", "string"],
"doc": ""}, {"name": "tumorSampleId", "type": ["null", "string"], "doc": ""}]}}, "doc": ""}]}}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "cancerParticipant",
        "candidateStructuralVariants",
        "candidateVariants",
        "genePanelsCoverage",
        "genomicInterpretation",
        "interpretationRequestID",
        "interpretationRequestVersion",
        "referenceDatabasesVersions",
        "references",
        "reportingDate",
        "softwareVersions",
        "user",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'cancerParticipant': CancerParticipant,
            'candidateStructuralVariants': ReportedStructuralVariant,
            'candidateVariants': ReportedVariant,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'cancerParticipant': CancerParticipant,
            'candidateStructuralVariants': ReportedStructuralVariant,
            'candidateVariants': ReportedVariant,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'cancerParticipant', 'candidateStructuralVariants',
        'candidateVariants', 'genePanelsCoverage',
        'genomicInterpretation', 'interpretationRequestID',
        'interpretationRequestVersion', 'referenceDatabasesVersions',
        'references', 'reportingDate', 'softwareVersions', 'user'
    ]

    def __init__(self, **kwargs):
        self.cancerParticipant = kwargs.get(
            'cancerParticipant', CancerParticipant())
        self.candidateStructuralVariants = kwargs.get(
            'candidateStructuralVariants', None)
        self.candidateVariants = kwargs.get(
            'candidateVariants', None)
        self.genePanelsCoverage = kwargs.get(
            'genePanelsCoverage', None)
        self.genomicInterpretation = kwargs.get(
            'genomicInterpretation', None)
        self.interpretationRequestID = kwargs.get(
            'interpretationRequestID', None)
        self.interpretationRequestVersion = kwargs.get(
            'interpretationRequestVersion', None)
        self.referenceDatabasesVersions = kwargs.get(
            'referenceDatabasesVersions', None)
        self.references = kwargs.get(
            'references', None)
        self.reportingDate = kwargs.get(
            'reportingDate', None)
        self.softwareVersions = kwargs.get(
            'softwareVersions', None)
        self.user = kwargs.get(
            'user', None)


class ClinicalReportRD(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "ClinicalReportRD", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"interpretationRequestID", "type": "string", "doc": ""}, {"name": "interpretationRequestVersion",
"type": "string", "doc": ""}, {"name": "interpretationRequestAnalysisVersion", "type": "string",
"doc": ""}, {"name": "reportingDate", "type": "string", "doc": ""}, {"name": "user", "type":
"string", "doc": ""}, {"name": "candidateVariants", "type": {"type": "array", "items": {"type":
"record", "name": "ReportedVariant", "fields": [{"name": "versionControl", "type": {"type":
"record", "name": "VersionControl", "fields": [{"name": "GitVersionControl", "type": "string",
"doc": "", "default": "2.1.0"}]}, "doc": ""}, {"name": "chromosome", "type": "string", "doc": ""},
{"name": "dbSNPid", "type": ["null", "string"], "doc": ""}, {"name": "position", "type": "int",
"doc": ""}, {"name": "reference", "type": "string", "doc": ""}, {"name": "alternate", "type":
"string", "doc": ""}, {"name": "calledGenotypes", "type": {"type": "array", "items": {"type":
"record", "name": "CalledGenotype", "doc": "", "fields": [{"name": "gelId", "type": "string", "doc":
""}, {"name": "sampleId", "type": "string", "doc": ""}, {"name": "genotype", "type": "string",
"doc": ""}, {"name": "phaseSet", "type": ["null", "int"], "doc": ""}, {"name": "depthReference",
"type": ["null", "int"], "doc": ""}, {"name": "depthAlternate", "type": ["null", "int"], "doc": ""},
{"name": "copyNumber", "type": ["null", "int"], "doc": ""}]}}, "doc": ""}, {"name": "reportEvents",
"type": {"type": "array", "items": {"type": "record", "name": "ReportEvent", "fields": [{"name":
"reportEventId", "type": "string", "doc": ""}, {"name": "phenotype", "type": "string", "doc": ""},
{"name": "panelName", "type": ["null", "string"], "doc": ""}, {"name": "panelVersion", "type":
["null", "string"], "doc": ""}, {"name": "modeOfInheritance", "type": {"type": "enum", "name":
"ReportedModeOfInheritance", "doc": "", "symbols": ["monoallelic", "monoallelic_not_imprinted",
"monoallelic_maternally_imprinted", "monoallelic_paternally_imprinted", "biallelic",
"monoallelic_and_biallelic", "monoallelic_and_more_severe_biallelic", "xlinked_biallelic",
"xlinked_monoallelic", "mitochondrial", "unknown"]}, "doc": ""}, {"name": "genomicFeature", "type":
{"type": "record", "name": "GenomicFeature", "fields": [{"name": "featureType", "type": {"type":
"enum", "name": "FeatureTypes", "symbols": ["RegulatoryRegion", "Gene", "Transcript"]}, "doc": ""},
{"name": "ensemblId", "type": "string", "doc": ""}, {"name": "ids", "type": {"type": "map",
"values": "string"}, "doc": ""}]}, "doc": ""}, {"name": "penetrance", "type": {"type": "enum",
"name": "Penetrance", "doc": "", "symbols": ["complete", "incomplete"]}, "doc": ""}, {"name":
"score", "type": "float", "doc": ""}, {"name": "vendorSpecificScores", "type": ["null", {"type":
"map", "values": "float"}], "doc": ""}, {"name": "variantClassification", "type": ["null", {"type":
"enum", "name": "VariantClassification", "doc": "", "symbols": ["BENIGN", "LIKELY_BENIGN", "VUS",
"LIKELY_PATHOGENIC", "PATHOGENIC"]}], "doc": ""}, {"name": "fullyExplainsPhenotype", "type":
["null", "boolean"], "doc": ""}, {"name": "groupOfVariants", "type": ["null", "int"], "doc": ""},
{"name": "eventJustification", "type": "string", "doc": ""}, {"name": "tier", "type": ["null",
{"type": "enum", "name": "Tier", "doc": "", "symbols": ["NONE", "TIER1", "TIER2", "TIER3"]}], "doc":
""}]}}, "doc": ""}, {"name": "additionalTextualVariantAnnotations", "type": ["null", {"type": "map",
"values": "string"}], "doc": ""}, {"name": "additionalNumericVariantAnnotations", "type": ["null",
{"type": "map", "values": "float"}], "doc": ""}, {"name": "comments", "type": ["null", {"type":
"array", "items": "string"}], "doc": ""}]}}, "doc": ""}, {"name": "candidateStructuralVariants",
"type": {"type": "array", "items": {"type": "record", "name": "ReportedStructuralVariant", "fields":
[{"name": "versionControl", "type": "VersionControl", "doc": ""}, {"name": "chromosome", "type":
"string", "doc": ""}, {"name": "start", "type": "int", "doc": ""}, {"name": "end", "type": "int",
"doc": ""}, {"name": "type", "type": "string", "doc": ""}, {"name": "reference", "type": "string",
"doc": ""}, {"name": "alternate", "type": "string", "doc": ""}, {"name": "calledGenotypes", "type":
{"type": "array", "items": "CalledGenotype"}}, {"name": "reportEvents", "type": {"type": "array",
"items": "ReportEvent"}, "doc": ""}, {"name": "additionalTextualVariantAnnotations", "type":
["null", {"type": "map", "values": "string"}], "doc": ""}, {"name":
"additionalNumericVariantAnnotations", "type": ["null", {"type": "map", "values": "float"}], "doc":
""}, {"name": "comments", "type": ["null", {"type": "array", "items": "string"}], "doc": ""}]}},
"doc": ""}, {"name": "genomicInterpretation", "type": "string", "doc": ""}, {"name":
"supportingEvidence", "type": ["null", {"type": "array", "items": "string"}], "doc": ""}, {"name":
"referenceDatabasesVersions", "type": {"type": "map", "values": "string"}, "doc": ""}, {"name":
"softwareVersions", "type": {"type": "map", "values": "string"}, "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "candidateStructuralVariants",
        "candidateVariants",
        "genomicInterpretation",
        "interpretationRequestAnalysisVersion",
        "interpretationRequestID",
        "interpretationRequestVersion",
        "referenceDatabasesVersions",
        "reportingDate",
        "softwareVersions",
        "supportingEvidence",
        "user",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'candidateStructuralVariants': ReportedStructuralVariant,
            'candidateVariants': ReportedVariant,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'candidateStructuralVariants': ReportedStructuralVariant,
            'candidateVariants': ReportedVariant,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'candidateStructuralVariants', 'candidateVariants',
        'genomicInterpretation',
        'interpretationRequestAnalysisVersion',
        'interpretationRequestID', 'interpretationRequestVersion',
        'referenceDatabasesVersions', 'reportingDate',
        'softwareVersions', 'supportingEvidence', 'user'
    ]

    def __init__(self, **kwargs):
        self.candidateStructuralVariants = kwargs.get(
            'candidateStructuralVariants', None)
        self.candidateVariants = kwargs.get(
            'candidateVariants', None)
        self.genomicInterpretation = kwargs.get(
            'genomicInterpretation', None)
        self.interpretationRequestAnalysisVersion = kwargs.get(
            'interpretationRequestAnalysisVersion', None)
        self.interpretationRequestID = kwargs.get(
            'interpretationRequestID', None)
        self.interpretationRequestVersion = kwargs.get(
            'interpretationRequestVersion', None)
        self.referenceDatabasesVersions = kwargs.get(
            'referenceDatabasesVersions', None)
        self.reportingDate = kwargs.get(
            'reportingDate', None)
        self.softwareVersions = kwargs.get(
            'softwareVersions', None)
        self.supportingEvidence = kwargs.get(
            'supportingEvidence', None)
        self.user = kwargs.get(
            'user', None)


class ComplexGeneticPhenomena(object):
    """
    No documentation
    """
    mosaicism = "mosaicism"
    monosomy = "monosomy"
    disomy = "disomy"
    uniparental_disomy = "uniparental_disomy"
    trisomy = "trisomy"
    other_aneuploidy = "other_aneuploidy"

    def __hash__(self):
        return str(self).__hash__()


class ConsentStatus(ProtocolElement):
    """
    Consent Status
    """
    _schemaSource = """
{"type": "record", "name": "ConsentStatus", "namespace": "Gel_BioInf_Models", "doc": "", "fields":
[{"name": "programmeConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"primaryFindingConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"secondaryFindingConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"carrierStatusConsent", "type": "boolean", "doc": "", "default": false}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {}

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'carrierStatusConsent', 'primaryFindingConsent',
        'programmeConsent', 'secondaryFindingConsent'
    ]

    def __init__(self, **kwargs):
        self.carrierStatusConsent = kwargs.get(
            'carrierStatusConsent', False)
        self.primaryFindingConsent = kwargs.get(
            'primaryFindingConsent', False)
        self.programmeConsent = kwargs.get(
            'programmeConsent', False)
        self.secondaryFindingConsent = kwargs.get(
            'secondaryFindingConsent', False)


class CoverageSummary(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "CoverageSummary", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"wellId", "type": "string"}, {"name": "n", "type": "double"}, {"name": "mean", "type": "double"},
{"name": "sd", "type": "double"}, {"name": "pct25", "type": "double"}, {"name": "median", "type":
"double"}, {"name": "pct75", "type": "double"}, {"name": "scope", "type": "string"}, {"name":
"localRMSD", "type": "double"}, {"name": "COSMIC30Xcov", "type": "double"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "COSMIC30Xcov",
        "localRMSD",
        "mean",
        "median",
        "n",
        "pct25",
        "pct75",
        "scope",
        "sd",
        "wellId",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'COSMIC30Xcov', 'localRMSD', 'mean', 'median', 'n', 'pct25',
        'pct75', 'scope', 'sd', 'wellId'
    ]

    def __init__(self, **kwargs):
        self.COSMIC30Xcov = kwargs.get(
            'COSMIC30Xcov', None)
        self.localRMSD = kwargs.get(
            'localRMSD', None)
        self.mean = kwargs.get(
            'mean', None)
        self.median = kwargs.get(
            'median', None)
        self.n = kwargs.get(
            'n', None)
        self.pct25 = kwargs.get(
            'pct25', None)
        self.pct75 = kwargs.get(
            'pct75', None)
        self.scope = kwargs.get(
            'scope', None)
        self.sd = kwargs.get(
            'sd', None)
        self.wellId = kwargs.get(
            'wellId', None)


class Disorder(ProtocolElement):
    """
    This is quite GEL specific. This is the way is stored in
    ModelCatalogue and PanelApp. Currently all specific disease titles
    are assigned to a disease subgroup so really only specificDisease
    needs to be completed but we add the others for generality
    """
    _schemaSource = """
{"type": "record", "name": "Disorder", "namespace": "Gel_BioInf_Models", "doc": "", "fields":
[{"name": "diseaseGroup", "type": ["null", "string"], "doc": ""}, {"name": "diseaseSubGroup",
"type": ["null", "string"], "doc": ""}, {"name": "specificDisease", "type": ["null", "string"],
"doc": ""}, {"name": "ageOfOnset", "type": ["null", "string"], "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "ageOfOnset",
        "diseaseGroup",
        "diseaseSubGroup",
        "specificDisease",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'ageOfOnset', 'diseaseGroup', 'diseaseSubGroup',
        'specificDisease'
    ]

    def __init__(self, **kwargs):
        self.ageOfOnset = kwargs.get(
            'ageOfOnset', None)
        self.diseaseGroup = kwargs.get(
            'diseaseGroup', None)
        self.diseaseSubGroup = kwargs.get(
            'diseaseSubGroup', None)
        self.specificDisease = kwargs.get(
            'specificDisease', None)


class EthnicCategory(object):
    """
    This is the list of ethnics in ONS16  * `D`:  Mixed: White and
    Black Caribbean * `E`:  Mixed: White and Black African * `F`:
    Mixed: White and Asian * `G`:  Mixed: Any other mixed background *
    `A`:  White: British * `B`:  White: Irish * `C`:  White: Any other
    White background * `L`:  Asian or Asian British: Any other Asian
    background * `M`:  Black or Black British: Caribbean * `N`:  Black
    or Black British: African * `H`:  Asian or Asian British: Indian *
    `J`:  Asian or Asian British: Pakistani * `K`:  Asian or Asian
    British: Bangladeshi * `P`:  Black or Black British: Any other
    Black background * `S`:  Other Ethnic Groups: Any other ethnic
    group * `R`:  Other Ethnic Groups: Chinese * `Z`:  Not stated
    """
    D = "D"
    E = "E"
    F = "F"
    G = "G"
    A = "A"
    B = "B"
    C = "C"
    L = "L"
    M = "M"
    N = "N"
    H = "H"
    J = "J"
    K = "K"
    P = "P"
    S = "S"
    R = "R"
    Z = "Z"

    def __hash__(self):
        return str(self).__hash__()


class ExonCoverage(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "ExonCoverage", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"coverageSummary", "type": {"type": "array", "items": {"type": "record", "name": "CoverageSummary",
"fields": [{"name": "wellId", "type": "string"}, {"name": "n", "type": "double"}, {"name": "mean",
"type": "double"}, {"name": "sd", "type": "double"}, {"name": "pct25", "type": "double"}, {"name":
"median", "type": "double"}, {"name": "pct75", "type": "double"}, {"name": "scope", "type":
"string"}, {"name": "localRMSD", "type": "double"}, {"name": "COSMIC30Xcov", "type": "double"}]}}}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "coverageSummary",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'coverageSummary': CoverageSummary,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'coverageSummary': CoverageSummary,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'coverageSummary'
    ]

    def __init__(self, **kwargs):
        self.coverageSummary = kwargs.get(
            'coverageSummary', None)


class FeatureTypes(object):
    """
    No documentation
    """
    RegulatoryRegion = "RegulatoryRegion"
    Gene = "Gene"
    Transcript = "Transcript"

    def __hash__(self):
        return str(self).__hash__()


class File(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "File", "namespace": "Gel_BioInf_Models", "fields": [{"name": "SampleId",
"type": ["null", "string", {"type": "array", "items": "string"}], "doc": ""}, {"name": "URIFile",
"type": "string", "doc": ""}, {"name": "fileType", "type": {"type": "enum", "name": "FileType",
"doc": "", "symbols": ["BAM", "gVCF", "VCF_small", "VCF_CNV", "VCF_SV", "VCF_SV_CNV", "SVG", "TIER",
"ANN", "BigWig"]}}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "SampleId",
        "URIFile",
        "fileType",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'SampleId', 'URIFile', 'fileType'
    ]

    def __init__(self, **kwargs):
        self.SampleId = kwargs.get(
            'SampleId', None)
        self.URIFile = kwargs.get(
            'URIFile', None)
        self.fileType = kwargs.get(
            'fileType', None)


class FileType(object):
    """
    This defines a file This Record is defined by the sampleID and a
    URI Currently SampleID can be a single String or an array of
    strings if multiple samples are associated with the same file
    """
    BAM = "BAM"
    gVCF = "gVCF"
    VCF_small = "VCF_small"
    VCF_CNV = "VCF_CNV"
    VCF_SV = "VCF_SV"
    VCF_SV_CNV = "VCF_SV_CNV"
    SVG = "SVG"
    TIER = "TIER"
    ANN = "ANN"
    BigWig = "BigWig"

    def __hash__(self):
        return str(self).__hash__()


class GelAtGcDrop(ProtocolElement):
    """
    GEL AT/GC dropout calculation
    """
    _schemaSource = """
{"type": "record", "name": "GelAtGcDrop", "namespace": "Gel_BioInf_Models", "doc": "", "fields":
[{"name": "at_drop", "type": "double"}, {"name": "gc_drop", "type": "double"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "at_drop",
        "gc_drop",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'at_drop', 'gc_drop'
    ]

    def __init__(self, **kwargs):
        self.at_drop = kwargs.get(
            'at_drop', None)
        self.gc_drop = kwargs.get(
            'gc_drop', None)


class GelMetrics(ProtocolElement):
    """
    These are contractual metrics calculated by GEL  * `BaseDir` -
    Base directory of the data * `GbQ30NoDupsNoClip`: The number of
    missing genotypes * `perc_bases_ge_15x_mapQ_ge11`: Percentage of
    bases that have >=15x coverage and mapping quality >=11 *
    `InputDir`: Input dir of the data * `DirectoryType`: Illumina
    directory type version * `nBases_samtools`: number of bases from
    samtools * `FileRelativePath`: relative path of bam file *
    `md5checksum`: md5 of the delivery
    """
    _schemaSource = """
{"type": "record", "name": "GelMetrics", "namespace": "Gel_BioInf_Models", "doc": "", "fields":
[{"name": "BaseDir", "type": "string"}, {"name": "GbQ30NoDupsNoClip", "type": "double"}, {"name":
"perc_bases_ge_15x_mapQ_ge11", "type": "float"}, {"name": "InputDir", "type": "string"}, {"name":
"DirectoryType", "type": "string"}, {"name": "nBases_samtools", "type": "double"}, {"name":
"FileRelativePath", "type": "string"}, {"name": "md5checksum", "type": "string"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "BaseDir",
        "DirectoryType",
        "FileRelativePath",
        "GbQ30NoDupsNoClip",
        "InputDir",
        "md5checksum",
        "nBases_samtools",
        "perc_bases_ge_15x_mapQ_ge11",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'BaseDir', 'DirectoryType', 'FileRelativePath',
        'GbQ30NoDupsNoClip', 'InputDir', 'md5checksum',
        'nBases_samtools', 'perc_bases_ge_15x_mapQ_ge11'
    ]

    def __init__(self, **kwargs):
        self.BaseDir = kwargs.get(
            'BaseDir', None)
        self.DirectoryType = kwargs.get(
            'DirectoryType', None)
        self.FileRelativePath = kwargs.get(
            'FileRelativePath', None)
        self.GbQ30NoDupsNoClip = kwargs.get(
            'GbQ30NoDupsNoClip', None)
        self.InputDir = kwargs.get(
            'InputDir', None)
        self.md5checksum = kwargs.get(
            'md5checksum', None)
        self.nBases_samtools = kwargs.get(
            'nBases_samtools', None)
        self.perc_bases_ge_15x_mapQ_ge11 = kwargs.get(
            'perc_bases_ge_15x_mapQ_ge11', None)


class GelPhase(object):
    """
    No documentation
    """
    CRUK = "CRUK"
    OXFORD = "OXFORD"
    CLL = "CLL"
    IIP = "IIP"
    MAIN = "MAIN"
    EXPT = "EXPT"

    def __hash__(self):
        return str(self).__hash__()


class GenomicFeature(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "GenomicFeature", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"featureType", "type": {"type": "enum", "name": "FeatureTypes", "symbols": ["RegulatoryRegion",
"Gene", "Transcript"]}, "doc": ""}, {"name": "ensemblId", "type": "string", "doc": ""}, {"name":
"ids", "type": {"type": "map", "values": "string"}, "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "ensemblId",
        "featureType",
        "ids",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'ensemblId', 'featureType', 'ids'
    ]

    def __init__(self, **kwargs):
        self.ensemblId = kwargs.get(
            'ensemblId', None)
        self.featureType = kwargs.get(
            'featureType', None)
        self.ids = kwargs.get(
            'ids', None)


class GenomicFeatureCancer(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "GenomicFeatureCancer", "namespace": "Gel_BioInf_Models", "fields":
[{"name": "featureType", "type": {"type": "enum", "name": "FeatureTypes", "symbols":
["RegulatoryRegion", "Gene", "Transcript"]}, "doc": ""}, {"name": "ensemblId", "type": "string",
"doc": ""}, {"name": "refSeqTranscriptId", "type": "string", "doc": ""}, {"name": "refSeqProteinId",
"type": "string", "doc": ""}, {"name": "geneName", "type": "string", "doc": ""}, {"name":
"roleInCancer", "type": ["null", "string"], "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "ensemblId",
        "featureType",
        "geneName",
        "refSeqProteinId",
        "refSeqTranscriptId",
        "roleInCancer",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'ensemblId', 'featureType', 'geneName', 'refSeqProteinId',
        'refSeqTranscriptId', 'roleInCancer'
    ]

    def __init__(self, **kwargs):
        self.ensemblId = kwargs.get(
            'ensemblId', None)
        self.featureType = kwargs.get(
            'featureType', None)
        self.geneName = kwargs.get(
            'geneName', None)
        self.refSeqProteinId = kwargs.get(
            'refSeqProteinId', None)
        self.refSeqTranscriptId = kwargs.get(
            'refSeqTranscriptId', None)
        self.roleInCancer = kwargs.get(
            'roleInCancer', None)


class HpoTerm(ProtocolElement):
    """
    This defines an HPO term and its modifiers (possibly multiple) If
    HPO term presence is unknown we don't have a entry on the list
    """
    _schemaSource = """
{"type": "record", "name": "HpoTerm", "namespace": "Gel_BioInf_Models", "doc": "", "fields":
[{"name": "term", "type": "string", "doc": ""}, {"name": "termPresence", "type": ["null",
"boolean"], "doc": ""}, {"name": "modifiers", "type": ["null", {"type": "map", "values": "string"}],
"doc": ""}, {"name": "ageOfOnset", "type": ["null", "string"], "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "ageOfOnset",
        "modifiers",
        "term",
        "termPresence",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'ageOfOnset', 'modifiers', 'term', 'termPresence'
    ]

    def __init__(self, **kwargs):
        self.ageOfOnset = kwargs.get(
            'ageOfOnset', None)
        self.modifiers = kwargs.get(
            'modifiers', None)
        self.term = kwargs.get(
            'term', None)
        self.termPresence = kwargs.get(
            'termPresence', None)


class IlluminaSummaryCancerV2(ProtocolElement):
    """
    This is the summary provided by illumina for Somatic calls in V2
    (current) of their pipeline  Fields of interest include
    `PURITY_TUMOR_PURITY` and `PURITY_TUMOR_PLOIDY`
    """
    _schemaSource = """
{"type": "record", "name": "IlluminaSummaryCancerV2", "namespace": "Gel_BioInf_Models", "doc": "",
"fields": [{"name": "illumina_version", "type": {"type": "enum", "name": "IlluminaVersion",
"symbols": ["IlluminaSummaryV1", "IlluminaSummaryV2", "IlluminaSummaryCancerV2"]}}, {"name":
"VARIANTSTATS_STOP_GAINED_SNVS", "type": "double"}, {"name": "VARIANTSTATS_MATURE_MIRNA_DELETIONS",
"type": "double"}, {"name": "VARIANTSTATS_SYNONYMOUS_SNVS", "type": "double"}, {"name":
"BAMSTATS_NORMAL_PERCENT_BASES_GE_Q30", "type": "double"}, {"name":
"VARIANTSTATS_NON_SYNONYMOUS_INSERTIONS", "type": "double"}, {"name":
"VARIANTSTATS_SYNONYMOUS_INSERTIONS", "type": "double"}, {"name":
"BAMSTATS_TUMOR_PERCENT_BASES_GE_Q30", "type": "double"}, {"name": "PURITY_TUMOR_PLOIDY", "type":
["null", "double"]}, {"name": "SVSTATS_TANDEM_DUPLICATION_TOTAL", "type": "double"}, {"name":
"SVSTATS_INSERTION_NUMBER_IN_GENES", "type": "double"}, {"name": "VARIANTSTATS_TOTAL_DELETIONS",
"type": "double"}, {"name": "VARIANTSTATS_NUMBER_IN_CODING_REGIONS_SNVS", "type": "double"},
{"name": "VARIANTSTATS_FRAMESHIFT_INSERTIONS", "type": "double"}, {"name": "PURITY_TUMOR_PURITY",
"type": ["null", "double"]}, {"name": "SVSTATS_DELETION_NUMBER_IN_GENES", "type": "double"},
{"name": "VARIANTSTATS_TOTAL_INSERTIONS", "type": "double"}, {"name":
"VARIANTSTATS_NON_SYNONYMOUS_SNVS", "type": "double"}, {"name": "VARIANTSTATS_MATURE_MIRNA_SNVS",
"type": "double"}, {"name": "VARIANTSTATS_STOP_GAINED_INSERTIONS", "type": "double"}, {"name":
"VARIANTSTATS_FRAMESHIFT_DELETIONS", "type": "double"}, {"name":
"SVSTATS_TANDEM_DUPLICATION_NUMBER_IN_GENES", "type": "double"}, {"name":
"VARIANTSTATS_NUMBER_IN_EXONS_DELETIONS", "type": "double"}, {"name":
"VARIANTSTATS_NUMBER_IN_GENES_DELETIONS", "type": "double"}, {"name":
"VARIANTSTATS_UTR_REGION_DELETIONS", "type": "double"}, {"name":
"BAMSTATS_TUMOR_GIGABASES_PASSING_FILTER", "type": "double"}, {"name":
"VARIANTSTATS_DBSNP_INSERTIONS", "type": "double"}, {"name": "SVSTATS_CNV_TOTAL", "type": "double"},
{"name": "VARIANTSTATS_STOP_LOST_INSERTIONS", "type": "double"}, {"name":
"VARIANTSTATS_SPLICE_SITE_REGION_INSERTIONS", "type": "double"}, {"name": "SVSTATS_INVERSION_TOTAL",
"type": "double"}, {"name": "VARIANTSTATS_SYNONYMOUS_DELETIONS", "type": "double"}, {"name":
"VARIANTSTATS_STOP_LOST_SNVS", "type": "double"}, {"name":
"BAMSTATS_NORMAL_GIGABASES_PASSING_FILTER", "type": "double"}, {"name":
"VARIANTSTATS_MATURE_MIRNA_INSERTIONS", "type": "double"}, {"name":
"VARIANTSTATS_SPLICE_SITE_REGION_SNVS", "type": "double"}, {"name":
"VARIANTSTATS_NUMBER_IN_CODING_REGIONS_INSERTIONS", "type": "double"}, {"name":
"VARIANTSTATS_NUMBER_IN_GENES_SNVS", "type": "double"}, {"name":
"VARIANTSTATS_STOP_GAINED_DELETIONS", "type": "double"}, {"name":
"VARIANTSTATS_STOP_LOST_DELETIONS", "type": "double"}, {"name": "NORMAL_ID", "type": "string"},
{"name": "VARIANTSTATS_SPLICE_SITE_REGION_DELETIONS", "type": "double"}, {"name":
"VARIANTSTATS_UTR_REGION_INSERTIONS", "type": "double"}, {"name":
"VARIANTSTATS_NUMBER_IN_CODING_REGIONS_DELETIONS", "type": "double"}, {"name":
"SVSTATS_TRANSLOCATION_BREAKEND_TOTAL", "type": "double"}, {"name": "SVSTATS_CNV_NUMBER_IN_GENES",
"type": "double"}, {"name": "VARIANTSTATS_UTR_REGION_SNVS", "type": "double"}, {"name":
"VARIANTSTATS_NUMBER_IN_EXONS_INSERTIONS", "type": "double"}, {"name": "VARIANTSTATS_DBSNP_SNVS",
"type": "double"}, {"name": "VARIANTSTATS_DBSNP_DELETIONS", "type": "double"}, {"name":
"VARIANTSTATS_FRAMESHIFT_SNVS", "type": "double"}, {"name": "VARIANTSTATS_NUMBER_IN_EXONS_SNVS",
"type": "double"}, {"name": "VARIANTSTATS_NUMBER_IN_GENES_INSERTIONS", "type": "double"}, {"name":
"VARIANTSTATS_TOTAL_SNVS", "type": "double"}, {"name": "SVSTATS_DELETION_TOTAL", "type": "double"},
{"name": "SVSTATS_INVERSION_NUMBER_IN_GENES", "type": "double"}, {"name": "TUMOR_ID", "type":
"string"}, {"name": "SVSTATS_INSERTION_TOTAL", "type": "double"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "BAMSTATS_NORMAL_GIGABASES_PASSING_FILTER",
        "BAMSTATS_NORMAL_PERCENT_BASES_GE_Q30",
        "BAMSTATS_TUMOR_GIGABASES_PASSING_FILTER",
        "BAMSTATS_TUMOR_PERCENT_BASES_GE_Q30",
        "NORMAL_ID",
        "PURITY_TUMOR_PLOIDY",
        "PURITY_TUMOR_PURITY",
        "SVSTATS_CNV_NUMBER_IN_GENES",
        "SVSTATS_CNV_TOTAL",
        "SVSTATS_DELETION_NUMBER_IN_GENES",
        "SVSTATS_DELETION_TOTAL",
        "SVSTATS_INSERTION_NUMBER_IN_GENES",
        "SVSTATS_INSERTION_TOTAL",
        "SVSTATS_INVERSION_NUMBER_IN_GENES",
        "SVSTATS_INVERSION_TOTAL",
        "SVSTATS_TANDEM_DUPLICATION_NUMBER_IN_GENES",
        "SVSTATS_TANDEM_DUPLICATION_TOTAL",
        "SVSTATS_TRANSLOCATION_BREAKEND_TOTAL",
        "TUMOR_ID",
        "VARIANTSTATS_DBSNP_DELETIONS",
        "VARIANTSTATS_DBSNP_INSERTIONS",
        "VARIANTSTATS_DBSNP_SNVS",
        "VARIANTSTATS_FRAMESHIFT_DELETIONS",
        "VARIANTSTATS_FRAMESHIFT_INSERTIONS",
        "VARIANTSTATS_FRAMESHIFT_SNVS",
        "VARIANTSTATS_MATURE_MIRNA_DELETIONS",
        "VARIANTSTATS_MATURE_MIRNA_INSERTIONS",
        "VARIANTSTATS_MATURE_MIRNA_SNVS",
        "VARIANTSTATS_NON_SYNONYMOUS_INSERTIONS",
        "VARIANTSTATS_NON_SYNONYMOUS_SNVS",
        "VARIANTSTATS_NUMBER_IN_CODING_REGIONS_DELETIONS",
        "VARIANTSTATS_NUMBER_IN_CODING_REGIONS_INSERTIONS",
        "VARIANTSTATS_NUMBER_IN_CODING_REGIONS_SNVS",
        "VARIANTSTATS_NUMBER_IN_EXONS_DELETIONS",
        "VARIANTSTATS_NUMBER_IN_EXONS_INSERTIONS",
        "VARIANTSTATS_NUMBER_IN_EXONS_SNVS",
        "VARIANTSTATS_NUMBER_IN_GENES_DELETIONS",
        "VARIANTSTATS_NUMBER_IN_GENES_INSERTIONS",
        "VARIANTSTATS_NUMBER_IN_GENES_SNVS",
        "VARIANTSTATS_SPLICE_SITE_REGION_DELETIONS",
        "VARIANTSTATS_SPLICE_SITE_REGION_INSERTIONS",
        "VARIANTSTATS_SPLICE_SITE_REGION_SNVS",
        "VARIANTSTATS_STOP_GAINED_DELETIONS",
        "VARIANTSTATS_STOP_GAINED_INSERTIONS",
        "VARIANTSTATS_STOP_GAINED_SNVS",
        "VARIANTSTATS_STOP_LOST_DELETIONS",
        "VARIANTSTATS_STOP_LOST_INSERTIONS",
        "VARIANTSTATS_STOP_LOST_SNVS",
        "VARIANTSTATS_SYNONYMOUS_DELETIONS",
        "VARIANTSTATS_SYNONYMOUS_INSERTIONS",
        "VARIANTSTATS_SYNONYMOUS_SNVS",
        "VARIANTSTATS_TOTAL_DELETIONS",
        "VARIANTSTATS_TOTAL_INSERTIONS",
        "VARIANTSTATS_TOTAL_SNVS",
        "VARIANTSTATS_UTR_REGION_DELETIONS",
        "VARIANTSTATS_UTR_REGION_INSERTIONS",
        "VARIANTSTATS_UTR_REGION_SNVS",
        "illumina_version",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'BAMSTATS_NORMAL_GIGABASES_PASSING_FILTER',
        'BAMSTATS_NORMAL_PERCENT_BASES_GE_Q30',
        'BAMSTATS_TUMOR_GIGABASES_PASSING_FILTER',
        'BAMSTATS_TUMOR_PERCENT_BASES_GE_Q30', 'NORMAL_ID',
        'PURITY_TUMOR_PLOIDY', 'PURITY_TUMOR_PURITY',
        'SVSTATS_CNV_NUMBER_IN_GENES', 'SVSTATS_CNV_TOTAL',
        'SVSTATS_DELETION_NUMBER_IN_GENES', 'SVSTATS_DELETION_TOTAL',
        'SVSTATS_INSERTION_NUMBER_IN_GENES',
        'SVSTATS_INSERTION_TOTAL',
        'SVSTATS_INVERSION_NUMBER_IN_GENES',
        'SVSTATS_INVERSION_TOTAL',
        'SVSTATS_TANDEM_DUPLICATION_NUMBER_IN_GENES',
        'SVSTATS_TANDEM_DUPLICATION_TOTAL',
        'SVSTATS_TRANSLOCATION_BREAKEND_TOTAL', 'TUMOR_ID',
        'VARIANTSTATS_DBSNP_DELETIONS',
        'VARIANTSTATS_DBSNP_INSERTIONS', 'VARIANTSTATS_DBSNP_SNVS',
        'VARIANTSTATS_FRAMESHIFT_DELETIONS',
        'VARIANTSTATS_FRAMESHIFT_INSERTIONS',
        'VARIANTSTATS_FRAMESHIFT_SNVS',
        'VARIANTSTATS_MATURE_MIRNA_DELETIONS',
        'VARIANTSTATS_MATURE_MIRNA_INSERTIONS',
        'VARIANTSTATS_MATURE_MIRNA_SNVS',
        'VARIANTSTATS_NON_SYNONYMOUS_INSERTIONS',
        'VARIANTSTATS_NON_SYNONYMOUS_SNVS',
        'VARIANTSTATS_NUMBER_IN_CODING_REGIONS_DELETIONS',
        'VARIANTSTATS_NUMBER_IN_CODING_REGIONS_INSERTIONS',
        'VARIANTSTATS_NUMBER_IN_CODING_REGIONS_SNVS',
        'VARIANTSTATS_NUMBER_IN_EXONS_DELETIONS',
        'VARIANTSTATS_NUMBER_IN_EXONS_INSERTIONS',
        'VARIANTSTATS_NUMBER_IN_EXONS_SNVS',
        'VARIANTSTATS_NUMBER_IN_GENES_DELETIONS',
        'VARIANTSTATS_NUMBER_IN_GENES_INSERTIONS',
        'VARIANTSTATS_NUMBER_IN_GENES_SNVS',
        'VARIANTSTATS_SPLICE_SITE_REGION_DELETIONS',
        'VARIANTSTATS_SPLICE_SITE_REGION_INSERTIONS',
        'VARIANTSTATS_SPLICE_SITE_REGION_SNVS',
        'VARIANTSTATS_STOP_GAINED_DELETIONS',
        'VARIANTSTATS_STOP_GAINED_INSERTIONS',
        'VARIANTSTATS_STOP_GAINED_SNVS',
        'VARIANTSTATS_STOP_LOST_DELETIONS',
        'VARIANTSTATS_STOP_LOST_INSERTIONS',
        'VARIANTSTATS_STOP_LOST_SNVS',
        'VARIANTSTATS_SYNONYMOUS_DELETIONS',
        'VARIANTSTATS_SYNONYMOUS_INSERTIONS',
        'VARIANTSTATS_SYNONYMOUS_SNVS',
        'VARIANTSTATS_TOTAL_DELETIONS',
        'VARIANTSTATS_TOTAL_INSERTIONS', 'VARIANTSTATS_TOTAL_SNVS',
        'VARIANTSTATS_UTR_REGION_DELETIONS',
        'VARIANTSTATS_UTR_REGION_INSERTIONS',
        'VARIANTSTATS_UTR_REGION_SNVS', 'illumina_version'
    ]

    def __init__(self, **kwargs):
        self.BAMSTATS_NORMAL_GIGABASES_PASSING_FILTER = kwargs.get(
            'BAMSTATS_NORMAL_GIGABASES_PASSING_FILTER', None)
        self.BAMSTATS_NORMAL_PERCENT_BASES_GE_Q30 = kwargs.get(
            'BAMSTATS_NORMAL_PERCENT_BASES_GE_Q30', None)
        self.BAMSTATS_TUMOR_GIGABASES_PASSING_FILTER = kwargs.get(
            'BAMSTATS_TUMOR_GIGABASES_PASSING_FILTER', None)
        self.BAMSTATS_TUMOR_PERCENT_BASES_GE_Q30 = kwargs.get(
            'BAMSTATS_TUMOR_PERCENT_BASES_GE_Q30', None)
        self.NORMAL_ID = kwargs.get(
            'NORMAL_ID', None)
        self.PURITY_TUMOR_PLOIDY = kwargs.get(
            'PURITY_TUMOR_PLOIDY', None)
        self.PURITY_TUMOR_PURITY = kwargs.get(
            'PURITY_TUMOR_PURITY', None)
        self.SVSTATS_CNV_NUMBER_IN_GENES = kwargs.get(
            'SVSTATS_CNV_NUMBER_IN_GENES', None)
        self.SVSTATS_CNV_TOTAL = kwargs.get(
            'SVSTATS_CNV_TOTAL', None)
        self.SVSTATS_DELETION_NUMBER_IN_GENES = kwargs.get(
            'SVSTATS_DELETION_NUMBER_IN_GENES', None)
        self.SVSTATS_DELETION_TOTAL = kwargs.get(
            'SVSTATS_DELETION_TOTAL', None)
        self.SVSTATS_INSERTION_NUMBER_IN_GENES = kwargs.get(
            'SVSTATS_INSERTION_NUMBER_IN_GENES', None)
        self.SVSTATS_INSERTION_TOTAL = kwargs.get(
            'SVSTATS_INSERTION_TOTAL', None)
        self.SVSTATS_INVERSION_NUMBER_IN_GENES = kwargs.get(
            'SVSTATS_INVERSION_NUMBER_IN_GENES', None)
        self.SVSTATS_INVERSION_TOTAL = kwargs.get(
            'SVSTATS_INVERSION_TOTAL', None)
        self.SVSTATS_TANDEM_DUPLICATION_NUMBER_IN_GENES = kwargs.get(
            'SVSTATS_TANDEM_DUPLICATION_NUMBER_IN_GENES', None)
        self.SVSTATS_TANDEM_DUPLICATION_TOTAL = kwargs.get(
            'SVSTATS_TANDEM_DUPLICATION_TOTAL', None)
        self.SVSTATS_TRANSLOCATION_BREAKEND_TOTAL = kwargs.get(
            'SVSTATS_TRANSLOCATION_BREAKEND_TOTAL', None)
        self.TUMOR_ID = kwargs.get(
            'TUMOR_ID', None)
        self.VARIANTSTATS_DBSNP_DELETIONS = kwargs.get(
            'VARIANTSTATS_DBSNP_DELETIONS', None)
        self.VARIANTSTATS_DBSNP_INSERTIONS = kwargs.get(
            'VARIANTSTATS_DBSNP_INSERTIONS', None)
        self.VARIANTSTATS_DBSNP_SNVS = kwargs.get(
            'VARIANTSTATS_DBSNP_SNVS', None)
        self.VARIANTSTATS_FRAMESHIFT_DELETIONS = kwargs.get(
            'VARIANTSTATS_FRAMESHIFT_DELETIONS', None)
        self.VARIANTSTATS_FRAMESHIFT_INSERTIONS = kwargs.get(
            'VARIANTSTATS_FRAMESHIFT_INSERTIONS', None)
        self.VARIANTSTATS_FRAMESHIFT_SNVS = kwargs.get(
            'VARIANTSTATS_FRAMESHIFT_SNVS', None)
        self.VARIANTSTATS_MATURE_MIRNA_DELETIONS = kwargs.get(
            'VARIANTSTATS_MATURE_MIRNA_DELETIONS', None)
        self.VARIANTSTATS_MATURE_MIRNA_INSERTIONS = kwargs.get(
            'VARIANTSTATS_MATURE_MIRNA_INSERTIONS', None)
        self.VARIANTSTATS_MATURE_MIRNA_SNVS = kwargs.get(
            'VARIANTSTATS_MATURE_MIRNA_SNVS', None)
        self.VARIANTSTATS_NON_SYNONYMOUS_INSERTIONS = kwargs.get(
            'VARIANTSTATS_NON_SYNONYMOUS_INSERTIONS', None)
        self.VARIANTSTATS_NON_SYNONYMOUS_SNVS = kwargs.get(
            'VARIANTSTATS_NON_SYNONYMOUS_SNVS', None)
        self.VARIANTSTATS_NUMBER_IN_CODING_REGIONS_DELETIONS = kwargs.get(
            'VARIANTSTATS_NUMBER_IN_CODING_REGIONS_DELETIONS', None)
        self.VARIANTSTATS_NUMBER_IN_CODING_REGIONS_INSERTIONS = kwargs.get(
            'VARIANTSTATS_NUMBER_IN_CODING_REGIONS_INSERTIONS', None)
        self.VARIANTSTATS_NUMBER_IN_CODING_REGIONS_SNVS = kwargs.get(
            'VARIANTSTATS_NUMBER_IN_CODING_REGIONS_SNVS', None)
        self.VARIANTSTATS_NUMBER_IN_EXONS_DELETIONS = kwargs.get(
            'VARIANTSTATS_NUMBER_IN_EXONS_DELETIONS', None)
        self.VARIANTSTATS_NUMBER_IN_EXONS_INSERTIONS = kwargs.get(
            'VARIANTSTATS_NUMBER_IN_EXONS_INSERTIONS', None)
        self.VARIANTSTATS_NUMBER_IN_EXONS_SNVS = kwargs.get(
            'VARIANTSTATS_NUMBER_IN_EXONS_SNVS', None)
        self.VARIANTSTATS_NUMBER_IN_GENES_DELETIONS = kwargs.get(
            'VARIANTSTATS_NUMBER_IN_GENES_DELETIONS', None)
        self.VARIANTSTATS_NUMBER_IN_GENES_INSERTIONS = kwargs.get(
            'VARIANTSTATS_NUMBER_IN_GENES_INSERTIONS', None)
        self.VARIANTSTATS_NUMBER_IN_GENES_SNVS = kwargs.get(
            'VARIANTSTATS_NUMBER_IN_GENES_SNVS', None)
        self.VARIANTSTATS_SPLICE_SITE_REGION_DELETIONS = kwargs.get(
            'VARIANTSTATS_SPLICE_SITE_REGION_DELETIONS', None)
        self.VARIANTSTATS_SPLICE_SITE_REGION_INSERTIONS = kwargs.get(
            'VARIANTSTATS_SPLICE_SITE_REGION_INSERTIONS', None)
        self.VARIANTSTATS_SPLICE_SITE_REGION_SNVS = kwargs.get(
            'VARIANTSTATS_SPLICE_SITE_REGION_SNVS', None)
        self.VARIANTSTATS_STOP_GAINED_DELETIONS = kwargs.get(
            'VARIANTSTATS_STOP_GAINED_DELETIONS', None)
        self.VARIANTSTATS_STOP_GAINED_INSERTIONS = kwargs.get(
            'VARIANTSTATS_STOP_GAINED_INSERTIONS', None)
        self.VARIANTSTATS_STOP_GAINED_SNVS = kwargs.get(
            'VARIANTSTATS_STOP_GAINED_SNVS', None)
        self.VARIANTSTATS_STOP_LOST_DELETIONS = kwargs.get(
            'VARIANTSTATS_STOP_LOST_DELETIONS', None)
        self.VARIANTSTATS_STOP_LOST_INSERTIONS = kwargs.get(
            'VARIANTSTATS_STOP_LOST_INSERTIONS', None)
        self.VARIANTSTATS_STOP_LOST_SNVS = kwargs.get(
            'VARIANTSTATS_STOP_LOST_SNVS', None)
        self.VARIANTSTATS_SYNONYMOUS_DELETIONS = kwargs.get(
            'VARIANTSTATS_SYNONYMOUS_DELETIONS', None)
        self.VARIANTSTATS_SYNONYMOUS_INSERTIONS = kwargs.get(
            'VARIANTSTATS_SYNONYMOUS_INSERTIONS', None)
        self.VARIANTSTATS_SYNONYMOUS_SNVS = kwargs.get(
            'VARIANTSTATS_SYNONYMOUS_SNVS', None)
        self.VARIANTSTATS_TOTAL_DELETIONS = kwargs.get(
            'VARIANTSTATS_TOTAL_DELETIONS', None)
        self.VARIANTSTATS_TOTAL_INSERTIONS = kwargs.get(
            'VARIANTSTATS_TOTAL_INSERTIONS', None)
        self.VARIANTSTATS_TOTAL_SNVS = kwargs.get(
            'VARIANTSTATS_TOTAL_SNVS', None)
        self.VARIANTSTATS_UTR_REGION_DELETIONS = kwargs.get(
            'VARIANTSTATS_UTR_REGION_DELETIONS', None)
        self.VARIANTSTATS_UTR_REGION_INSERTIONS = kwargs.get(
            'VARIANTSTATS_UTR_REGION_INSERTIONS', None)
        self.VARIANTSTATS_UTR_REGION_SNVS = kwargs.get(
            'VARIANTSTATS_UTR_REGION_SNVS', None)
        self.illumina_version = kwargs.get(
            'illumina_version', None)


class IlluminaSummaryV1(ProtocolElement):
    """
    This is the summary provided by illumina for all V1 (old version)
    of their pipeline
    """
    _schemaSource = """
{"type": "record", "name": "IlluminaSummaryV1", "namespace": "Gel_BioInf_Models", "doc": "",
"fields": [{"name": "illumina_version", "type": {"type": "enum", "name": "IlluminaVersion",
"symbols": ["IlluminaSummaryV1", "IlluminaSummaryV2", "IlluminaSummaryCancerV2"]}}, {"name":
"PERCENT_Q30_R1", "type": "double"}, {"name": "PERCENT_Q30_R2", "type": "double"}, {"name":
"PERCENT_ALIGNED_R1", "type": "double"}, {"name": "INVERSIONCOUNT", "type": "double"}, {"name":
"DELETIONS_PERCENT_FOUND_IN_DBSNP", "type": "double"}, {"name": "PERCENT_ALIGNED_R2", "type":
"double"}, {"name": "INDELS_PERCENT_FOUND_IN_DBSNP", "type": "double"}, {"name": "STOPGAINEDINS",
"type": "double"}, {"name": "SYNONYMOUSSNVS", "type": "double"}, {"name": "SPLICESITEREGIONDEL",
"type": "double"}, {"name": "STOPLOSTSNVS", "type": "double"}, {"name": "SYNONYMOUSINS", "type":
"double"}, {"name": "SYNONYMOUSDEL", "type": "double"}, {"name": "NUMSNVSINCODINGREGIONS", "type":
"double"}, {"name": "STOPGAINEDSNVS", "type": "double"}, {"name": "SAMPLE_ID", "type": "string"},
{"name": "NUMDELINEXONS", "type": "double"}, {"name": "TANDEMDUPLICATIONPERCENTINGENES", "type":
"double"}, {"name": "SNVS_ALL", "type": "double"}, {"name": "CNVNUMBERINGENES", "type": "double"},
{"name": "CNVPERCENTINGENES", "type": "double"}, {"name": "NUMDELINGENES", "type": "double"},
{"name": "NUMSNVSINGENES", "type": "double"}, {"name": "INSERTIONS_PERCENT_FOUND_IN_DBSNP", "type":
"double"}, {"name": "SPLICESITEREGIONSNVS", "type": "double"}, {"name": "INSERTIONS", "type":
"double"}, {"name": "FRAGMENT_LENGTH_MIN", "type": "double"}, {"name": "STOPLOSTDEL", "type":
"double"}, {"name": "FRAGMENT_LENGTH_SD", "type": "double"}, {"name": "NONSYNONYMOUSINS", "type":
"double"}, {"name": "FRAGMENT_LENGTH_MAX", "type": "double"}, {"name": "FRAMESHIFTDEL", "type":
"double"}, {"name": "DELETIONPERCENTINGENES", "type": "double"}, {"name": "INDEL_HET_HOM_RATIO",
"type": "double"}, {"name": "UTRREGIONDEL", "type": "double"}, {"name": "RUNFOLDER", "type":
"string"}, {"name": "MATUREMIRNAINS", "type": "double"}, {"name": "INVERSIONNUMBERINGENES", "type":
"double"}, {"name": "NUMINSINCODINGREGIONS", "type": "double"}, {"name": "INSERTIONPERCENTINGENES",
"type": "double"}, {"name": "CALLABLE_AUTOSOMAL_FRACTION", "type": "double"}, {"name":
"NONSYNONYMOUSDEL", "type": "double"}, {"name": "FRAMESHIFTSNVS", "type": "double"}, {"name":
"SVSTATISTICSFLAG", "type": "string"}, {"name": "CNVCOUNT", "type": "double"}, {"name":
"NONSYNONYMOUSSNVS", "type": "double"}, {"name": "DELETIONCOUNT", "type": "double"}, {"name":
"CNVSTATISTICSFLAG", "type": "string"}, {"name": "INSERTIONNUMBERINGENES", "type": "double"},
{"name": "STOPGAINEDDEL", "type": "double"}, {"name": "MISMATCHRATE_R1", "type": "double"}, {"name":
"SNVS", "type": "double"}, {"name": "SNV_HET_HOM_RATIO", "type": "double"}, {"name":
"SPLICESITEREGIONINS", "type": "double"}, {"name": "NUMINSINEXONS", "type": "double"}, {"name":
"MISMATCHRATE_R2", "type": "double"}, {"name": "NUMDELINCODINGREGIONS", "type": "double"}, {"name":
"INSERTION_HET_HOM_RATIO", "type": "double"}, {"name": "NUMINSINGENES", "type": "double"}, {"name":
"SNV_TS_TV_RATIO", "type": "double"}, {"name": "TANDEMDUPLICATIONNUMBERINGENES", "type": "double"},
{"name": "SNVS_PERCENT_FOUND_IN_DBSNP", "type": "double"}, {"name": "UTRREGIONINS", "type":
"double"}, {"name": "MATUREMIRNASNVS", "type": "double"}, {"name": "DELETIONNUMBERINGENES", "type":
"double"}, {"name": "DELETIONS", "type": "double"}, {"name": "TOTALNUMBERINS", "type": "double"},
{"name": "TANDEMDUPLICATIONCOUNT", "type": "double"}, {"name": "PAIRED_END", "type": "string"},
{"name": "UTRREGIONSNVS", "type": "double"}, {"name": "INDELS", "type": "double"}, {"name":
"NUMSNVSINEXONS", "type": "double"}, {"name": "FRAMESHIFTINS", "type": "double"}, {"name":
"TOTALNUMBERSNVS", "type": "double"}, {"name": "DELETIONS_HET_HOM_RATIO", "type": "double"},
{"name": "FRAGMENT_LENGTH_MEDIAN", "type": "double"}, {"name": "REFERENCE_GENOME", "type":
"string"}, {"name": "SMALLVARIANTSTATISTICSFLAG", "type": "string"}, {"name": "TOTALNUMBERDEL",
"type": "double"}, {"name": "MATUREMIRNADEL", "type": "double"}, {"name": "INVERSIONPERCENTINGENES",
"type": "double"}, {"name": "STOPLOSTINS", "type": "double"}, {"name": "INSERTIONCOUNT", "type":
"double"}, {"name": "PERCENT_DUPLICATE_PAIRED_READS", "type": "double"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "CALLABLE_AUTOSOMAL_FRACTION",
        "CNVCOUNT",
        "CNVNUMBERINGENES",
        "CNVPERCENTINGENES",
        "CNVSTATISTICSFLAG",
        "DELETIONCOUNT",
        "DELETIONNUMBERINGENES",
        "DELETIONPERCENTINGENES",
        "DELETIONS",
        "DELETIONS_HET_HOM_RATIO",
        "DELETIONS_PERCENT_FOUND_IN_DBSNP",
        "FRAGMENT_LENGTH_MAX",
        "FRAGMENT_LENGTH_MEDIAN",
        "FRAGMENT_LENGTH_MIN",
        "FRAGMENT_LENGTH_SD",
        "FRAMESHIFTDEL",
        "FRAMESHIFTINS",
        "FRAMESHIFTSNVS",
        "INDELS",
        "INDELS_PERCENT_FOUND_IN_DBSNP",
        "INDEL_HET_HOM_RATIO",
        "INSERTIONCOUNT",
        "INSERTIONNUMBERINGENES",
        "INSERTIONPERCENTINGENES",
        "INSERTIONS",
        "INSERTIONS_PERCENT_FOUND_IN_DBSNP",
        "INSERTION_HET_HOM_RATIO",
        "INVERSIONCOUNT",
        "INVERSIONNUMBERINGENES",
        "INVERSIONPERCENTINGENES",
        "MATUREMIRNADEL",
        "MATUREMIRNAINS",
        "MATUREMIRNASNVS",
        "MISMATCHRATE_R1",
        "MISMATCHRATE_R2",
        "NONSYNONYMOUSDEL",
        "NONSYNONYMOUSINS",
        "NONSYNONYMOUSSNVS",
        "NUMDELINCODINGREGIONS",
        "NUMDELINEXONS",
        "NUMDELINGENES",
        "NUMINSINCODINGREGIONS",
        "NUMINSINEXONS",
        "NUMINSINGENES",
        "NUMSNVSINCODINGREGIONS",
        "NUMSNVSINEXONS",
        "NUMSNVSINGENES",
        "PAIRED_END",
        "PERCENT_ALIGNED_R1",
        "PERCENT_ALIGNED_R2",
        "PERCENT_DUPLICATE_PAIRED_READS",
        "PERCENT_Q30_R1",
        "PERCENT_Q30_R2",
        "REFERENCE_GENOME",
        "RUNFOLDER",
        "SAMPLE_ID",
        "SMALLVARIANTSTATISTICSFLAG",
        "SNVS",
        "SNVS_ALL",
        "SNVS_PERCENT_FOUND_IN_DBSNP",
        "SNV_HET_HOM_RATIO",
        "SNV_TS_TV_RATIO",
        "SPLICESITEREGIONDEL",
        "SPLICESITEREGIONINS",
        "SPLICESITEREGIONSNVS",
        "STOPGAINEDDEL",
        "STOPGAINEDINS",
        "STOPGAINEDSNVS",
        "STOPLOSTDEL",
        "STOPLOSTINS",
        "STOPLOSTSNVS",
        "SVSTATISTICSFLAG",
        "SYNONYMOUSDEL",
        "SYNONYMOUSINS",
        "SYNONYMOUSSNVS",
        "TANDEMDUPLICATIONCOUNT",
        "TANDEMDUPLICATIONNUMBERINGENES",
        "TANDEMDUPLICATIONPERCENTINGENES",
        "TOTALNUMBERDEL",
        "TOTALNUMBERINS",
        "TOTALNUMBERSNVS",
        "UTRREGIONDEL",
        "UTRREGIONINS",
        "UTRREGIONSNVS",
        "illumina_version",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'CALLABLE_AUTOSOMAL_FRACTION', 'CNVCOUNT', 'CNVNUMBERINGENES',
        'CNVPERCENTINGENES', 'CNVSTATISTICSFLAG', 'DELETIONCOUNT',
        'DELETIONNUMBERINGENES', 'DELETIONPERCENTINGENES',
        'DELETIONS', 'DELETIONS_HET_HOM_RATIO',
        'DELETIONS_PERCENT_FOUND_IN_DBSNP', 'FRAGMENT_LENGTH_MAX',
        'FRAGMENT_LENGTH_MEDIAN', 'FRAGMENT_LENGTH_MIN',
        'FRAGMENT_LENGTH_SD', 'FRAMESHIFTDEL', 'FRAMESHIFTINS',
        'FRAMESHIFTSNVS', 'INDELS', 'INDELS_PERCENT_FOUND_IN_DBSNP',
        'INDEL_HET_HOM_RATIO', 'INSERTIONCOUNT',
        'INSERTIONNUMBERINGENES', 'INSERTIONPERCENTINGENES',
        'INSERTIONS', 'INSERTIONS_PERCENT_FOUND_IN_DBSNP',
        'INSERTION_HET_HOM_RATIO', 'INVERSIONCOUNT',
        'INVERSIONNUMBERINGENES', 'INVERSIONPERCENTINGENES',
        'MATUREMIRNADEL', 'MATUREMIRNAINS', 'MATUREMIRNASNVS',
        'MISMATCHRATE_R1', 'MISMATCHRATE_R2', 'NONSYNONYMOUSDEL',
        'NONSYNONYMOUSINS', 'NONSYNONYMOUSSNVS',
        'NUMDELINCODINGREGIONS', 'NUMDELINEXONS', 'NUMDELINGENES',
        'NUMINSINCODINGREGIONS', 'NUMINSINEXONS', 'NUMINSINGENES',
        'NUMSNVSINCODINGREGIONS', 'NUMSNVSINEXONS', 'NUMSNVSINGENES',
        'PAIRED_END', 'PERCENT_ALIGNED_R1', 'PERCENT_ALIGNED_R2',
        'PERCENT_DUPLICATE_PAIRED_READS', 'PERCENT_Q30_R1',
        'PERCENT_Q30_R2', 'REFERENCE_GENOME', 'RUNFOLDER',
        'SAMPLE_ID', 'SMALLVARIANTSTATISTICSFLAG', 'SNVS', 'SNVS_ALL',
        'SNVS_PERCENT_FOUND_IN_DBSNP', 'SNV_HET_HOM_RATIO',
        'SNV_TS_TV_RATIO', 'SPLICESITEREGIONDEL',
        'SPLICESITEREGIONINS', 'SPLICESITEREGIONSNVS',
        'STOPGAINEDDEL', 'STOPGAINEDINS', 'STOPGAINEDSNVS',
        'STOPLOSTDEL', 'STOPLOSTINS', 'STOPLOSTSNVS',
        'SVSTATISTICSFLAG', 'SYNONYMOUSDEL', 'SYNONYMOUSINS',
        'SYNONYMOUSSNVS', 'TANDEMDUPLICATIONCOUNT',
        'TANDEMDUPLICATIONNUMBERINGENES',
        'TANDEMDUPLICATIONPERCENTINGENES', 'TOTALNUMBERDEL',
        'TOTALNUMBERINS', 'TOTALNUMBERSNVS', 'UTRREGIONDEL',
        'UTRREGIONINS', 'UTRREGIONSNVS', 'illumina_version'
    ]

    def __init__(self, **kwargs):
        self.CALLABLE_AUTOSOMAL_FRACTION = kwargs.get(
            'CALLABLE_AUTOSOMAL_FRACTION', None)
        self.CNVCOUNT = kwargs.get(
            'CNVCOUNT', None)
        self.CNVNUMBERINGENES = kwargs.get(
            'CNVNUMBERINGENES', None)
        self.CNVPERCENTINGENES = kwargs.get(
            'CNVPERCENTINGENES', None)
        self.CNVSTATISTICSFLAG = kwargs.get(
            'CNVSTATISTICSFLAG', None)
        self.DELETIONCOUNT = kwargs.get(
            'DELETIONCOUNT', None)
        self.DELETIONNUMBERINGENES = kwargs.get(
            'DELETIONNUMBERINGENES', None)
        self.DELETIONPERCENTINGENES = kwargs.get(
            'DELETIONPERCENTINGENES', None)
        self.DELETIONS = kwargs.get(
            'DELETIONS', None)
        self.DELETIONS_HET_HOM_RATIO = kwargs.get(
            'DELETIONS_HET_HOM_RATIO', None)
        self.DELETIONS_PERCENT_FOUND_IN_DBSNP = kwargs.get(
            'DELETIONS_PERCENT_FOUND_IN_DBSNP', None)
        self.FRAGMENT_LENGTH_MAX = kwargs.get(
            'FRAGMENT_LENGTH_MAX', None)
        self.FRAGMENT_LENGTH_MEDIAN = kwargs.get(
            'FRAGMENT_LENGTH_MEDIAN', None)
        self.FRAGMENT_LENGTH_MIN = kwargs.get(
            'FRAGMENT_LENGTH_MIN', None)
        self.FRAGMENT_LENGTH_SD = kwargs.get(
            'FRAGMENT_LENGTH_SD', None)
        self.FRAMESHIFTDEL = kwargs.get(
            'FRAMESHIFTDEL', None)
        self.FRAMESHIFTINS = kwargs.get(
            'FRAMESHIFTINS', None)
        self.FRAMESHIFTSNVS = kwargs.get(
            'FRAMESHIFTSNVS', None)
        self.INDELS = kwargs.get(
            'INDELS', None)
        self.INDELS_PERCENT_FOUND_IN_DBSNP = kwargs.get(
            'INDELS_PERCENT_FOUND_IN_DBSNP', None)
        self.INDEL_HET_HOM_RATIO = kwargs.get(
            'INDEL_HET_HOM_RATIO', None)
        self.INSERTIONCOUNT = kwargs.get(
            'INSERTIONCOUNT', None)
        self.INSERTIONNUMBERINGENES = kwargs.get(
            'INSERTIONNUMBERINGENES', None)
        self.INSERTIONPERCENTINGENES = kwargs.get(
            'INSERTIONPERCENTINGENES', None)
        self.INSERTIONS = kwargs.get(
            'INSERTIONS', None)
        self.INSERTIONS_PERCENT_FOUND_IN_DBSNP = kwargs.get(
            'INSERTIONS_PERCENT_FOUND_IN_DBSNP', None)
        self.INSERTION_HET_HOM_RATIO = kwargs.get(
            'INSERTION_HET_HOM_RATIO', None)
        self.INVERSIONCOUNT = kwargs.get(
            'INVERSIONCOUNT', None)
        self.INVERSIONNUMBERINGENES = kwargs.get(
            'INVERSIONNUMBERINGENES', None)
        self.INVERSIONPERCENTINGENES = kwargs.get(
            'INVERSIONPERCENTINGENES', None)
        self.MATUREMIRNADEL = kwargs.get(
            'MATUREMIRNADEL', None)
        self.MATUREMIRNAINS = kwargs.get(
            'MATUREMIRNAINS', None)
        self.MATUREMIRNASNVS = kwargs.get(
            'MATUREMIRNASNVS', None)
        self.MISMATCHRATE_R1 = kwargs.get(
            'MISMATCHRATE_R1', None)
        self.MISMATCHRATE_R2 = kwargs.get(
            'MISMATCHRATE_R2', None)
        self.NONSYNONYMOUSDEL = kwargs.get(
            'NONSYNONYMOUSDEL', None)
        self.NONSYNONYMOUSINS = kwargs.get(
            'NONSYNONYMOUSINS', None)
        self.NONSYNONYMOUSSNVS = kwargs.get(
            'NONSYNONYMOUSSNVS', None)
        self.NUMDELINCODINGREGIONS = kwargs.get(
            'NUMDELINCODINGREGIONS', None)
        self.NUMDELINEXONS = kwargs.get(
            'NUMDELINEXONS', None)
        self.NUMDELINGENES = kwargs.get(
            'NUMDELINGENES', None)
        self.NUMINSINCODINGREGIONS = kwargs.get(
            'NUMINSINCODINGREGIONS', None)
        self.NUMINSINEXONS = kwargs.get(
            'NUMINSINEXONS', None)
        self.NUMINSINGENES = kwargs.get(
            'NUMINSINGENES', None)
        self.NUMSNVSINCODINGREGIONS = kwargs.get(
            'NUMSNVSINCODINGREGIONS', None)
        self.NUMSNVSINEXONS = kwargs.get(
            'NUMSNVSINEXONS', None)
        self.NUMSNVSINGENES = kwargs.get(
            'NUMSNVSINGENES', None)
        self.PAIRED_END = kwargs.get(
            'PAIRED_END', None)
        self.PERCENT_ALIGNED_R1 = kwargs.get(
            'PERCENT_ALIGNED_R1', None)
        self.PERCENT_ALIGNED_R2 = kwargs.get(
            'PERCENT_ALIGNED_R2', None)
        self.PERCENT_DUPLICATE_PAIRED_READS = kwargs.get(
            'PERCENT_DUPLICATE_PAIRED_READS', None)
        self.PERCENT_Q30_R1 = kwargs.get(
            'PERCENT_Q30_R1', None)
        self.PERCENT_Q30_R2 = kwargs.get(
            'PERCENT_Q30_R2', None)
        self.REFERENCE_GENOME = kwargs.get(
            'REFERENCE_GENOME', None)
        self.RUNFOLDER = kwargs.get(
            'RUNFOLDER', None)
        self.SAMPLE_ID = kwargs.get(
            'SAMPLE_ID', None)
        self.SMALLVARIANTSTATISTICSFLAG = kwargs.get(
            'SMALLVARIANTSTATISTICSFLAG', None)
        self.SNVS = kwargs.get(
            'SNVS', None)
        self.SNVS_ALL = kwargs.get(
            'SNVS_ALL', None)
        self.SNVS_PERCENT_FOUND_IN_DBSNP = kwargs.get(
            'SNVS_PERCENT_FOUND_IN_DBSNP', None)
        self.SNV_HET_HOM_RATIO = kwargs.get(
            'SNV_HET_HOM_RATIO', None)
        self.SNV_TS_TV_RATIO = kwargs.get(
            'SNV_TS_TV_RATIO', None)
        self.SPLICESITEREGIONDEL = kwargs.get(
            'SPLICESITEREGIONDEL', None)
        self.SPLICESITEREGIONINS = kwargs.get(
            'SPLICESITEREGIONINS', None)
        self.SPLICESITEREGIONSNVS = kwargs.get(
            'SPLICESITEREGIONSNVS', None)
        self.STOPGAINEDDEL = kwargs.get(
            'STOPGAINEDDEL', None)
        self.STOPGAINEDINS = kwargs.get(
            'STOPGAINEDINS', None)
        self.STOPGAINEDSNVS = kwargs.get(
            'STOPGAINEDSNVS', None)
        self.STOPLOSTDEL = kwargs.get(
            'STOPLOSTDEL', None)
        self.STOPLOSTINS = kwargs.get(
            'STOPLOSTINS', None)
        self.STOPLOSTSNVS = kwargs.get(
            'STOPLOSTSNVS', None)
        self.SVSTATISTICSFLAG = kwargs.get(
            'SVSTATISTICSFLAG', None)
        self.SYNONYMOUSDEL = kwargs.get(
            'SYNONYMOUSDEL', None)
        self.SYNONYMOUSINS = kwargs.get(
            'SYNONYMOUSINS', None)
        self.SYNONYMOUSSNVS = kwargs.get(
            'SYNONYMOUSSNVS', None)
        self.TANDEMDUPLICATIONCOUNT = kwargs.get(
            'TANDEMDUPLICATIONCOUNT', None)
        self.TANDEMDUPLICATIONNUMBERINGENES = kwargs.get(
            'TANDEMDUPLICATIONNUMBERINGENES', None)
        self.TANDEMDUPLICATIONPERCENTINGENES = kwargs.get(
            'TANDEMDUPLICATIONPERCENTINGENES', None)
        self.TOTALNUMBERDEL = kwargs.get(
            'TOTALNUMBERDEL', None)
        self.TOTALNUMBERINS = kwargs.get(
            'TOTALNUMBERINS', None)
        self.TOTALNUMBERSNVS = kwargs.get(
            'TOTALNUMBERSNVS', None)
        self.UTRREGIONDEL = kwargs.get(
            'UTRREGIONDEL', None)
        self.UTRREGIONINS = kwargs.get(
            'UTRREGIONINS', None)
        self.UTRREGIONSNVS = kwargs.get(
            'UTRREGIONSNVS', None)
        self.illumina_version = kwargs.get(
            'illumina_version', None)


class IlluminaSummaryV2(ProtocolElement):
    """
    This is the summary provided by Illumina V2 (current) germline
    samples
    """
    _schemaSource = """
{"type": "record", "name": "IlluminaSummaryV2", "namespace": "Gel_BioInf_Models", "doc": "",
"fields": [{"name": "illumina_version", "type": {"type": "enum", "name": "IlluminaVersion",
"symbols": ["IlluminaSummaryV1", "IlluminaSummaryV2", "IlluminaSummaryCancerV2"]}}, {"name":
"STOP_LOST_SNVS", "type": "double"}, {"name": "SV_INSERTIONS_IN_GENES", "type": "double"}, {"name":
"SAMPLE_ID", "type": "string"}, {"name": "DELETION_HET_HOM_RATIO", "type": "double"}, {"name":
"STOP_GAINED_DELETIONS", "type": "double"}, {"name": "INSERTIONS_PERCENT_FOUND_IN_DBSNP", "type":
"double"}, {"name": "FRAGMENT_LENGTH_SD", "type": "double"}, {"name": "PERCENT_Q30_BASES_READ_1",
"type": "double"}, {"name": "PERCENT_Q30_BASES_READ_2", "type": "double"}, {"name":
"STOP_GAINED_SNVS", "type": "double"}, {"name": "CNV_IN_GENES", "type": "double"}, {"name":
"TOTAL_PF_READS", "type": "double"}, {"name": "DELETIONS_IN_GENES", "type": "double"}, {"name":
"DELETIONS_IN_SPLICE_SITE_REGIONS", "type": "double"}, {"name": "PERCENT_Q30_BASES", "type":
"double"}, {"name": "TOTAL_ALIGNED_BASES_READ_1", "type": "double"}, {"name":
"TOTAL_ALIGNED_BASES_READ_2", "type": "double"}, {"name": "SV_INVERSIONS_IN_GENES", "type":
"double"}, {"name": "SNV_TS_TV_RATIO", "type": "double"}, {"name": "INSERTIONS_IN_MATURE_MIRNA",
"type": "double"}, {"name": "DELETIONS", "type": "double"}, {"name": "PERCENT_ALIGNED_BASES_READ_1",
"type": "double"}, {"name": "INDELS_ALL", "type": "double"}, {"name":
"PERCENT_ALIGNED_BASES_READ_2", "type": "double"}, {"name": "NON_SYNONYMOUS_DELETIONS", "type":
"double"}, {"name": "TOTAL_PF_BASES", "type": "double"}, {"name": "PERCENT_DUPLICATE_PAIRED_READS",
"type": "double"}, {"name": "PERCENT_ALIGNED_READ_2", "type": "double"}, {"name":
"PERCENT_ALIGNED_READ_1", "type": "double"}, {"name": "NON_SYNONYMOUS_SNVS", "type": "double"},
{"name": "SV_PERCENT_DELETIONS_IN_GENES", "type": "double"}, {"name": "MISMATCH_RATE_READ_1",
"type": "double"}, {"name": "SV_DELETIONS_IN_GENES", "type": "double"}, {"name":
"DELETIONS_IN_CODING_REGIONS", "type": "double"}, {"name": "INSERTIONS_IN_SPLICE_SITE_REGIONS",
"type": "double"}, {"name": "SV_DELETIONS", "type": "double"}, {"name": "INSERTIONS", "type":
"double"}, {"name": "FRAGMENT_LENGTH_MIN", "type": "double"}, {"name": "FRAGMENT_LENGTH_MAX",
"type": "double"}, {"name": "RUNFOLDER", "type": "string"}, {"name": "SAMPLE_NAME", "type":
"string"}, {"name": "INSERTIONS_IN_GENES", "type": "double"}, {"name":
"SV_PERCENT_INVERSIONS_IN_GENES", "type": "double"}, {"name": "DELETIONS_IN_EXONS", "type":
"double"}, {"name": "STOP_LOST_DELETIONS", "type": "double"}, {"name":
"SNVS_PERCENT_FOUND_IN_DBSNP", "type": "double"}, {"name": "NON_SYNONYMOUS_INSERTIONS", "type":
"double"}, {"name": "SNVS_IN_MATURE_MIRNA", "type": "double"}, {"name": "FRAMESHIFT_DELETIONS",
"type": "double"}, {"name": "FRAGMENT_LENGTH_MEDIAN", "type": "double"}, {"name":
"SNVS_IN_UTR_REGIONS", "type": "double"}, {"name": "SV_PERCENT_TANDEM_DUPLICATIONS_IN_GENES",
"type": "double"}, {"name": "SV_INVERSIONS", "type": "double"}, {"name":
"SV_TANDEM_DUPLICATIONS_IN_GENES", "type": "double"}, {"name": "DELETIONS_PERCENT_FOUND_IN_DBSNP",
"type": "double"}, {"name": "SV_TANDEM_DUPLICATIONS", "type": "double"}, {"name":
"INDELS_PERCENT_FOUND_IN_DBSNP", "type": "double"}, {"name": "SV_PERCENT_INSERTIONS_IN_GENES",
"type": "double"}, {"name": "STOP_GAINED_INSERTIONS", "type": "double"}, {"name":
"MISMATCH_RATE_READ_2", "type": "double"}, {"name": "DELETIONS_ALL", "type": "double"}, {"name":
"TOTAL_PF_BASES_READ_2", "type": "double"}, {"name": "TOTAL_PF_BASES_READ_1", "type": "double"},
{"name": "INDEL_HET_HOM_RATIO", "type": "double"}, {"name": "SYNONYMOUS_SNVS", "type": "double"},
{"name": "DELETIONS_IN_MATURE_MIRNA", "type": "double"}, {"name": "SNV_HET_HOM_RATIO", "type":
"double"}, {"name": "INSERTIONS_IN_CODING_REGIONS", "type": "double"}, {"name": "SNVS_IN_GENES",
"type": "double"}, {"name": "STOP_LOST_INSERTIONS", "type": "double"}, {"name": "PAIRED_END",
"type": "string"}, {"name": "DIVERSITY", "type": "double"}, {"name": "REFERENCE_GENOME", "type":
"string"}, {"name": "SNVS_IN_EXONS", "type": "double"}, {"name": "SNVS_IN_SPLICE_SITE_REGIONS",
"type": "double"}, {"name": "FRAMESHIFT_INSERTIONS", "type": "double"}, {"name": "INSERTIONS_ALL",
"type": "double"}, {"name": "SNVS_ALL", "type": "double"}, {"name": "INSERTIONS_IN_UTR_REGIONS",
"type": "double"}, {"name": "CNV_PERCENT_IN_GENES", "type": "double"}, {"name": "MEAN_COVERAGE",
"type": "double"}, {"name": "CNV", "type": "double"}, {"name": "SV_INSERTIONS", "type": "double"},
{"name": "SNVS_IN_CODING_REGIONS", "type": "double"}, {"name": "SNVS", "type": "double"}, {"name":
"INSERTION_HET_HOM_RATIO", "type": "double"}, {"name": "INDELS", "type": "double"}, {"name":
"TOTAL_ALIGNED_READ_2", "type": "double"}, {"name": "TOTAL_ALIGNED_READ_1", "type": "double"},
{"name": "DELETIONS_IN_UTR_REGIONS", "type": "double"}, {"name": "INSERTIONS_IN_EXONS", "type":
"double"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "CNV",
        "CNV_IN_GENES",
        "CNV_PERCENT_IN_GENES",
        "DELETIONS",
        "DELETIONS_ALL",
        "DELETIONS_IN_CODING_REGIONS",
        "DELETIONS_IN_EXONS",
        "DELETIONS_IN_GENES",
        "DELETIONS_IN_MATURE_MIRNA",
        "DELETIONS_IN_SPLICE_SITE_REGIONS",
        "DELETIONS_IN_UTR_REGIONS",
        "DELETIONS_PERCENT_FOUND_IN_DBSNP",
        "DELETION_HET_HOM_RATIO",
        "DIVERSITY",
        "FRAGMENT_LENGTH_MAX",
        "FRAGMENT_LENGTH_MEDIAN",
        "FRAGMENT_LENGTH_MIN",
        "FRAGMENT_LENGTH_SD",
        "FRAMESHIFT_DELETIONS",
        "FRAMESHIFT_INSERTIONS",
        "INDELS",
        "INDELS_ALL",
        "INDELS_PERCENT_FOUND_IN_DBSNP",
        "INDEL_HET_HOM_RATIO",
        "INSERTIONS",
        "INSERTIONS_ALL",
        "INSERTIONS_IN_CODING_REGIONS",
        "INSERTIONS_IN_EXONS",
        "INSERTIONS_IN_GENES",
        "INSERTIONS_IN_MATURE_MIRNA",
        "INSERTIONS_IN_SPLICE_SITE_REGIONS",
        "INSERTIONS_IN_UTR_REGIONS",
        "INSERTIONS_PERCENT_FOUND_IN_DBSNP",
        "INSERTION_HET_HOM_RATIO",
        "MEAN_COVERAGE",
        "MISMATCH_RATE_READ_1",
        "MISMATCH_RATE_READ_2",
        "NON_SYNONYMOUS_DELETIONS",
        "NON_SYNONYMOUS_INSERTIONS",
        "NON_SYNONYMOUS_SNVS",
        "PAIRED_END",
        "PERCENT_ALIGNED_BASES_READ_1",
        "PERCENT_ALIGNED_BASES_READ_2",
        "PERCENT_ALIGNED_READ_1",
        "PERCENT_ALIGNED_READ_2",
        "PERCENT_DUPLICATE_PAIRED_READS",
        "PERCENT_Q30_BASES",
        "PERCENT_Q30_BASES_READ_1",
        "PERCENT_Q30_BASES_READ_2",
        "REFERENCE_GENOME",
        "RUNFOLDER",
        "SAMPLE_ID",
        "SAMPLE_NAME",
        "SNVS",
        "SNVS_ALL",
        "SNVS_IN_CODING_REGIONS",
        "SNVS_IN_EXONS",
        "SNVS_IN_GENES",
        "SNVS_IN_MATURE_MIRNA",
        "SNVS_IN_SPLICE_SITE_REGIONS",
        "SNVS_IN_UTR_REGIONS",
        "SNVS_PERCENT_FOUND_IN_DBSNP",
        "SNV_HET_HOM_RATIO",
        "SNV_TS_TV_RATIO",
        "STOP_GAINED_DELETIONS",
        "STOP_GAINED_INSERTIONS",
        "STOP_GAINED_SNVS",
        "STOP_LOST_DELETIONS",
        "STOP_LOST_INSERTIONS",
        "STOP_LOST_SNVS",
        "SV_DELETIONS",
        "SV_DELETIONS_IN_GENES",
        "SV_INSERTIONS",
        "SV_INSERTIONS_IN_GENES",
        "SV_INVERSIONS",
        "SV_INVERSIONS_IN_GENES",
        "SV_PERCENT_DELETIONS_IN_GENES",
        "SV_PERCENT_INSERTIONS_IN_GENES",
        "SV_PERCENT_INVERSIONS_IN_GENES",
        "SV_PERCENT_TANDEM_DUPLICATIONS_IN_GENES",
        "SV_TANDEM_DUPLICATIONS",
        "SV_TANDEM_DUPLICATIONS_IN_GENES",
        "SYNONYMOUS_SNVS",
        "TOTAL_ALIGNED_BASES_READ_1",
        "TOTAL_ALIGNED_BASES_READ_2",
        "TOTAL_ALIGNED_READ_1",
        "TOTAL_ALIGNED_READ_2",
        "TOTAL_PF_BASES",
        "TOTAL_PF_BASES_READ_1",
        "TOTAL_PF_BASES_READ_2",
        "TOTAL_PF_READS",
        "illumina_version",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'CNV', 'CNV_IN_GENES', 'CNV_PERCENT_IN_GENES', 'DELETIONS',
        'DELETIONS_ALL', 'DELETIONS_IN_CODING_REGIONS',
        'DELETIONS_IN_EXONS', 'DELETIONS_IN_GENES',
        'DELETIONS_IN_MATURE_MIRNA',
        'DELETIONS_IN_SPLICE_SITE_REGIONS',
        'DELETIONS_IN_UTR_REGIONS',
        'DELETIONS_PERCENT_FOUND_IN_DBSNP', 'DELETION_HET_HOM_RATIO',
        'DIVERSITY', 'FRAGMENT_LENGTH_MAX', 'FRAGMENT_LENGTH_MEDIAN',
        'FRAGMENT_LENGTH_MIN', 'FRAGMENT_LENGTH_SD',
        'FRAMESHIFT_DELETIONS', 'FRAMESHIFT_INSERTIONS', 'INDELS',
        'INDELS_ALL', 'INDELS_PERCENT_FOUND_IN_DBSNP',
        'INDEL_HET_HOM_RATIO', 'INSERTIONS', 'INSERTIONS_ALL',
        'INSERTIONS_IN_CODING_REGIONS', 'INSERTIONS_IN_EXONS',
        'INSERTIONS_IN_GENES', 'INSERTIONS_IN_MATURE_MIRNA',
        'INSERTIONS_IN_SPLICE_SITE_REGIONS',
        'INSERTIONS_IN_UTR_REGIONS',
        'INSERTIONS_PERCENT_FOUND_IN_DBSNP',
        'INSERTION_HET_HOM_RATIO', 'MEAN_COVERAGE',
        'MISMATCH_RATE_READ_1', 'MISMATCH_RATE_READ_2',
        'NON_SYNONYMOUS_DELETIONS', 'NON_SYNONYMOUS_INSERTIONS',
        'NON_SYNONYMOUS_SNVS', 'PAIRED_END',
        'PERCENT_ALIGNED_BASES_READ_1',
        'PERCENT_ALIGNED_BASES_READ_2', 'PERCENT_ALIGNED_READ_1',
        'PERCENT_ALIGNED_READ_2', 'PERCENT_DUPLICATE_PAIRED_READS',
        'PERCENT_Q30_BASES', 'PERCENT_Q30_BASES_READ_1',
        'PERCENT_Q30_BASES_READ_2', 'REFERENCE_GENOME', 'RUNFOLDER',
        'SAMPLE_ID', 'SAMPLE_NAME', 'SNVS', 'SNVS_ALL',
        'SNVS_IN_CODING_REGIONS', 'SNVS_IN_EXONS', 'SNVS_IN_GENES',
        'SNVS_IN_MATURE_MIRNA', 'SNVS_IN_SPLICE_SITE_REGIONS',
        'SNVS_IN_UTR_REGIONS', 'SNVS_PERCENT_FOUND_IN_DBSNP',
        'SNV_HET_HOM_RATIO', 'SNV_TS_TV_RATIO',
        'STOP_GAINED_DELETIONS', 'STOP_GAINED_INSERTIONS',
        'STOP_GAINED_SNVS', 'STOP_LOST_DELETIONS',
        'STOP_LOST_INSERTIONS', 'STOP_LOST_SNVS', 'SV_DELETIONS',
        'SV_DELETIONS_IN_GENES', 'SV_INSERTIONS',
        'SV_INSERTIONS_IN_GENES', 'SV_INVERSIONS',
        'SV_INVERSIONS_IN_GENES', 'SV_PERCENT_DELETIONS_IN_GENES',
        'SV_PERCENT_INSERTIONS_IN_GENES',
        'SV_PERCENT_INVERSIONS_IN_GENES',
        'SV_PERCENT_TANDEM_DUPLICATIONS_IN_GENES',
        'SV_TANDEM_DUPLICATIONS', 'SV_TANDEM_DUPLICATIONS_IN_GENES',
        'SYNONYMOUS_SNVS', 'TOTAL_ALIGNED_BASES_READ_1',
        'TOTAL_ALIGNED_BASES_READ_2', 'TOTAL_ALIGNED_READ_1',
        'TOTAL_ALIGNED_READ_2', 'TOTAL_PF_BASES',
        'TOTAL_PF_BASES_READ_1', 'TOTAL_PF_BASES_READ_2',
        'TOTAL_PF_READS', 'illumina_version'
    ]

    def __init__(self, **kwargs):
        self.CNV = kwargs.get(
            'CNV', None)
        self.CNV_IN_GENES = kwargs.get(
            'CNV_IN_GENES', None)
        self.CNV_PERCENT_IN_GENES = kwargs.get(
            'CNV_PERCENT_IN_GENES', None)
        self.DELETIONS = kwargs.get(
            'DELETIONS', None)
        self.DELETIONS_ALL = kwargs.get(
            'DELETIONS_ALL', None)
        self.DELETIONS_IN_CODING_REGIONS = kwargs.get(
            'DELETIONS_IN_CODING_REGIONS', None)
        self.DELETIONS_IN_EXONS = kwargs.get(
            'DELETIONS_IN_EXONS', None)
        self.DELETIONS_IN_GENES = kwargs.get(
            'DELETIONS_IN_GENES', None)
        self.DELETIONS_IN_MATURE_MIRNA = kwargs.get(
            'DELETIONS_IN_MATURE_MIRNA', None)
        self.DELETIONS_IN_SPLICE_SITE_REGIONS = kwargs.get(
            'DELETIONS_IN_SPLICE_SITE_REGIONS', None)
        self.DELETIONS_IN_UTR_REGIONS = kwargs.get(
            'DELETIONS_IN_UTR_REGIONS', None)
        self.DELETIONS_PERCENT_FOUND_IN_DBSNP = kwargs.get(
            'DELETIONS_PERCENT_FOUND_IN_DBSNP', None)
        self.DELETION_HET_HOM_RATIO = kwargs.get(
            'DELETION_HET_HOM_RATIO', None)
        self.DIVERSITY = kwargs.get(
            'DIVERSITY', None)
        self.FRAGMENT_LENGTH_MAX = kwargs.get(
            'FRAGMENT_LENGTH_MAX', None)
        self.FRAGMENT_LENGTH_MEDIAN = kwargs.get(
            'FRAGMENT_LENGTH_MEDIAN', None)
        self.FRAGMENT_LENGTH_MIN = kwargs.get(
            'FRAGMENT_LENGTH_MIN', None)
        self.FRAGMENT_LENGTH_SD = kwargs.get(
            'FRAGMENT_LENGTH_SD', None)
        self.FRAMESHIFT_DELETIONS = kwargs.get(
            'FRAMESHIFT_DELETIONS', None)
        self.FRAMESHIFT_INSERTIONS = kwargs.get(
            'FRAMESHIFT_INSERTIONS', None)
        self.INDELS = kwargs.get(
            'INDELS', None)
        self.INDELS_ALL = kwargs.get(
            'INDELS_ALL', None)
        self.INDELS_PERCENT_FOUND_IN_DBSNP = kwargs.get(
            'INDELS_PERCENT_FOUND_IN_DBSNP', None)
        self.INDEL_HET_HOM_RATIO = kwargs.get(
            'INDEL_HET_HOM_RATIO', None)
        self.INSERTIONS = kwargs.get(
            'INSERTIONS', None)
        self.INSERTIONS_ALL = kwargs.get(
            'INSERTIONS_ALL', None)
        self.INSERTIONS_IN_CODING_REGIONS = kwargs.get(
            'INSERTIONS_IN_CODING_REGIONS', None)
        self.INSERTIONS_IN_EXONS = kwargs.get(
            'INSERTIONS_IN_EXONS', None)
        self.INSERTIONS_IN_GENES = kwargs.get(
            'INSERTIONS_IN_GENES', None)
        self.INSERTIONS_IN_MATURE_MIRNA = kwargs.get(
            'INSERTIONS_IN_MATURE_MIRNA', None)
        self.INSERTIONS_IN_SPLICE_SITE_REGIONS = kwargs.get(
            'INSERTIONS_IN_SPLICE_SITE_REGIONS', None)
        self.INSERTIONS_IN_UTR_REGIONS = kwargs.get(
            'INSERTIONS_IN_UTR_REGIONS', None)
        self.INSERTIONS_PERCENT_FOUND_IN_DBSNP = kwargs.get(
            'INSERTIONS_PERCENT_FOUND_IN_DBSNP', None)
        self.INSERTION_HET_HOM_RATIO = kwargs.get(
            'INSERTION_HET_HOM_RATIO', None)
        self.MEAN_COVERAGE = kwargs.get(
            'MEAN_COVERAGE', None)
        self.MISMATCH_RATE_READ_1 = kwargs.get(
            'MISMATCH_RATE_READ_1', None)
        self.MISMATCH_RATE_READ_2 = kwargs.get(
            'MISMATCH_RATE_READ_2', None)
        self.NON_SYNONYMOUS_DELETIONS = kwargs.get(
            'NON_SYNONYMOUS_DELETIONS', None)
        self.NON_SYNONYMOUS_INSERTIONS = kwargs.get(
            'NON_SYNONYMOUS_INSERTIONS', None)
        self.NON_SYNONYMOUS_SNVS = kwargs.get(
            'NON_SYNONYMOUS_SNVS', None)
        self.PAIRED_END = kwargs.get(
            'PAIRED_END', None)
        self.PERCENT_ALIGNED_BASES_READ_1 = kwargs.get(
            'PERCENT_ALIGNED_BASES_READ_1', None)
        self.PERCENT_ALIGNED_BASES_READ_2 = kwargs.get(
            'PERCENT_ALIGNED_BASES_READ_2', None)
        self.PERCENT_ALIGNED_READ_1 = kwargs.get(
            'PERCENT_ALIGNED_READ_1', None)
        self.PERCENT_ALIGNED_READ_2 = kwargs.get(
            'PERCENT_ALIGNED_READ_2', None)
        self.PERCENT_DUPLICATE_PAIRED_READS = kwargs.get(
            'PERCENT_DUPLICATE_PAIRED_READS', None)
        self.PERCENT_Q30_BASES = kwargs.get(
            'PERCENT_Q30_BASES', None)
        self.PERCENT_Q30_BASES_READ_1 = kwargs.get(
            'PERCENT_Q30_BASES_READ_1', None)
        self.PERCENT_Q30_BASES_READ_2 = kwargs.get(
            'PERCENT_Q30_BASES_READ_2', None)
        self.REFERENCE_GENOME = kwargs.get(
            'REFERENCE_GENOME', None)
        self.RUNFOLDER = kwargs.get(
            'RUNFOLDER', None)
        self.SAMPLE_ID = kwargs.get(
            'SAMPLE_ID', None)
        self.SAMPLE_NAME = kwargs.get(
            'SAMPLE_NAME', None)
        self.SNVS = kwargs.get(
            'SNVS', None)
        self.SNVS_ALL = kwargs.get(
            'SNVS_ALL', None)
        self.SNVS_IN_CODING_REGIONS = kwargs.get(
            'SNVS_IN_CODING_REGIONS', None)
        self.SNVS_IN_EXONS = kwargs.get(
            'SNVS_IN_EXONS', None)
        self.SNVS_IN_GENES = kwargs.get(
            'SNVS_IN_GENES', None)
        self.SNVS_IN_MATURE_MIRNA = kwargs.get(
            'SNVS_IN_MATURE_MIRNA', None)
        self.SNVS_IN_SPLICE_SITE_REGIONS = kwargs.get(
            'SNVS_IN_SPLICE_SITE_REGIONS', None)
        self.SNVS_IN_UTR_REGIONS = kwargs.get(
            'SNVS_IN_UTR_REGIONS', None)
        self.SNVS_PERCENT_FOUND_IN_DBSNP = kwargs.get(
            'SNVS_PERCENT_FOUND_IN_DBSNP', None)
        self.SNV_HET_HOM_RATIO = kwargs.get(
            'SNV_HET_HOM_RATIO', None)
        self.SNV_TS_TV_RATIO = kwargs.get(
            'SNV_TS_TV_RATIO', None)
        self.STOP_GAINED_DELETIONS = kwargs.get(
            'STOP_GAINED_DELETIONS', None)
        self.STOP_GAINED_INSERTIONS = kwargs.get(
            'STOP_GAINED_INSERTIONS', None)
        self.STOP_GAINED_SNVS = kwargs.get(
            'STOP_GAINED_SNVS', None)
        self.STOP_LOST_DELETIONS = kwargs.get(
            'STOP_LOST_DELETIONS', None)
        self.STOP_LOST_INSERTIONS = kwargs.get(
            'STOP_LOST_INSERTIONS', None)
        self.STOP_LOST_SNVS = kwargs.get(
            'STOP_LOST_SNVS', None)
        self.SV_DELETIONS = kwargs.get(
            'SV_DELETIONS', None)
        self.SV_DELETIONS_IN_GENES = kwargs.get(
            'SV_DELETIONS_IN_GENES', None)
        self.SV_INSERTIONS = kwargs.get(
            'SV_INSERTIONS', None)
        self.SV_INSERTIONS_IN_GENES = kwargs.get(
            'SV_INSERTIONS_IN_GENES', None)
        self.SV_INVERSIONS = kwargs.get(
            'SV_INVERSIONS', None)
        self.SV_INVERSIONS_IN_GENES = kwargs.get(
            'SV_INVERSIONS_IN_GENES', None)
        self.SV_PERCENT_DELETIONS_IN_GENES = kwargs.get(
            'SV_PERCENT_DELETIONS_IN_GENES', None)
        self.SV_PERCENT_INSERTIONS_IN_GENES = kwargs.get(
            'SV_PERCENT_INSERTIONS_IN_GENES', None)
        self.SV_PERCENT_INVERSIONS_IN_GENES = kwargs.get(
            'SV_PERCENT_INVERSIONS_IN_GENES', None)
        self.SV_PERCENT_TANDEM_DUPLICATIONS_IN_GENES = kwargs.get(
            'SV_PERCENT_TANDEM_DUPLICATIONS_IN_GENES', None)
        self.SV_TANDEM_DUPLICATIONS = kwargs.get(
            'SV_TANDEM_DUPLICATIONS', None)
        self.SV_TANDEM_DUPLICATIONS_IN_GENES = kwargs.get(
            'SV_TANDEM_DUPLICATIONS_IN_GENES', None)
        self.SYNONYMOUS_SNVS = kwargs.get(
            'SYNONYMOUS_SNVS', None)
        self.TOTAL_ALIGNED_BASES_READ_1 = kwargs.get(
            'TOTAL_ALIGNED_BASES_READ_1', None)
        self.TOTAL_ALIGNED_BASES_READ_2 = kwargs.get(
            'TOTAL_ALIGNED_BASES_READ_2', None)
        self.TOTAL_ALIGNED_READ_1 = kwargs.get(
            'TOTAL_ALIGNED_READ_1', None)
        self.TOTAL_ALIGNED_READ_2 = kwargs.get(
            'TOTAL_ALIGNED_READ_2', None)
        self.TOTAL_PF_BASES = kwargs.get(
            'TOTAL_PF_BASES', None)
        self.TOTAL_PF_BASES_READ_1 = kwargs.get(
            'TOTAL_PF_BASES_READ_1', None)
        self.TOTAL_PF_BASES_READ_2 = kwargs.get(
            'TOTAL_PF_BASES_READ_2', None)
        self.TOTAL_PF_READS = kwargs.get(
            'TOTAL_PF_READS', None)
        self.illumina_version = kwargs.get(
            'illumina_version', None)


class IlluminaVersion(object):
    """
    No documentation
    """
    IlluminaSummaryV1 = "IlluminaSummaryV1"
    IlluminaSummaryV2 = "IlluminaSummaryV2"
    IlluminaSummaryCancerV2 = "IlluminaSummaryCancerV2"

    def __hash__(self):
        return str(self).__hash__()


class InbreedingCoefficient(ProtocolElement):
    """
    Inbreeding coefficient
    """
    _schemaSource = """
{"type": "record", "name": "InbreedingCoefficient", "namespace": "Gel_BioInf_Models", "doc": "",
"fields": [{"name": "sampleId", "type": "string", "doc": ""}, {"name": "program", "type": "string",
"doc": ""}, {"name": "version", "type": "string", "doc": ""}, {"name": "estimationMethod", "type":
"string", "doc": ""}, {"name": "coefficient", "type": "double", "doc": ""}, {"name":
"standardError", "type": ["null", "double"], "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "coefficient",
        "estimationMethod",
        "program",
        "sampleId",
        "standardError",
        "version",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'coefficient', 'estimationMethod', 'program', 'sampleId',
        'standardError', 'version'
    ]

    def __init__(self, **kwargs):
        self.coefficient = kwargs.get(
            'coefficient', None)
        self.estimationMethod = kwargs.get(
            'estimationMethod', None)
        self.program = kwargs.get(
            'program', None)
        self.sampleId = kwargs.get(
            'sampleId', None)
        self.standardError = kwargs.get(
            'standardError', None)
        self.version = kwargs.get(
            'version', None)


class InbreedingCoefficientEstimates(ProtocolElement):
    """
    Inbreeding coefficient estimates
    """
    _schemaSource = """
{"type": "record", "name": "InbreedingCoefficientEstimates", "namespace": "Gel_BioInf_Models",
"doc": "", "fields": [{"name": "FID", "type": "string"}, {"name": "IID", "type": "string"}, {"name":
"O_HOM", "type": "double"}, {"name": "E_HOM", "type": "double"}, {"name": "N_NM", "type": "double"},
{"name": "F", "type": "double"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "E_HOM",
        "F",
        "FID",
        "IID",
        "N_NM",
        "O_HOM",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'E_HOM', 'F', 'FID', 'IID', 'N_NM', 'O_HOM'
    ]

    def __init__(self, **kwargs):
        self.E_HOM = kwargs.get(
            'E_HOM', None)
        self.F = kwargs.get(
            'F', None)
        self.FID = kwargs.get(
            'FID', None)
        self.IID = kwargs.get(
            'IID', None)
        self.N_NM = kwargs.get(
            'N_NM', None)
        self.O_HOM = kwargs.get(
            'O_HOM', None)


class InsertSizeGel(ProtocolElement):
    """
    GEL Insert size calculation
    """
    _schemaSource = """
{"type": "record", "name": "InsertSizeGel", "namespace": "Gel_BioInf_Models", "doc": "", "fields":
[{"name": "median_inward", "type": "double"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "median_inward",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'median_inward'
    ]

    def __init__(self, **kwargs):
        self.median_inward = kwargs.get(
            'median_inward', None)


class InterpretationRequestRD(ProtocolElement):
    """
    This record represents basic information for this report
    """
    _schemaSource = """
{"type": "record", "name": "InterpretationRequestRD", "namespace": "Gel_BioInf_Models", "doc": "",
"fields": [{"name": "versionControl", "type": {"type": "record", "name": "VersionControl", "fields":
[{"name": "GitVersionControl", "type": "string", "doc": "", "default": "2.1.0"}]}, "doc": ""},
{"name": "InterpretationRequestID", "type": "string", "doc": ""}, {"name": "genomeAssemblyVersion",
"type": "string", "doc": "", "default": "GRCh37.p13"}, {"name": "cellbaseVersion", "type": "string",
"doc": "", "default": "4.0"}, {"name": "InterpretationRequestVersion", "type": "int", "doc": ""},
{"name": "interpretGenome", "type": "boolean", "doc": "", "default": false}, {"name": "workspace",
"type": {"type": "array", "items": "string"}, "doc": ""}, {"name": "BAMs", "type": {"type": "array",
"items": {"type": "record", "name": "File", "fields": [{"name": "SampleId", "type": ["null",
"string", {"type": "array", "items": "string"}], "doc": ""}, {"name": "URIFile", "type": "string",
"doc": ""}, {"name": "fileType", "type": {"type": "enum", "name": "FileType", "doc": "", "symbols":
["BAM", "gVCF", "VCF_small", "VCF_CNV", "VCF_SV", "VCF_SV_CNV", "SVG", "TIER", "ANN",
"BigWig"]}}]}}, "doc": ""}, {"name": "VCFs", "type": {"type": "array", "items": "File"}, "doc": ""},
{"name": "bigWigs", "type": ["null", {"type": "array", "items": "File"}], "doc": ""}, {"name":
"pedigreeDiagram", "type": ["null", "File"], "doc": ""}, {"name": "annotationFile", "type": ["null",
"File"], "doc": ""}, {"name": "pedigree", "type": {"type": "record", "name": "Pedigree", "doc": "",
"fields": [{"name": "versionControl", "type": "VersionControl", "doc": ""}, {"name": "gelFamilyId",
"type": "string", "doc": ""}, {"name": "participants", "type": {"type": "array", "items": {"type":
"record", "name": "RDParticipant", "doc": "", "fields": [{"name": "versionControl", "type":
"VersionControl", "doc": ""}, {"name": "pedigreeId", "type": "int", "doc": ""}, {"name":
"isProband", "type": "boolean", "doc": ""}, {"name": "gelId", "type": ["null", "string"], "doc":
""}, {"name": "gelFamilyId", "type": "string", "doc": ""}, {"name": "gelSuperFamilyId", "type":
["null", "string"], "doc": ""}, {"name": "sex", "type": {"type": "enum", "name": "Sex", "doc": "",
"symbols": ["male", "female", "unknown", "undetermined"]}, "doc": ""}, {"name":
"personKaryotipicSex", "type": ["null", {"type": "enum", "name": "PersonKaryotipicSex", "doc": "",
"symbols": ["unknown", "XX", "XY", "XO", "XXY", "XXX", "XXYY", "XXXY", "XXXX", "XYY", "other"]}],
"doc": ""}, {"name": "fatherId", "type": ["null", "int"], "doc": ""}, {"name": "motherId", "type":
["null", "int"], "doc": ""}, {"name": "superFatherId", "type": ["null", "int"], "doc": ""}, {"name":
"superMotherId", "type": ["null", "int"], "doc": ""}, {"name": "twinGroup", "type": ["null", "int"],
"doc": ""}, {"name": "monozygotic", "type": ["null", {"type": "enum", "name": "TernaryOption",
"doc": "", "symbols": ["yes", "no", "unknown"]}], "doc": ""}, {"name": "adoptedStatus", "type":
{"type": "enum", "name": "AdoptedStatus", "doc": "", "symbols": ["not_adopted", "adoptedin",
"adoptedout"]}, "doc": ""}, {"name": "lifeStatus", "type": {"type": "enum", "name": "LifeStatus",
"doc": "", "symbols": ["alive", "aborted", "deceased", "unborn", "stillborn", "miscarriage"]},
"doc": ""}, {"name": "consanguineousParents", "type": "TernaryOption", "doc": ""}, {"name":
"consanguineousPopulation", "type": ["null", "string"], "doc": ""}, {"name": "affectionStatus",
"type": {"type": "enum", "name": "AffectionStatus", "doc": "", "symbols": ["unaffected", "affected",
"unknown"]}, "doc": ""}, {"name": "disorderList", "type": {"type": "array", "items": {"type":
"record", "name": "Disorder", "doc": "", "fields": [{"name": "diseaseGroup", "type": ["null",
"string"], "doc": ""}, {"name": "diseaseSubGroup", "type": ["null", "string"], "doc": ""}, {"name":
"specificDisease", "type": ["null", "string"], "doc": ""}, {"name": "ageOfOnset", "type": ["null",
"string"], "doc": ""}]}}, "doc": ""}, {"name": "hpoTermList", "type": {"type": "array", "items":
{"type": "record", "name": "HpoTerm", "doc": "", "fields": [{"name": "term", "type": "string",
"doc": ""}, {"name": "termPresence", "type": ["null", "boolean"], "doc": ""}, {"name": "modifiers",
"type": ["null", {"type": "map", "values": "string"}], "doc": ""}, {"name": "ageOfOnset", "type":
["null", "string"], "doc": ""}]}}, "doc": ""}, {"name": "ancestries", "type": {"type": "record",
"name": "Ancestries", "fields": [{"name": "mothersEthnicOrigin", "type": ["null", {"type": "enum",
"name": "EthnicCategory", "doc": "", "symbols": ["D", "E", "F", "G", "A", "B", "C", "L", "M", "N",
"H", "J", "K", "P", "S", "R", "Z"]}], "doc": ""}, {"name": "mothersOtherRelevantAncestry", "type":
["null", "string"], "doc": ""}, {"name": "fathersEthnicOrigin", "type": ["null", "EthnicCategory"],
"doc": ""}, {"name": "fathersOtherRelevantAncestry", "type": ["null", "string"], "doc": ""},
{"name": "chiSquare1KGenomesPhase3Pop", "type": ["null", {"type": "array", "items": {"type":
"record", "name": "ChiSquare1KGenomesPhase3Pop", "doc": "", "fields": [{"name":
"kGSuperPopCategory", "type": {"type": "enum", "name": "KGSuperPopCategory", "doc": "", "symbols":
["AFR", "AMR", "EAS", "EUR", "SAS"]}, "doc": ""}, {"name": "kGPopCategory", "type": ["null",
{"type": "enum", "name": "KGPopCategory", "doc": "", "symbols": ["ACB", "ASW", "BEB", "CDX", "CEU",
"CHB", "CHS", "CLM", "ESN", "FIN", "GBR", "GIH", "GWD", "IBS", "ITU", "JPT", "KHV", "LWK", "MSL",
"MXL", "PEL", "PJL", "PUR", "STU", "TSI", "YRI"]}], "doc": ""}, {"name": "chiSquare", "type":
"float", "doc": ""}]}}], "doc": ""}]}, "doc": ""}, {"name": "dataModelCatalogueVersion", "type":
"string", "doc": "", "default": "v4.2"}, {"name": "consentStatus", "type": {"type": "record",
"name": "ConsentStatus", "doc": "", "fields": [{"name": "programmeConsent", "type": "boolean",
"doc": "", "default": false}, {"name": "primaryFindingConsent", "type": "boolean", "doc": "",
"default": false}, {"name": "secondaryFindingConsent", "type": "boolean", "doc": "", "default":
false}, {"name": "carrierStatusConsent", "type": "boolean", "doc": "", "default": false}]}, "doc":
""}, {"name": "samples", "type": ["null", {"type": "array", "items": "string"}], "doc": ""},
{"name": "inbreedingCoefficient", "type": ["null", {"type": "record", "name":
"InbreedingCoefficient", "doc": "", "fields": [{"name": "sampleId", "type": "string", "doc": ""},
{"name": "program", "type": "string", "doc": ""}, {"name": "version", "type": "string", "doc": ""},
{"name": "estimationMethod", "type": "string", "doc": ""}, {"name": "coefficient", "type": "double",
"doc": ""}, {"name": "standardError", "type": ["null", "double"], "doc": ""}]}], "doc": ""},
{"name": "additionalInformation", "type": ["null", {"type": "map", "values": "string"}], "doc":
""}]}}}, {"name": "analysisPanels", "type": ["null", {"type": "array", "items": {"type": "record",
"name": "AnalysisPanel", "fields": [{"name": "specificDisease", "type": "string"}, {"name":
"panelName", "type": "string"}, {"name": "panelVersion", "type": ["null", "string"]}, {"name":
"review_outcome", "type": "string"}, {"name": "multiple_genetic_origins", "type": "string"}]}}]}]},
"doc": ""}, {"name": "TieredVariants", "type": {"type": "array", "items": {"type": "record", "name":
"ReportedVariant", "fields": [{"name": "versionControl", "type": "VersionControl", "doc": ""},
{"name": "chromosome", "type": "string", "doc": ""}, {"name": "dbSNPid", "type": ["null", "string"],
"doc": ""}, {"name": "position", "type": "int", "doc": ""}, {"name": "reference", "type": "string",
"doc": ""}, {"name": "alternate", "type": "string", "doc": ""}, {"name": "calledGenotypes", "type":
{"type": "array", "items": {"type": "record", "name": "CalledGenotype", "doc": "", "fields":
[{"name": "gelId", "type": "string", "doc": ""}, {"name": "sampleId", "type": "string", "doc": ""},
{"name": "genotype", "type": "string", "doc": ""}, {"name": "phaseSet", "type": ["null", "int"],
"doc": ""}, {"name": "depthReference", "type": ["null", "int"], "doc": ""}, {"name":
"depthAlternate", "type": ["null", "int"], "doc": ""}, {"name": "copyNumber", "type": ["null",
"int"], "doc": ""}]}}, "doc": ""}, {"name": "reportEvents", "type": {"type": "array", "items":
{"type": "record", "name": "ReportEvent", "fields": [{"name": "reportEventId", "type": "string",
"doc": ""}, {"name": "phenotype", "type": "string", "doc": ""}, {"name": "panelName", "type":
["null", "string"], "doc": ""}, {"name": "panelVersion", "type": ["null", "string"], "doc": ""},
{"name": "modeOfInheritance", "type": {"type": "enum", "name": "ReportedModeOfInheritance", "doc":
"", "symbols": ["monoallelic", "monoallelic_not_imprinted", "monoallelic_maternally_imprinted",
"monoallelic_paternally_imprinted", "biallelic", "monoallelic_and_biallelic",
"monoallelic_and_more_severe_biallelic", "xlinked_biallelic", "xlinked_monoallelic",
"mitochondrial", "unknown"]}, "doc": ""}, {"name": "genomicFeature", "type": {"type": "record",
"name": "GenomicFeature", "fields": [{"name": "featureType", "type": {"type": "enum", "name":
"FeatureTypes", "symbols": ["RegulatoryRegion", "Gene", "Transcript"]}, "doc": ""}, {"name":
"ensemblId", "type": "string", "doc": ""}, {"name": "ids", "type": {"type": "map", "values":
"string"}, "doc": ""}]}, "doc": ""}, {"name": "penetrance", "type": {"type": "enum", "name":
"Penetrance", "doc": "", "symbols": ["complete", "incomplete"]}, "doc": ""}, {"name": "score",
"type": "float", "doc": ""}, {"name": "vendorSpecificScores", "type": ["null", {"type": "map",
"values": "float"}], "doc": ""}, {"name": "variantClassification", "type": ["null", {"type": "enum",
"name": "VariantClassification", "doc": "", "symbols": ["BENIGN", "LIKELY_BENIGN", "VUS",
"LIKELY_PATHOGENIC", "PATHOGENIC"]}], "doc": ""}, {"name": "fullyExplainsPhenotype", "type":
["null", "boolean"], "doc": ""}, {"name": "groupOfVariants", "type": ["null", "int"], "doc": ""},
{"name": "eventJustification", "type": "string", "doc": ""}, {"name": "tier", "type": ["null",
{"type": "enum", "name": "Tier", "doc": "", "symbols": ["NONE", "TIER1", "TIER2", "TIER3"]}], "doc":
""}]}}, "doc": ""}, {"name": "additionalTextualVariantAnnotations", "type": ["null", {"type": "map",
"values": "string"}], "doc": ""}, {"name": "additionalNumericVariantAnnotations", "type": ["null",
{"type": "map", "values": "float"}], "doc": ""}, {"name": "comments", "type": ["null", {"type":
"array", "items": "string"}], "doc": ""}]}}, "doc": ""}, {"name": "TieringVersion", "type":
"string", "doc": ""}, {"name": "complexGeneticPhenomena", "type": ["null", {"type": "enum", "name":
"ComplexGeneticPhenomena", "symbols": ["mosaicism", "monosomy", "disomy", "uniparental_disomy",
"trisomy", "other_aneuploidy"]}], "doc": ""}, {"name": "modeOfInheritance", "type": {"type": "enum",
"name": "ModeOfInheritance", "doc": "", "symbols": ["monoallelic", "monoallelic_not_imprinted",
"monoallelic_maternally_imprinted", "monoallelic_paternally_imprinted", "biallelic",
"monoallelic_and_biallelic", "monoallelic_and_more_severe_biallelic", "xlinked_biallelic",
"xlinked_monoallelic", "mitochondrial", "unknown"]}, "doc": ""}, {"name": "virtualPanel", "type":
["null", {"type": "array", "items": {"type": "record", "name": "VirtualPanel", "doc": "", "fields":
[{"name": "panelName", "type": "string", "doc": ""}, {"name": "panelVersion", "type": "string",
"doc": ""}, {"name": "ensemblVersion", "type": ["null", "string"], "doc": ""}, {"name":
"dataModelCatalogueVersion", "type": "string", "doc": "", "default": "4.2"}, {"name": "geneIds",
"type": {"type": "array", "items": "string"}, "doc": ""}, {"name": "Transcripts", "type": ["null",
{"type": "array", "items": "string"}], "doc": ""}, {"name": "relevantRegions", "type": ["null",
"File"], "doc": ""}, {"name": "clinicalRelevantVariants", "type": ["null", "File"], "doc": ""}]}}],
"doc": ""}, {"name": "otherFamilyHistory", "type": ["null", {"type": "record", "name":
"OtherFamilyHistory", "doc": "", "fields": [{"name": "maternalFamilyHistory", "type": ["null",
{"type": "array", "items": "string"}], "doc": ""}, {"name": "paternalFamilyHistory", "type":
["null", {"type": "array", "items": "string"}], "doc": ""}]}], "doc": ""}, {"name":
"analysisReturnURI", "type": "string", "doc": ""}, {"name": "analysisVersion", "type": "string",
"doc": "", "default": "1"}, {"name": "additionalInfo", "type": ["null", {"type": "map", "values":
"string"}], "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "BAMs",
        "InterpretationRequestID",
        "InterpretationRequestVersion",
        "TieredVariants",
        "TieringVersion",
        "VCFs",
        "additionalInfo",
        "analysisReturnURI",
        "annotationFile",
        "bigWigs",
        "complexGeneticPhenomena",
        "modeOfInheritance",
        "otherFamilyHistory",
        "pedigree",
        "pedigreeDiagram",
        "versionControl",
        "virtualPanel",
        "workspace",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'BAMs': File,
            'TieredVariants': ReportedVariant,
            'VCFs': File,
            'annotationFile': File,
            'bigWigs': File,
            'otherFamilyHistory': OtherFamilyHistory,
            'pedigree': Pedigree,
            'pedigreeDiagram': File,
            'versionControl': VersionControl,
            'virtualPanel': VirtualPanel,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'BAMs': File,
            'TieredVariants': ReportedVariant,
            'VCFs': File,
            'annotationFile': File,
            'bigWigs': File,
            'otherFamilyHistory': OtherFamilyHistory,
            'pedigree': Pedigree,
            'pedigreeDiagram': File,
            'versionControl': VersionControl,
            'virtualPanel': VirtualPanel,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'BAMs', 'InterpretationRequestID',
        'InterpretationRequestVersion', 'TieredVariants',
        'TieringVersion', 'VCFs', 'additionalInfo',
        'analysisReturnURI', 'analysisVersion', 'annotationFile',
        'bigWigs', 'cellbaseVersion', 'complexGeneticPhenomena',
        'genomeAssemblyVersion', 'interpretGenome',
        'modeOfInheritance', 'otherFamilyHistory', 'pedigree',
        'pedigreeDiagram', 'versionControl', 'virtualPanel',
        'workspace'
    ]

    def __init__(self, **kwargs):
        self.BAMs = kwargs.get(
            'BAMs', None)
        self.InterpretationRequestID = kwargs.get(
            'InterpretationRequestID', None)
        self.InterpretationRequestVersion = kwargs.get(
            'InterpretationRequestVersion', None)
        self.TieredVariants = kwargs.get(
            'TieredVariants', None)
        self.TieringVersion = kwargs.get(
            'TieringVersion', None)
        self.VCFs = kwargs.get(
            'VCFs', None)
        self.additionalInfo = kwargs.get(
            'additionalInfo', None)
        self.analysisReturnURI = kwargs.get(
            'analysisReturnURI', None)
        self.analysisVersion = kwargs.get(
            'analysisVersion', '1')
        self.annotationFile = kwargs.get(
            'annotationFile', None)
        self.bigWigs = kwargs.get(
            'bigWigs', None)
        self.cellbaseVersion = kwargs.get(
            'cellbaseVersion', '4.0')
        self.complexGeneticPhenomena = kwargs.get(
            'complexGeneticPhenomena', None)
        self.genomeAssemblyVersion = kwargs.get(
            'genomeAssemblyVersion', 'GRCh37.p13')
        self.interpretGenome = kwargs.get(
            'interpretGenome', False)
        self.modeOfInheritance = kwargs.get(
            'modeOfInheritance', None)
        self.otherFamilyHistory = kwargs.get(
            'otherFamilyHistory', None)
        self.pedigree = kwargs.get(
            'pedigree', Pedigree())
        self.pedigreeDiagram = kwargs.get(
            'pedigreeDiagram', None)
        self.versionControl = kwargs.get(
            'versionControl', VersionControl())
        self.virtualPanel = kwargs.get(
            'virtualPanel', None)
        self.workspace = kwargs.get(
            'workspace', None)


class InterpretedGenomeRD(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "InterpretedGenomeRD", "namespace": "Gel_BioInf_Models", "fields":
[{"name": "versionControl", "type": {"type": "record", "name": "VersionControl", "fields": [{"name":
"GitVersionControl", "type": "string", "doc": "", "default": "2.1.0"}]}, "doc": ""}, {"name":
"InterpretationRequestID", "type": "string", "doc": ""}, {"name": "analysisId", "type": "string",
"doc": ""}, {"name": "companyName", "type": "string", "doc": ""}, {"name": "reportURI", "type":
"string", "doc": ""}, {"name": "reportURL", "type": "string", "doc": ""}, {"name":
"reportedVariants", "type": {"type": "array", "items": {"type": "record", "name": "ReportedVariant",
"fields": [{"name": "versionControl", "type": "VersionControl", "doc": ""}, {"name": "chromosome",
"type": "string", "doc": ""}, {"name": "dbSNPid", "type": ["null", "string"], "doc": ""}, {"name":
"position", "type": "int", "doc": ""}, {"name": "reference", "type": "string", "doc": ""}, {"name":
"alternate", "type": "string", "doc": ""}, {"name": "calledGenotypes", "type": {"type": "array",
"items": {"type": "record", "name": "CalledGenotype", "doc": "", "fields": [{"name": "gelId",
"type": "string", "doc": ""}, {"name": "sampleId", "type": "string", "doc": ""}, {"name":
"genotype", "type": "string", "doc": ""}, {"name": "phaseSet", "type": ["null", "int"], "doc": ""},
{"name": "depthReference", "type": ["null", "int"], "doc": ""}, {"name": "depthAlternate", "type":
["null", "int"], "doc": ""}, {"name": "copyNumber", "type": ["null", "int"], "doc": ""}]}}, "doc":
""}, {"name": "reportEvents", "type": {"type": "array", "items": {"type": "record", "name":
"ReportEvent", "fields": [{"name": "reportEventId", "type": "string", "doc": ""}, {"name":
"phenotype", "type": "string", "doc": ""}, {"name": "panelName", "type": ["null", "string"], "doc":
""}, {"name": "panelVersion", "type": ["null", "string"], "doc": ""}, {"name": "modeOfInheritance",
"type": {"type": "enum", "name": "ReportedModeOfInheritance", "doc": "", "symbols": ["monoallelic",
"monoallelic_not_imprinted", "monoallelic_maternally_imprinted", "monoallelic_paternally_imprinted",
"biallelic", "monoallelic_and_biallelic", "monoallelic_and_more_severe_biallelic",
"xlinked_biallelic", "xlinked_monoallelic", "mitochondrial", "unknown"]}, "doc": ""}, {"name":
"genomicFeature", "type": {"type": "record", "name": "GenomicFeature", "fields": [{"name":
"featureType", "type": {"type": "enum", "name": "FeatureTypes", "symbols": ["RegulatoryRegion",
"Gene", "Transcript"]}, "doc": ""}, {"name": "ensemblId", "type": "string", "doc": ""}, {"name":
"ids", "type": {"type": "map", "values": "string"}, "doc": ""}]}, "doc": ""}, {"name": "penetrance",
"type": {"type": "enum", "name": "Penetrance", "doc": "", "symbols": ["complete", "incomplete"]},
"doc": ""}, {"name": "score", "type": "float", "doc": ""}, {"name": "vendorSpecificScores", "type":
["null", {"type": "map", "values": "float"}], "doc": ""}, {"name": "variantClassification", "type":
["null", {"type": "enum", "name": "VariantClassification", "doc": "", "symbols": ["BENIGN",
"LIKELY_BENIGN", "VUS", "LIKELY_PATHOGENIC", "PATHOGENIC"]}], "doc": ""}, {"name":
"fullyExplainsPhenotype", "type": ["null", "boolean"], "doc": ""}, {"name": "groupOfVariants",
"type": ["null", "int"], "doc": ""}, {"name": "eventJustification", "type": "string", "doc": ""},
{"name": "tier", "type": ["null", {"type": "enum", "name": "Tier", "doc": "", "symbols": ["NONE",
"TIER1", "TIER2", "TIER3"]}], "doc": ""}]}}, "doc": ""}, {"name":
"additionalTextualVariantAnnotations", "type": ["null", {"type": "map", "values": "string"}], "doc":
""}, {"name": "additionalNumericVariantAnnotations", "type": ["null", {"type": "map", "values":
"float"}], "doc": ""}, {"name": "comments", "type": ["null", {"type": "array", "items": "string"}],
"doc": ""}]}}, "doc": ""}, {"name": "reportedStructuralVariants", "type": ["null", {"type": "array",
"items": {"type": "record", "name": "ReportedStructuralVariant", "fields": [{"name":
"versionControl", "type": "VersionControl", "doc": ""}, {"name": "chromosome", "type": "string",
"doc": ""}, {"name": "start", "type": "int", "doc": ""}, {"name": "end", "type": "int", "doc": ""},
{"name": "type", "type": "string", "doc": ""}, {"name": "reference", "type": "string", "doc": ""},
{"name": "alternate", "type": "string", "doc": ""}, {"name": "calledGenotypes", "type": {"type":
"array", "items": "CalledGenotype"}}, {"name": "reportEvents", "type": {"type": "array", "items":
"ReportEvent"}, "doc": ""}, {"name": "additionalTextualVariantAnnotations", "type": ["null",
{"type": "map", "values": "string"}], "doc": ""}, {"name": "additionalNumericVariantAnnotations",
"type": ["null", {"type": "map", "values": "float"}], "doc": ""}, {"name": "comments", "type":
["null", {"type": "array", "items": "string"}], "doc": ""}]}}], "doc": ""}, {"name": "comments",
"type": ["null", {"type": "array", "items": "string"}], "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "InterpretationRequestID",
        "analysisId",
        "comments",
        "companyName",
        "reportURI",
        "reportURL",
        "reportedStructuralVariants",
        "reportedVariants",
        "versionControl",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'reportedStructuralVariants': ReportedStructuralVariant,
            'reportedVariants': ReportedVariant,
            'versionControl': VersionControl,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'reportedStructuralVariants': ReportedStructuralVariant,
            'reportedVariants': ReportedVariant,
            'versionControl': VersionControl,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'InterpretationRequestID', 'analysisId', 'comments',
        'companyName', 'reportURI', 'reportURL',
        'reportedStructuralVariants', 'reportedVariants',
        'versionControl'
    ]

    def __init__(self, **kwargs):
        self.InterpretationRequestID = kwargs.get(
            'InterpretationRequestID', None)
        self.analysisId = kwargs.get(
            'analysisId', None)
        self.comments = kwargs.get(
            'comments', None)
        self.companyName = kwargs.get(
            'companyName', None)
        self.reportURI = kwargs.get(
            'reportURI', None)
        self.reportURL = kwargs.get(
            'reportURL', None)
        self.reportedStructuralVariants = kwargs.get(
            'reportedStructuralVariants', None)
        self.reportedVariants = kwargs.get(
            'reportedVariants', None)
        self.versionControl = kwargs.get(
            'versionControl', VersionControl())


class KGPopCategory(object):
    """
    1K Super Population
    """
    ACB = "ACB"
    ASW = "ASW"
    BEB = "BEB"
    CDX = "CDX"
    CEU = "CEU"
    CHB = "CHB"
    CHS = "CHS"
    CLM = "CLM"
    ESN = "ESN"
    FIN = "FIN"
    GBR = "GBR"
    GIH = "GIH"
    GWD = "GWD"
    IBS = "IBS"
    ITU = "ITU"
    JPT = "JPT"
    KHV = "KHV"
    LWK = "LWK"
    MSL = "MSL"
    MXL = "MXL"
    PEL = "PEL"
    PJL = "PJL"
    PUR = "PUR"
    STU = "STU"
    TSI = "TSI"
    YRI = "YRI"

    def __hash__(self):
        return str(self).__hash__()


class KGSuperPopCategory(object):
    """
    1K Population
    """
    AFR = "AFR"
    AMR = "AMR"
    EAS = "EAS"
    EUR = "EUR"
    SAS = "SAS"

    def __hash__(self):
        return str(self).__hash__()


class LifeStatus(object):
    """
    Life Status
    """
    alive = "alive"
    aborted = "aborted"
    deceased = "deceased"
    unborn = "unborn"
    stillborn = "stillborn"
    miscarriage = "miscarriage"

    def __hash__(self):
        return str(self).__hash__()


class Machine(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "Machine", "namespace": "Gel_BioInf_Models", "fields": [{"name": "DATE",
"type": "double"}, {"name": "MACHINE", "type": "string"}, {"name": "FLOWCELL", "type": "string"},
{"name": "RUN", "type": "string"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "DATE",
        "FLOWCELL",
        "MACHINE",
        "RUN",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'DATE', 'FLOWCELL', 'MACHINE', 'RUN'
    ]

    def __init__(self, **kwargs):
        self.DATE = kwargs.get(
            'DATE', None)
        self.FLOWCELL = kwargs.get(
            'FLOWCELL', None)
        self.MACHINE = kwargs.get(
            'MACHINE', None)
        self.RUN = kwargs.get(
            'RUN', None)


class MatchedSamples(ProtocolElement):
    """
    This define a pair of germline and tumor, this pair should/must be
    analyzed together
    """
    _schemaSource = """
{"type": "record", "name": "MatchedSamples", "namespace": "Gel_BioInf_Models", "doc": "", "fields":
[{"name": "germlineSampleId", "type": ["null", "string"], "doc": ""}, {"name": "tumorSampleId",
"type": ["null", "string"], "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "germlineSampleId",
        "tumorSampleId",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'germlineSampleId', 'tumorSampleId'
    ]

    def __init__(self, **kwargs):
        self.germlineSampleId = kwargs.get(
            'germlineSampleId', None)
        self.tumorSampleId = kwargs.get(
            'tumorSampleId', None)


class Method(object):
    """
    No documentation
    """
    RESECTION = "RESECTION"
    BIOPSY = "BIOPSY"
    BLOOD = "BLOOD"

    def __hash__(self):
        return str(self).__hash__()


class ModeOfInheritance(object):
    """
    An enumeration for the different mode of inheritances:  *
    `monoallelic_not_imprinted`: MONOALLELIC, autosomal or
    pseudoautosomal, not imprinted *
    `monoallelic_maternally_imprinted`: MONOALLELIC, autosomal or
    pseudoautosomal, maternally imprinted (paternal allele expressed)
    * `monoallelic_paternally_imprinted`: MONOALLELIC, autosomal or
    pseudoautosomal, paternally imprinted (maternal allele expressed)
    * `monoallelic`: MONOALLELIC, autosomal or pseudoautosomal,
    imprinted status unknown * `biallelic`: BIALLELIC, autosomal or
    pseudoautosomal * `monoallelic_and_biallelic`: BOTH monoallelic
    and biallelic, autosomal or pseudoautosomal *
    `monoallelic_and_more_severe_biallelic`: BOTH monoallelic and
    biallelic, autosomal or pseudoautosomal (but BIALLELIC mutations
    cause a more SEVERE disease form), autosomal or pseudoautosomal *
    `xlinked_biallelic`: X-LINKED: hemizygous mutation in males,
    biallelic mutations in females * `xlinked_monoallelic`: X linked:
    hemizygous mutation in males, monoallelic mutations in females may
    cause disease (may be less severe, later onset than males) *
    `mitochondrial`: MITOCHONDRIAL * `unknown`: Unknown
    """
    monoallelic = "monoallelic"
    monoallelic_not_imprinted = "monoallelic_not_imprinted"
    monoallelic_maternally_imprinted = "monoallelic_maternally_imprinted"
    monoallelic_paternally_imprinted = "monoallelic_paternally_imprinted"
    biallelic = "biallelic"
    monoallelic_and_biallelic = "monoallelic_and_biallelic"
    monoallelic_and_more_severe_biallelic = "monoallelic_and_more_severe_biallelic"
    xlinked_biallelic = "xlinked_biallelic"
    xlinked_monoallelic = "xlinked_monoallelic"
    mitochondrial = "mitochondrial"
    unknown = "unknown"

    def __hash__(self):
        return str(self).__hash__()


class OtherFamilyHistory(ProtocolElement):
    """
    Family history for secondary findings Arrays of strings describing
    discrete family history phenotypes
    """
    _schemaSource = """
{"type": "record", "name": "OtherFamilyHistory", "namespace": "Gel_BioInf_Models", "doc": "",
"fields": [{"name": "maternalFamilyHistory", "type": ["null", {"type": "array", "items": "string"}],
"doc": ""}, {"name": "paternalFamilyHistory", "type": ["null", {"type": "array", "items":
"string"}], "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "maternalFamilyHistory",
        "paternalFamilyHistory",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'maternalFamilyHistory', 'paternalFamilyHistory'
    ]

    def __init__(self, **kwargs):
        self.maternalFamilyHistory = kwargs.get(
            'maternalFamilyHistory', None)
        self.paternalFamilyHistory = kwargs.get(
            'paternalFamilyHistory', None)


class Pedigree(ProtocolElement):
    """
    This is the concept of a family with associated phenotypes as
    present in the record RDParticipant
    """
    _schemaSource = """
{"type": "record", "name": "Pedigree", "namespace": "Gel_BioInf_Models", "doc": "", "fields":
[{"name": "versionControl", "type": {"type": "record", "name": "VersionControl", "fields": [{"name":
"GitVersionControl", "type": "string", "doc": "", "default": "2.1.0"}]}, "doc": ""}, {"name":
"gelFamilyId", "type": "string", "doc": ""}, {"name": "participants", "type": {"type": "array",
"items": {"type": "record", "name": "RDParticipant", "doc": "", "fields": [{"name":
"versionControl", "type": "VersionControl", "doc": ""}, {"name": "pedigreeId", "type": "int", "doc":
""}, {"name": "isProband", "type": "boolean", "doc": ""}, {"name": "gelId", "type": ["null",
"string"], "doc": ""}, {"name": "gelFamilyId", "type": "string", "doc": ""}, {"name":
"gelSuperFamilyId", "type": ["null", "string"], "doc": ""}, {"name": "sex", "type": {"type": "enum",
"name": "Sex", "doc": "", "symbols": ["male", "female", "unknown", "undetermined"]}, "doc": ""},
{"name": "personKaryotipicSex", "type": ["null", {"type": "enum", "name": "PersonKaryotipicSex",
"doc": "", "symbols": ["unknown", "XX", "XY", "XO", "XXY", "XXX", "XXYY", "XXXY", "XXXX", "XYY",
"other"]}], "doc": ""}, {"name": "fatherId", "type": ["null", "int"], "doc": ""}, {"name":
"motherId", "type": ["null", "int"], "doc": ""}, {"name": "superFatherId", "type": ["null", "int"],
"doc": ""}, {"name": "superMotherId", "type": ["null", "int"], "doc": ""}, {"name": "twinGroup",
"type": ["null", "int"], "doc": ""}, {"name": "monozygotic", "type": ["null", {"type": "enum",
"name": "TernaryOption", "doc": "", "symbols": ["yes", "no", "unknown"]}], "doc": ""}, {"name":
"adoptedStatus", "type": {"type": "enum", "name": "AdoptedStatus", "doc": "", "symbols":
["not_adopted", "adoptedin", "adoptedout"]}, "doc": ""}, {"name": "lifeStatus", "type": {"type":
"enum", "name": "LifeStatus", "doc": "", "symbols": ["alive", "aborted", "deceased", "unborn",
"stillborn", "miscarriage"]}, "doc": ""}, {"name": "consanguineousParents", "type": "TernaryOption",
"doc": ""}, {"name": "consanguineousPopulation", "type": ["null", "string"], "doc": ""}, {"name":
"affectionStatus", "type": {"type": "enum", "name": "AffectionStatus", "doc": "", "symbols":
["unaffected", "affected", "unknown"]}, "doc": ""}, {"name": "disorderList", "type": {"type":
"array", "items": {"type": "record", "name": "Disorder", "doc": "", "fields": [{"name":
"diseaseGroup", "type": ["null", "string"], "doc": ""}, {"name": "diseaseSubGroup", "type": ["null",
"string"], "doc": ""}, {"name": "specificDisease", "type": ["null", "string"], "doc": ""}, {"name":
"ageOfOnset", "type": ["null", "string"], "doc": ""}]}}, "doc": ""}, {"name": "hpoTermList", "type":
{"type": "array", "items": {"type": "record", "name": "HpoTerm", "doc": "", "fields": [{"name":
"term", "type": "string", "doc": ""}, {"name": "termPresence", "type": ["null", "boolean"], "doc":
""}, {"name": "modifiers", "type": ["null", {"type": "map", "values": "string"}], "doc": ""},
{"name": "ageOfOnset", "type": ["null", "string"], "doc": ""}]}}, "doc": ""}, {"name": "ancestries",
"type": {"type": "record", "name": "Ancestries", "fields": [{"name": "mothersEthnicOrigin", "type":
["null", {"type": "enum", "name": "EthnicCategory", "doc": "", "symbols": ["D", "E", "F", "G", "A",
"B", "C", "L", "M", "N", "H", "J", "K", "P", "S", "R", "Z"]}], "doc": ""}, {"name":
"mothersOtherRelevantAncestry", "type": ["null", "string"], "doc": ""}, {"name":
"fathersEthnicOrigin", "type": ["null", "EthnicCategory"], "doc": ""}, {"name":
"fathersOtherRelevantAncestry", "type": ["null", "string"], "doc": ""}, {"name":
"chiSquare1KGenomesPhase3Pop", "type": ["null", {"type": "array", "items": {"type": "record",
"name": "ChiSquare1KGenomesPhase3Pop", "doc": "", "fields": [{"name": "kGSuperPopCategory", "type":
{"type": "enum", "name": "KGSuperPopCategory", "doc": "", "symbols": ["AFR", "AMR", "EAS", "EUR",
"SAS"]}, "doc": ""}, {"name": "kGPopCategory", "type": ["null", {"type": "enum", "name":
"KGPopCategory", "doc": "", "symbols": ["ACB", "ASW", "BEB", "CDX", "CEU", "CHB", "CHS", "CLM",
"ESN", "FIN", "GBR", "GIH", "GWD", "IBS", "ITU", "JPT", "KHV", "LWK", "MSL", "MXL", "PEL", "PJL",
"PUR", "STU", "TSI", "YRI"]}], "doc": ""}, {"name": "chiSquare", "type": "float", "doc": ""}]}}],
"doc": ""}]}, "doc": ""}, {"name": "dataModelCatalogueVersion", "type": "string", "doc": "",
"default": "v4.2"}, {"name": "consentStatus", "type": {"type": "record", "name": "ConsentStatus",
"doc": "", "fields": [{"name": "programmeConsent", "type": "boolean", "doc": "", "default": false},
{"name": "primaryFindingConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"secondaryFindingConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"carrierStatusConsent", "type": "boolean", "doc": "", "default": false}]}, "doc": ""}, {"name":
"samples", "type": ["null", {"type": "array", "items": "string"}], "doc": ""}, {"name":
"inbreedingCoefficient", "type": ["null", {"type": "record", "name": "InbreedingCoefficient", "doc":
"", "fields": [{"name": "sampleId", "type": "string", "doc": ""}, {"name": "program", "type":
"string", "doc": ""}, {"name": "version", "type": "string", "doc": ""}, {"name": "estimationMethod",
"type": "string", "doc": ""}, {"name": "coefficient", "type": "double", "doc": ""}, {"name":
"standardError", "type": ["null", "double"], "doc": ""}]}], "doc": ""}, {"name":
"additionalInformation", "type": ["null", {"type": "map", "values": "string"}], "doc": ""}]}}},
{"name": "analysisPanels", "type": ["null", {"type": "array", "items": {"type": "record", "name":
"AnalysisPanel", "fields": [{"name": "specificDisease", "type": "string"}, {"name": "panelName",
"type": "string"}, {"name": "panelVersion", "type": ["null", "string"]}, {"name": "review_outcome",
"type": "string"}, {"name": "multiple_genetic_origins", "type": "string"}]}}]}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "analysisPanels",
        "gelFamilyId",
        "participants",
        "versionControl",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'analysisPanels': AnalysisPanel,
            'participants': RDParticipant,
            'versionControl': VersionControl,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'analysisPanels': AnalysisPanel,
            'participants': RDParticipant,
            'versionControl': VersionControl,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'analysisPanels', 'gelFamilyId', 'participants',
        'versionControl'
    ]

    def __init__(self, **kwargs):
        self.analysisPanels = kwargs.get(
            'analysisPanels', None)
        self.gelFamilyId = kwargs.get(
            'gelFamilyId', None)
        self.participants = kwargs.get(
            'participants', None)
        self.versionControl = kwargs.get(
            'versionControl', VersionControl())


class Penetrance(object):
    """
    Penetrance assumed in the analysis
    """
    complete = "complete"
    incomplete = "incomplete"

    def __hash__(self):
        return str(self).__hash__()


class PersonKaryotipicSex(object):
    """
    Karyotipic Sex
    """
    unknown = "unknown"
    XX = "XX"
    XY = "XY"
    XO = "XO"
    XXY = "XXY"
    XXX = "XXX"
    XXYY = "XXYY"
    XXXY = "XXXY"
    XXXX = "XXXX"
    XYY = "XYY"
    other = "other"

    def __hash__(self):
        return str(self).__hash__()


class Phase(object):
    """
    No documentation
    """
    PRIMARY = "PRIMARY"
    METASTATIC = "METASTATIC"
    RECURRENCE = "RECURRENCE"

    def __hash__(self):
        return str(self).__hash__()


class PlinkROH(ProtocolElement):
    """
    Plink runs of homozygosity
    """
    _schemaSource = """
{"type": "record", "name": "PlinkROH", "namespace": "Gel_BioInf_Models", "doc": "", "fields":
[{"name": "FID", "type": "string"}, {"name": "IID", "type": "string"}, {"name": "PHE", "type":
"double"}, {"name": "NSEG", "type": "double"}, {"name": "KB", "type": "double"}, {"name": "KBAVG",
"type": "double"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "FID",
        "IID",
        "KB",
        "KBAVG",
        "NSEG",
        "PHE",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'FID', 'IID', 'KB', 'KBAVG', 'NSEG', 'PHE'
    ]

    def __init__(self, **kwargs):
        self.FID = kwargs.get(
            'FID', None)
        self.IID = kwargs.get(
            'IID', None)
        self.KB = kwargs.get(
            'KB', None)
        self.KBAVG = kwargs.get(
            'KBAVG', None)
        self.NSEG = kwargs.get(
            'NSEG', None)
        self.PHE = kwargs.get(
            'PHE', None)


class PlinkSexCheck(ProtocolElement):
    """
    Plink sex check
    """
    _schemaSource = """
{"type": "record", "name": "PlinkSexCheck", "namespace": "Gel_BioInf_Models", "doc": "", "fields":
[{"name": "FID", "type": "string"}, {"name": "IID", "type": "string"}, {"name": "F", "type":
"double"}, {"name": "YCOUNT", "type": "double"}, {"name": "STATUS", "type": "string"}, {"name":
"SNPSEX", "type": "double"}, {"name": "PEDSEX", "type": "double"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "F",
        "FID",
        "IID",
        "PEDSEX",
        "SNPSEX",
        "STATUS",
        "YCOUNT",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'F', 'FID', 'IID', 'PEDSEX', 'SNPSEX', 'STATUS', 'YCOUNT'
    ]

    def __init__(self, **kwargs):
        self.F = kwargs.get(
            'F', None)
        self.FID = kwargs.get(
            'FID', None)
        self.IID = kwargs.get(
            'IID', None)
        self.PEDSEX = kwargs.get(
            'PEDSEX', None)
        self.SNPSEX = kwargs.get(
            'SNPSEX', None)
        self.STATUS = kwargs.get(
            'STATUS', None)
        self.YCOUNT = kwargs.get(
            'YCOUNT', None)


class PreservationMethod(object):
    """
    No documentation
    """
    FFPE = "FFPE"
    FF = "FF"
    UNKNOWN = "UNKNOWN"
    BLOOD = "BLOOD"
    GL = "GL"
    SALIVA = "SALIVA"
    LEUK = "LEUK"

    def __hash__(self):
        return str(self).__hash__()


class RDParticipant(ProtocolElement):
    """
    This defines a RD Participant (demographics and pedigree
    information)
    """
    _schemaSource = """
{"type": "record", "name": "RDParticipant", "namespace": "Gel_BioInf_Models", "doc": "", "fields":
[{"name": "versionControl", "type": {"type": "record", "name": "VersionControl", "fields": [{"name":
"GitVersionControl", "type": "string", "doc": "", "default": "2.1.0"}]}, "doc": ""}, {"name":
"pedigreeId", "type": "int", "doc": ""}, {"name": "isProband", "type": "boolean", "doc": ""},
{"name": "gelId", "type": ["null", "string"], "doc": ""}, {"name": "gelFamilyId", "type": "string",
"doc": ""}, {"name": "gelSuperFamilyId", "type": ["null", "string"], "doc": ""}, {"name": "sex",
"type": {"type": "enum", "name": "Sex", "doc": "", "symbols": ["male", "female", "unknown",
"undetermined"]}, "doc": ""}, {"name": "personKaryotipicSex", "type": ["null", {"type": "enum",
"name": "PersonKaryotipicSex", "doc": "", "symbols": ["unknown", "XX", "XY", "XO", "XXY", "XXX",
"XXYY", "XXXY", "XXXX", "XYY", "other"]}], "doc": ""}, {"name": "fatherId", "type": ["null", "int"],
"doc": ""}, {"name": "motherId", "type": ["null", "int"], "doc": ""}, {"name": "superFatherId",
"type": ["null", "int"], "doc": ""}, {"name": "superMotherId", "type": ["null", "int"], "doc": ""},
{"name": "twinGroup", "type": ["null", "int"], "doc": ""}, {"name": "monozygotic", "type": ["null",
{"type": "enum", "name": "TernaryOption", "doc": "", "symbols": ["yes", "no", "unknown"]}], "doc":
""}, {"name": "adoptedStatus", "type": {"type": "enum", "name": "AdoptedStatus", "doc": "",
"symbols": ["not_adopted", "adoptedin", "adoptedout"]}, "doc": ""}, {"name": "lifeStatus", "type":
{"type": "enum", "name": "LifeStatus", "doc": "", "symbols": ["alive", "aborted", "deceased",
"unborn", "stillborn", "miscarriage"]}, "doc": ""}, {"name": "consanguineousParents", "type":
"TernaryOption", "doc": ""}, {"name": "consanguineousPopulation", "type": ["null", "string"], "doc":
""}, {"name": "affectionStatus", "type": {"type": "enum", "name": "AffectionStatus", "doc": "",
"symbols": ["unaffected", "affected", "unknown"]}, "doc": ""}, {"name": "disorderList", "type":
{"type": "array", "items": {"type": "record", "name": "Disorder", "doc": "", "fields": [{"name":
"diseaseGroup", "type": ["null", "string"], "doc": ""}, {"name": "diseaseSubGroup", "type": ["null",
"string"], "doc": ""}, {"name": "specificDisease", "type": ["null", "string"], "doc": ""}, {"name":
"ageOfOnset", "type": ["null", "string"], "doc": ""}]}}, "doc": ""}, {"name": "hpoTermList", "type":
{"type": "array", "items": {"type": "record", "name": "HpoTerm", "doc": "", "fields": [{"name":
"term", "type": "string", "doc": ""}, {"name": "termPresence", "type": ["null", "boolean"], "doc":
""}, {"name": "modifiers", "type": ["null", {"type": "map", "values": "string"}], "doc": ""},
{"name": "ageOfOnset", "type": ["null", "string"], "doc": ""}]}}, "doc": ""}, {"name": "ancestries",
"type": {"type": "record", "name": "Ancestries", "fields": [{"name": "mothersEthnicOrigin", "type":
["null", {"type": "enum", "name": "EthnicCategory", "doc": "", "symbols": ["D", "E", "F", "G", "A",
"B", "C", "L", "M", "N", "H", "J", "K", "P", "S", "R", "Z"]}], "doc": ""}, {"name":
"mothersOtherRelevantAncestry", "type": ["null", "string"], "doc": ""}, {"name":
"fathersEthnicOrigin", "type": ["null", "EthnicCategory"], "doc": ""}, {"name":
"fathersOtherRelevantAncestry", "type": ["null", "string"], "doc": ""}, {"name":
"chiSquare1KGenomesPhase3Pop", "type": ["null", {"type": "array", "items": {"type": "record",
"name": "ChiSquare1KGenomesPhase3Pop", "doc": "", "fields": [{"name": "kGSuperPopCategory", "type":
{"type": "enum", "name": "KGSuperPopCategory", "doc": "", "symbols": ["AFR", "AMR", "EAS", "EUR",
"SAS"]}, "doc": ""}, {"name": "kGPopCategory", "type": ["null", {"type": "enum", "name":
"KGPopCategory", "doc": "", "symbols": ["ACB", "ASW", "BEB", "CDX", "CEU", "CHB", "CHS", "CLM",
"ESN", "FIN", "GBR", "GIH", "GWD", "IBS", "ITU", "JPT", "KHV", "LWK", "MSL", "MXL", "PEL", "PJL",
"PUR", "STU", "TSI", "YRI"]}], "doc": ""}, {"name": "chiSquare", "type": "float", "doc": ""}]}}],
"doc": ""}]}, "doc": ""}, {"name": "dataModelCatalogueVersion", "type": "string", "doc": "",
"default": "v4.2"}, {"name": "consentStatus", "type": {"type": "record", "name": "ConsentStatus",
"doc": "", "fields": [{"name": "programmeConsent", "type": "boolean", "doc": "", "default": false},
{"name": "primaryFindingConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"secondaryFindingConsent", "type": "boolean", "doc": "", "default": false}, {"name":
"carrierStatusConsent", "type": "boolean", "doc": "", "default": false}]}, "doc": ""}, {"name":
"samples", "type": ["null", {"type": "array", "items": "string"}], "doc": ""}, {"name":
"inbreedingCoefficient", "type": ["null", {"type": "record", "name": "InbreedingCoefficient", "doc":
"", "fields": [{"name": "sampleId", "type": "string", "doc": ""}, {"name": "program", "type":
"string", "doc": ""}, {"name": "version", "type": "string", "doc": ""}, {"name": "estimationMethod",
"type": "string", "doc": ""}, {"name": "coefficient", "type": "double", "doc": ""}, {"name":
"standardError", "type": ["null", "double"], "doc": ""}]}], "doc": ""}, {"name":
"additionalInformation", "type": ["null", {"type": "map", "values": "string"}], "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "additionalInformation",
        "adoptedStatus",
        "affectionStatus",
        "ancestries",
        "consanguineousParents",
        "consanguineousPopulation",
        "consentStatus",
        "disorderList",
        "fatherId",
        "gelFamilyId",
        "gelId",
        "gelSuperFamilyId",
        "hpoTermList",
        "inbreedingCoefficient",
        "isProband",
        "lifeStatus",
        "monozygotic",
        "motherId",
        "pedigreeId",
        "personKaryotipicSex",
        "samples",
        "sex",
        "superFatherId",
        "superMotherId",
        "twinGroup",
        "versionControl",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'ancestries': Ancestries,
            'consentStatus': ConsentStatus,
            'disorderList': Disorder,
            'hpoTermList': HpoTerm,
            'inbreedingCoefficient': InbreedingCoefficient,
            'versionControl': VersionControl,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'ancestries': Ancestries,
            'consentStatus': ConsentStatus,
            'disorderList': Disorder,
            'hpoTermList': HpoTerm,
            'inbreedingCoefficient': InbreedingCoefficient,
            'versionControl': VersionControl,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'additionalInformation', 'adoptedStatus', 'affectionStatus',
        'ancestries', 'consanguineousParents',
        'consanguineousPopulation', 'consentStatus',
        'dataModelCatalogueVersion', 'disorderList', 'fatherId',
        'gelFamilyId', 'gelId', 'gelSuperFamilyId', 'hpoTermList',
        'inbreedingCoefficient', 'isProband', 'lifeStatus',
        'monozygotic', 'motherId', 'pedigreeId',
        'personKaryotipicSex', 'samples', 'sex', 'superFatherId',
        'superMotherId', 'twinGroup', 'versionControl'
    ]

    def __init__(self, **kwargs):
        self.additionalInformation = kwargs.get(
            'additionalInformation', None)
        self.adoptedStatus = kwargs.get(
            'adoptedStatus', None)
        self.affectionStatus = kwargs.get(
            'affectionStatus', None)
        self.ancestries = kwargs.get(
            'ancestries', Ancestries())
        self.consanguineousParents = kwargs.get(
            'consanguineousParents', None)
        self.consanguineousPopulation = kwargs.get(
            'consanguineousPopulation', None)
        self.consentStatus = kwargs.get(
            'consentStatus', ConsentStatus())
        self.dataModelCatalogueVersion = kwargs.get(
            'dataModelCatalogueVersion', 'v4.2')
        self.disorderList = kwargs.get(
            'disorderList', None)
        self.fatherId = kwargs.get(
            'fatherId', None)
        self.gelFamilyId = kwargs.get(
            'gelFamilyId', None)
        self.gelId = kwargs.get(
            'gelId', None)
        self.gelSuperFamilyId = kwargs.get(
            'gelSuperFamilyId', None)
        self.hpoTermList = kwargs.get(
            'hpoTermList', None)
        self.inbreedingCoefficient = kwargs.get(
            'inbreedingCoefficient', None)
        self.isProband = kwargs.get(
            'isProband', None)
        self.lifeStatus = kwargs.get(
            'lifeStatus', None)
        self.monozygotic = kwargs.get(
            'monozygotic', None)
        self.motherId = kwargs.get(
            'motherId', None)
        self.pedigreeId = kwargs.get(
            'pedigreeId', None)
        self.personKaryotipicSex = kwargs.get(
            'personKaryotipicSex', None)
        self.samples = kwargs.get(
            'samples', None)
        self.sex = kwargs.get(
            'sex', None)
        self.superFatherId = kwargs.get(
            'superFatherId', None)
        self.superMotherId = kwargs.get(
            'superMotherId', None)
        self.twinGroup = kwargs.get(
            'twinGroup', None)
        self.versionControl = kwargs.get(
            'versionControl', VersionControl())


class Reason(object):
    """
    No documentation
    """
    median_coverage = "median_coverage"
    in_analysis = "in_analysis"
    duplicate = "duplicate"
    pedigree_mendelian_errors = "pedigree_mendelian_errors"
    pedigree_ibd_sharing = "pedigree_ibd_sharing"
    contamination = "contamination"
    quality = "quality"
    sex_query = "sex_query"
    perc_bases_ge_15x_mapQ_ge11 = "perc_bases_ge_15x_mapQ_ge11"
    GbQ30NoDupsNoClip = "GbQ30NoDupsNoClip"
    arrayconcordance = "arrayconcordance"
    high_cnv = "high_cnv"
    in_qc = "in_qc"
    pass_qc = "pass_qc"
    other = "other"

    def __hash__(self):
        return str(self).__hash__()


class ReportEvent(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "ReportEvent", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"reportEventId", "type": "string", "doc": ""}, {"name": "phenotype", "type": "string", "doc": ""},
{"name": "panelName", "type": ["null", "string"], "doc": ""}, {"name": "panelVersion", "type":
["null", "string"], "doc": ""}, {"name": "modeOfInheritance", "type": {"type": "enum", "name":
"ReportedModeOfInheritance", "doc": "", "symbols": ["monoallelic", "monoallelic_not_imprinted",
"monoallelic_maternally_imprinted", "monoallelic_paternally_imprinted", "biallelic",
"monoallelic_and_biallelic", "monoallelic_and_more_severe_biallelic", "xlinked_biallelic",
"xlinked_monoallelic", "mitochondrial", "unknown"]}, "doc": ""}, {"name": "genomicFeature", "type":
{"type": "record", "name": "GenomicFeature", "fields": [{"name": "featureType", "type": {"type":
"enum", "name": "FeatureTypes", "symbols": ["RegulatoryRegion", "Gene", "Transcript"]}, "doc": ""},
{"name": "ensemblId", "type": "string", "doc": ""}, {"name": "ids", "type": {"type": "map",
"values": "string"}, "doc": ""}]}, "doc": ""}, {"name": "penetrance", "type": {"type": "enum",
"name": "Penetrance", "doc": "", "symbols": ["complete", "incomplete"]}, "doc": ""}, {"name":
"score", "type": "float", "doc": ""}, {"name": "vendorSpecificScores", "type": ["null", {"type":
"map", "values": "float"}], "doc": ""}, {"name": "variantClassification", "type": ["null", {"type":
"enum", "name": "VariantClassification", "doc": "", "symbols": ["BENIGN", "LIKELY_BENIGN", "VUS",
"LIKELY_PATHOGENIC", "PATHOGENIC"]}], "doc": ""}, {"name": "fullyExplainsPhenotype", "type":
["null", "boolean"], "doc": ""}, {"name": "groupOfVariants", "type": ["null", "int"], "doc": ""},
{"name": "eventJustification", "type": "string", "doc": ""}, {"name": "tier", "type": ["null",
{"type": "enum", "name": "Tier", "doc": "", "symbols": ["NONE", "TIER1", "TIER2", "TIER3"]}], "doc":
""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "eventJustification",
        "fullyExplainsPhenotype",
        "genomicFeature",
        "groupOfVariants",
        "modeOfInheritance",
        "panelName",
        "panelVersion",
        "penetrance",
        "phenotype",
        "reportEventId",
        "score",
        "tier",
        "variantClassification",
        "vendorSpecificScores",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'genomicFeature': GenomicFeature,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'genomicFeature': GenomicFeature,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'eventJustification', 'fullyExplainsPhenotype',
        'genomicFeature', 'groupOfVariants', 'modeOfInheritance',
        'panelName', 'panelVersion', 'penetrance', 'phenotype',
        'reportEventId', 'score', 'tier', 'variantClassification',
        'vendorSpecificScores'
    ]

    def __init__(self, **kwargs):
        self.eventJustification = kwargs.get(
            'eventJustification', None)
        self.fullyExplainsPhenotype = kwargs.get(
            'fullyExplainsPhenotype', None)
        self.genomicFeature = kwargs.get(
            'genomicFeature', GenomicFeature())
        self.groupOfVariants = kwargs.get(
            'groupOfVariants', None)
        self.modeOfInheritance = kwargs.get(
            'modeOfInheritance', None)
        self.panelName = kwargs.get(
            'panelName', None)
        self.panelVersion = kwargs.get(
            'panelVersion', None)
        self.penetrance = kwargs.get(
            'penetrance', None)
        self.phenotype = kwargs.get(
            'phenotype', None)
        self.reportEventId = kwargs.get(
            'reportEventId', None)
        self.score = kwargs.get(
            'score', None)
        self.tier = kwargs.get(
            'tier', None)
        self.variantClassification = kwargs.get(
            'variantClassification', None)
        self.vendorSpecificScores = kwargs.get(
            'vendorSpecificScores', None)


class ReportEventCancer(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "ReportEventCancer", "namespace": "Gel_BioInf_Models", "fields":
[{"name": "reportEventId", "type": "string", "doc": ""}, {"name": "genomicFeatureCancer", "type":
{"type": "record", "name": "GenomicFeatureCancer", "fields": [{"name": "featureType", "type":
{"type": "enum", "name": "FeatureTypes", "symbols": ["RegulatoryRegion", "Gene", "Transcript"]},
"doc": ""}, {"name": "ensemblId", "type": "string", "doc": ""}, {"name": "refSeqTranscriptId",
"type": "string", "doc": ""}, {"name": "refSeqProteinId", "type": "string", "doc": ""}, {"name":
"geneName", "type": "string", "doc": ""}, {"name": "roleInCancer", "type": ["null", "string"],
"doc": ""}]}, "doc": ""}, {"name": "soTerm", "type": "string", "doc": ""}, {"name": "soName",
"type": "string", "doc": ""}, {"name": "actions", "type": {"type": "array", "items": {"type":
"record", "name": "Actions", "fields": [{"name": "actionType", "type": {"type": "enum", "name":
"ActionType", "doc": "", "symbols": ["Therapy", "Prognosis", "Diagnosis"]}}, {"name": "tumorType",
"type": "string", "doc": ""}, {"name": "evidence", "type": ["null", {"type": "array", "items":
"string"}], "doc": ""}, {"name": "drug", "type": ["null", "string"], "doc": ""}, {"name": "status",
"type": ["null", "string"], "doc": ""}, {"name": "comments", "type": ["null", {"type": "array",
"items": "string"}], "doc": ""}, {"name": "evidenceType", "type": ["null", "string"], "doc": ""},
{"name": "source", "type": "string", "doc": ""}]}}, "doc": ""}, {"name": "tier", "type": ["null",
{"type": "enum", "name": "Tier", "doc": "", "symbols": ["NONE", "TIER1", "TIER2", "TIER3"]}], "doc":
""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "actions",
        "genomicFeatureCancer",
        "reportEventId",
        "soName",
        "soTerm",
        "tier",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'actions': Actions,
            'genomicFeatureCancer': GenomicFeatureCancer,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'actions': Actions,
            'genomicFeatureCancer': GenomicFeatureCancer,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'actions', 'genomicFeatureCancer', 'reportEventId', 'soName',
        'soTerm', 'tier'
    ]

    def __init__(self, **kwargs):
        self.actions = kwargs.get(
            'actions', None)
        self.genomicFeatureCancer = kwargs.get(
            'genomicFeatureCancer', GenomicFeatureCancer())
        self.reportEventId = kwargs.get(
            'reportEventId', None)
        self.soName = kwargs.get(
            'soName', None)
        self.soTerm = kwargs.get(
            'soTerm', None)
        self.tier = kwargs.get(
            'tier', None)


class ReportedModeOfInheritance(object):
    """
    An enumeration for the different mode of inheritances:  *
    `monoallelic_not_imprinted`: MONOALLELIC, autosomal or
    pseudoautosomal, not imprinted *
    `monoallelic_maternally_imprinted`: MONOALLELIC, autosomal or
    pseudoautosomal, maternally imprinted (paternal allele expressed)
    * `monoallelic_paternally_imprinted`: MONOALLELIC, autosomal or
    pseudoautosomal, paternally imprinted (maternal allele expressed)
    * `monoallelic`: MONOALLELIC, autosomal or pseudoautosomal,
    imprinted status unknown * `biallelic`: BIALLELIC, autosomal or
    pseudoautosomal * `monoallelic_and_biallelic`: BOTH monoallelic
    and biallelic, autosomal or pseudoautosomal *
    `monoallelic_and_more_severe_biallelic`: BOTH monoallelic and
    biallelic, autosomal or pseudoautosomal (but BIALLELIC mutations
    cause a more SEVERE disease form), autosomal or pseudoautosomal *
    `xlinked_biallelic`: X-LINKED: hemizygous mutation in males,
    biallelic mutations in females * `xlinked_monoallelic`: X linked:
    hemizygous mutation in males, monoallelic mutations in females may
    cause disease (may be less severe, later onset than males) *
    `mitochondrial`: MITOCHONDRIAL * `unknown`: Unknown
    """
    monoallelic = "monoallelic"
    monoallelic_not_imprinted = "monoallelic_not_imprinted"
    monoallelic_maternally_imprinted = "monoallelic_maternally_imprinted"
    monoallelic_paternally_imprinted = "monoallelic_paternally_imprinted"
    biallelic = "biallelic"
    monoallelic_and_biallelic = "monoallelic_and_biallelic"
    monoallelic_and_more_severe_biallelic = "monoallelic_and_more_severe_biallelic"
    xlinked_biallelic = "xlinked_biallelic"
    xlinked_monoallelic = "xlinked_monoallelic"
    mitochondrial = "mitochondrial"
    unknown = "unknown"

    def __hash__(self):
        return str(self).__hash__()


class ReportedSomaticStructuralVariants(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "ReportedSomaticStructuralVariants", "namespace": "Gel_BioInf_Models",
"fields": [{"name": "reportedStructuralVariant", "type": {"type": "record", "name":
"ReportedStructuralVariant", "fields": [{"name": "versionControl", "type": {"type": "record",
"name": "VersionControl", "fields": [{"name": "GitVersionControl", "type": "string", "doc": "",
"default": "2.1.0"}]}, "doc": ""}, {"name": "chromosome", "type": "string", "doc": ""}, {"name":
"start", "type": "int", "doc": ""}, {"name": "end", "type": "int", "doc": ""}, {"name": "type",
"type": "string", "doc": ""}, {"name": "reference", "type": "string", "doc": ""}, {"name":
"alternate", "type": "string", "doc": ""}, {"name": "calledGenotypes", "type": {"type": "array",
"items": {"type": "record", "name": "CalledGenotype", "doc": "", "fields": [{"name": "gelId",
"type": "string", "doc": ""}, {"name": "sampleId", "type": "string", "doc": ""}, {"name":
"genotype", "type": "string", "doc": ""}, {"name": "phaseSet", "type": ["null", "int"], "doc": ""},
{"name": "depthReference", "type": ["null", "int"], "doc": ""}, {"name": "depthAlternate", "type":
["null", "int"], "doc": ""}, {"name": "copyNumber", "type": ["null", "int"], "doc": ""}]}}},
{"name": "reportEvents", "type": {"type": "array", "items": {"type": "record", "name":
"ReportEvent", "fields": [{"name": "reportEventId", "type": "string", "doc": ""}, {"name":
"phenotype", "type": "string", "doc": ""}, {"name": "panelName", "type": ["null", "string"], "doc":
""}, {"name": "panelVersion", "type": ["null", "string"], "doc": ""}, {"name": "modeOfInheritance",
"type": {"type": "enum", "name": "ReportedModeOfInheritance", "doc": "", "symbols": ["monoallelic",
"monoallelic_not_imprinted", "monoallelic_maternally_imprinted", "monoallelic_paternally_imprinted",
"biallelic", "monoallelic_and_biallelic", "monoallelic_and_more_severe_biallelic",
"xlinked_biallelic", "xlinked_monoallelic", "mitochondrial", "unknown"]}, "doc": ""}, {"name":
"genomicFeature", "type": {"type": "record", "name": "GenomicFeature", "fields": [{"name":
"featureType", "type": {"type": "enum", "name": "FeatureTypes", "symbols": ["RegulatoryRegion",
"Gene", "Transcript"]}, "doc": ""}, {"name": "ensemblId", "type": "string", "doc": ""}, {"name":
"ids", "type": {"type": "map", "values": "string"}, "doc": ""}]}, "doc": ""}, {"name": "penetrance",
"type": {"type": "enum", "name": "Penetrance", "doc": "", "symbols": ["complete", "incomplete"]},
"doc": ""}, {"name": "score", "type": "float", "doc": ""}, {"name": "vendorSpecificScores", "type":
["null", {"type": "map", "values": "float"}], "doc": ""}, {"name": "variantClassification", "type":
["null", {"type": "enum", "name": "VariantClassification", "doc": "", "symbols": ["BENIGN",
"LIKELY_BENIGN", "VUS", "LIKELY_PATHOGENIC", "PATHOGENIC"]}], "doc": ""}, {"name":
"fullyExplainsPhenotype", "type": ["null", "boolean"], "doc": ""}, {"name": "groupOfVariants",
"type": ["null", "int"], "doc": ""}, {"name": "eventJustification", "type": "string", "doc": ""},
{"name": "tier", "type": ["null", {"type": "enum", "name": "Tier", "doc": "", "symbols": ["NONE",
"TIER1", "TIER2", "TIER3"]}], "doc": ""}]}}, "doc": ""}, {"name":
"additionalTextualVariantAnnotations", "type": ["null", {"type": "map", "values": "string"}], "doc":
""}, {"name": "additionalNumericVariantAnnotations", "type": ["null", {"type": "map", "values":
"float"}], "doc": ""}, {"name": "comments", "type": ["null", {"type": "array", "items": "string"}],
"doc": ""}]}, "doc": ""}, {"name": "somaticOrGermline", "type": {"type": "enum", "name":
"SomaticOrGermline", "doc": "", "symbols": ["unknown", "somatic", "germline"]}, "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "reportedStructuralVariant",
        "somaticOrGermline",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'reportedStructuralVariant': ReportedStructuralVariant,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'reportedStructuralVariant': ReportedStructuralVariant,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'reportedStructuralVariant', 'somaticOrGermline'
    ]

    def __init__(self, **kwargs):
        self.reportedStructuralVariant = kwargs.get(
            'reportedStructuralVariant', ReportedStructuralVariant())
        self.somaticOrGermline = kwargs.get(
            'somaticOrGermline', None)


class ReportedSomaticVariants(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "ReportedSomaticVariants", "namespace": "Gel_BioInf_Models", "fields":
[{"name": "reportedVariant", "type": {"type": "record", "name": "ReportedVariant", "fields":
[{"name": "versionControl", "type": {"type": "record", "name": "VersionControl", "fields": [{"name":
"GitVersionControl", "type": "string", "doc": "", "default": "2.1.0"}]}, "doc": ""}, {"name":
"chromosome", "type": "string", "doc": ""}, {"name": "dbSNPid", "type": ["null", "string"], "doc":
""}, {"name": "position", "type": "int", "doc": ""}, {"name": "reference", "type": "string", "doc":
""}, {"name": "alternate", "type": "string", "doc": ""}, {"name": "calledGenotypes", "type":
{"type": "array", "items": {"type": "record", "name": "CalledGenotype", "doc": "", "fields":
[{"name": "gelId", "type": "string", "doc": ""}, {"name": "sampleId", "type": "string", "doc": ""},
{"name": "genotype", "type": "string", "doc": ""}, {"name": "phaseSet", "type": ["null", "int"],
"doc": ""}, {"name": "depthReference", "type": ["null", "int"], "doc": ""}, {"name":
"depthAlternate", "type": ["null", "int"], "doc": ""}, {"name": "copyNumber", "type": ["null",
"int"], "doc": ""}]}}, "doc": ""}, {"name": "reportEvents", "type": {"type": "array", "items":
{"type": "record", "name": "ReportEvent", "fields": [{"name": "reportEventId", "type": "string",
"doc": ""}, {"name": "phenotype", "type": "string", "doc": ""}, {"name": "panelName", "type":
["null", "string"], "doc": ""}, {"name": "panelVersion", "type": ["null", "string"], "doc": ""},
{"name": "modeOfInheritance", "type": {"type": "enum", "name": "ReportedModeOfInheritance", "doc":
"", "symbols": ["monoallelic", "monoallelic_not_imprinted", "monoallelic_maternally_imprinted",
"monoallelic_paternally_imprinted", "biallelic", "monoallelic_and_biallelic",
"monoallelic_and_more_severe_biallelic", "xlinked_biallelic", "xlinked_monoallelic",
"mitochondrial", "unknown"]}, "doc": ""}, {"name": "genomicFeature", "type": {"type": "record",
"name": "GenomicFeature", "fields": [{"name": "featureType", "type": {"type": "enum", "name":
"FeatureTypes", "symbols": ["RegulatoryRegion", "Gene", "Transcript"]}, "doc": ""}, {"name":
"ensemblId", "type": "string", "doc": ""}, {"name": "ids", "type": {"type": "map", "values":
"string"}, "doc": ""}]}, "doc": ""}, {"name": "penetrance", "type": {"type": "enum", "name":
"Penetrance", "doc": "", "symbols": ["complete", "incomplete"]}, "doc": ""}, {"name": "score",
"type": "float", "doc": ""}, {"name": "vendorSpecificScores", "type": ["null", {"type": "map",
"values": "float"}], "doc": ""}, {"name": "variantClassification", "type": ["null", {"type": "enum",
"name": "VariantClassification", "doc": "", "symbols": ["BENIGN", "LIKELY_BENIGN", "VUS",
"LIKELY_PATHOGENIC", "PATHOGENIC"]}], "doc": ""}, {"name": "fullyExplainsPhenotype", "type":
["null", "boolean"], "doc": ""}, {"name": "groupOfVariants", "type": ["null", "int"], "doc": ""},
{"name": "eventJustification", "type": "string", "doc": ""}, {"name": "tier", "type": ["null",
{"type": "enum", "name": "Tier", "doc": "", "symbols": ["NONE", "TIER1", "TIER2", "TIER3"]}], "doc":
""}]}}, "doc": ""}, {"name": "additionalTextualVariantAnnotations", "type": ["null", {"type": "map",
"values": "string"}], "doc": ""}, {"name": "additionalNumericVariantAnnotations", "type": ["null",
{"type": "map", "values": "float"}], "doc": ""}, {"name": "comments", "type": ["null", {"type":
"array", "items": "string"}], "doc": ""}]}, "doc": ""}, {"name": "somaticOrGermline", "type":
{"type": "enum", "name": "SomaticOrGermline", "doc": "", "symbols": ["unknown", "somatic",
"germline"]}, "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "reportedVariant",
        "somaticOrGermline",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'reportedVariant': ReportedVariant,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'reportedVariant': ReportedVariant,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'reportedVariant', 'somaticOrGermline'
    ]

    def __init__(self, **kwargs):
        self.reportedVariant = kwargs.get(
            'reportedVariant', ReportedVariant())
        self.somaticOrGermline = kwargs.get(
            'somaticOrGermline', None)


class ReportedStructuralVariant(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "ReportedStructuralVariant", "namespace": "Gel_BioInf_Models", "fields":
[{"name": "versionControl", "type": {"type": "record", "name": "VersionControl", "fields": [{"name":
"GitVersionControl", "type": "string", "doc": "", "default": "2.1.0"}]}, "doc": ""}, {"name":
"chromosome", "type": "string", "doc": ""}, {"name": "start", "type": "int", "doc": ""}, {"name":
"end", "type": "int", "doc": ""}, {"name": "type", "type": "string", "doc": ""}, {"name":
"reference", "type": "string", "doc": ""}, {"name": "alternate", "type": "string", "doc": ""},
{"name": "calledGenotypes", "type": {"type": "array", "items": {"type": "record", "name":
"CalledGenotype", "doc": "", "fields": [{"name": "gelId", "type": "string", "doc": ""}, {"name":
"sampleId", "type": "string", "doc": ""}, {"name": "genotype", "type": "string", "doc": ""},
{"name": "phaseSet", "type": ["null", "int"], "doc": ""}, {"name": "depthReference", "type":
["null", "int"], "doc": ""}, {"name": "depthAlternate", "type": ["null", "int"], "doc": ""},
{"name": "copyNumber", "type": ["null", "int"], "doc": ""}]}}}, {"name": "reportEvents", "type":
{"type": "array", "items": {"type": "record", "name": "ReportEvent", "fields": [{"name":
"reportEventId", "type": "string", "doc": ""}, {"name": "phenotype", "type": "string", "doc": ""},
{"name": "panelName", "type": ["null", "string"], "doc": ""}, {"name": "panelVersion", "type":
["null", "string"], "doc": ""}, {"name": "modeOfInheritance", "type": {"type": "enum", "name":
"ReportedModeOfInheritance", "doc": "", "symbols": ["monoallelic", "monoallelic_not_imprinted",
"monoallelic_maternally_imprinted", "monoallelic_paternally_imprinted", "biallelic",
"monoallelic_and_biallelic", "monoallelic_and_more_severe_biallelic", "xlinked_biallelic",
"xlinked_monoallelic", "mitochondrial", "unknown"]}, "doc": ""}, {"name": "genomicFeature", "type":
{"type": "record", "name": "GenomicFeature", "fields": [{"name": "featureType", "type": {"type":
"enum", "name": "FeatureTypes", "symbols": ["RegulatoryRegion", "Gene", "Transcript"]}, "doc": ""},
{"name": "ensemblId", "type": "string", "doc": ""}, {"name": "ids", "type": {"type": "map",
"values": "string"}, "doc": ""}]}, "doc": ""}, {"name": "penetrance", "type": {"type": "enum",
"name": "Penetrance", "doc": "", "symbols": ["complete", "incomplete"]}, "doc": ""}, {"name":
"score", "type": "float", "doc": ""}, {"name": "vendorSpecificScores", "type": ["null", {"type":
"map", "values": "float"}], "doc": ""}, {"name": "variantClassification", "type": ["null", {"type":
"enum", "name": "VariantClassification", "doc": "", "symbols": ["BENIGN", "LIKELY_BENIGN", "VUS",
"LIKELY_PATHOGENIC", "PATHOGENIC"]}], "doc": ""}, {"name": "fullyExplainsPhenotype", "type":
["null", "boolean"], "doc": ""}, {"name": "groupOfVariants", "type": ["null", "int"], "doc": ""},
{"name": "eventJustification", "type": "string", "doc": ""}, {"name": "tier", "type": ["null",
{"type": "enum", "name": "Tier", "doc": "", "symbols": ["NONE", "TIER1", "TIER2", "TIER3"]}], "doc":
""}]}}, "doc": ""}, {"name": "additionalTextualVariantAnnotations", "type": ["null", {"type": "map",
"values": "string"}], "doc": ""}, {"name": "additionalNumericVariantAnnotations", "type": ["null",
{"type": "map", "values": "float"}], "doc": ""}, {"name": "comments", "type": ["null", {"type":
"array", "items": "string"}], "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "additionalNumericVariantAnnotations",
        "additionalTextualVariantAnnotations",
        "alternate",
        "calledGenotypes",
        "chromosome",
        "comments",
        "end",
        "reference",
        "reportEvents",
        "start",
        "type",
        "versionControl",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'calledGenotypes': CalledGenotype,
            'reportEvents': ReportEvent,
            'versionControl': VersionControl,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'calledGenotypes': CalledGenotype,
            'reportEvents': ReportEvent,
            'versionControl': VersionControl,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'additionalNumericVariantAnnotations',
        'additionalTextualVariantAnnotations', 'alternate',
        'calledGenotypes', 'chromosome', 'comments', 'end',
        'reference', 'reportEvents', 'start', 'type', 'versionControl'
    ]

    def __init__(self, **kwargs):
        self.additionalNumericVariantAnnotations = kwargs.get(
            'additionalNumericVariantAnnotations', None)
        self.additionalTextualVariantAnnotations = kwargs.get(
            'additionalTextualVariantAnnotations', None)
        self.alternate = kwargs.get(
            'alternate', None)
        self.calledGenotypes = kwargs.get(
            'calledGenotypes', None)
        self.chromosome = kwargs.get(
            'chromosome', None)
        self.comments = kwargs.get(
            'comments', None)
        self.end = kwargs.get(
            'end', None)
        self.reference = kwargs.get(
            'reference', None)
        self.reportEvents = kwargs.get(
            'reportEvents', None)
        self.start = kwargs.get(
            'start', None)
        self.type = kwargs.get(
            'type', None)
        self.versionControl = kwargs.get(
            'versionControl', VersionControl())


class ReportedStructuralVariantCancer(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "ReportedStructuralVariantCancer", "namespace": "Gel_BioInf_Models",
"fields": [{"name": "chromosome", "type": "string", "doc": ""}, {"name": "start", "type": "int",
"doc": ""}, {"name": "end", "type": "int", "doc": ""}, {"name": "type", "type": "string", "doc":
""}, {"name": "reference", "type": "string", "doc": ""}, {"name": "alternate", "type": "string",
"doc": ""}, {"name": "additionalTextualVariantAnnotations", "type": ["null", {"type": "map",
"values": "string"}], "doc": ""}, {"name": "additionalNumericVariantAnnotations", "type": ["null",
{"type": "map", "values": "float"}], "doc": ""}, {"name": "comments", "type": ["null", {"type":
"array", "items": "string"}], "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "additionalNumericVariantAnnotations",
        "additionalTextualVariantAnnotations",
        "alternate",
        "chromosome",
        "comments",
        "end",
        "reference",
        "start",
        "type",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'additionalNumericVariantAnnotations',
        'additionalTextualVariantAnnotations', 'alternate',
        'chromosome', 'comments', 'end', 'reference', 'start', 'type'
    ]

    def __init__(self, **kwargs):
        self.additionalNumericVariantAnnotations = kwargs.get(
            'additionalNumericVariantAnnotations', None)
        self.additionalTextualVariantAnnotations = kwargs.get(
            'additionalTextualVariantAnnotations', None)
        self.alternate = kwargs.get(
            'alternate', None)
        self.chromosome = kwargs.get(
            'chromosome', None)
        self.comments = kwargs.get(
            'comments', None)
        self.end = kwargs.get(
            'end', None)
        self.reference = kwargs.get(
            'reference', None)
        self.start = kwargs.get(
            'start', None)
        self.type = kwargs.get(
            'type', None)


class ReportedVariant(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "ReportedVariant", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"versionControl", "type": {"type": "record", "name": "VersionControl", "fields": [{"name":
"GitVersionControl", "type": "string", "doc": "", "default": "2.1.0"}]}, "doc": ""}, {"name":
"chromosome", "type": "string", "doc": ""}, {"name": "dbSNPid", "type": ["null", "string"], "doc":
""}, {"name": "position", "type": "int", "doc": ""}, {"name": "reference", "type": "string", "doc":
""}, {"name": "alternate", "type": "string", "doc": ""}, {"name": "calledGenotypes", "type":
{"type": "array", "items": {"type": "record", "name": "CalledGenotype", "doc": "", "fields":
[{"name": "gelId", "type": "string", "doc": ""}, {"name": "sampleId", "type": "string", "doc": ""},
{"name": "genotype", "type": "string", "doc": ""}, {"name": "phaseSet", "type": ["null", "int"],
"doc": ""}, {"name": "depthReference", "type": ["null", "int"], "doc": ""}, {"name":
"depthAlternate", "type": ["null", "int"], "doc": ""}, {"name": "copyNumber", "type": ["null",
"int"], "doc": ""}]}}, "doc": ""}, {"name": "reportEvents", "type": {"type": "array", "items":
{"type": "record", "name": "ReportEvent", "fields": [{"name": "reportEventId", "type": "string",
"doc": ""}, {"name": "phenotype", "type": "string", "doc": ""}, {"name": "panelName", "type":
["null", "string"], "doc": ""}, {"name": "panelVersion", "type": ["null", "string"], "doc": ""},
{"name": "modeOfInheritance", "type": {"type": "enum", "name": "ReportedModeOfInheritance", "doc":
"", "symbols": ["monoallelic", "monoallelic_not_imprinted", "monoallelic_maternally_imprinted",
"monoallelic_paternally_imprinted", "biallelic", "monoallelic_and_biallelic",
"monoallelic_and_more_severe_biallelic", "xlinked_biallelic", "xlinked_monoallelic",
"mitochondrial", "unknown"]}, "doc": ""}, {"name": "genomicFeature", "type": {"type": "record",
"name": "GenomicFeature", "fields": [{"name": "featureType", "type": {"type": "enum", "name":
"FeatureTypes", "symbols": ["RegulatoryRegion", "Gene", "Transcript"]}, "doc": ""}, {"name":
"ensemblId", "type": "string", "doc": ""}, {"name": "ids", "type": {"type": "map", "values":
"string"}, "doc": ""}]}, "doc": ""}, {"name": "penetrance", "type": {"type": "enum", "name":
"Penetrance", "doc": "", "symbols": ["complete", "incomplete"]}, "doc": ""}, {"name": "score",
"type": "float", "doc": ""}, {"name": "vendorSpecificScores", "type": ["null", {"type": "map",
"values": "float"}], "doc": ""}, {"name": "variantClassification", "type": ["null", {"type": "enum",
"name": "VariantClassification", "doc": "", "symbols": ["BENIGN", "LIKELY_BENIGN", "VUS",
"LIKELY_PATHOGENIC", "PATHOGENIC"]}], "doc": ""}, {"name": "fullyExplainsPhenotype", "type":
["null", "boolean"], "doc": ""}, {"name": "groupOfVariants", "type": ["null", "int"], "doc": ""},
{"name": "eventJustification", "type": "string", "doc": ""}, {"name": "tier", "type": ["null",
{"type": "enum", "name": "Tier", "doc": "", "symbols": ["NONE", "TIER1", "TIER2", "TIER3"]}], "doc":
""}]}}, "doc": ""}, {"name": "additionalTextualVariantAnnotations", "type": ["null", {"type": "map",
"values": "string"}], "doc": ""}, {"name": "additionalNumericVariantAnnotations", "type": ["null",
{"type": "map", "values": "float"}], "doc": ""}, {"name": "comments", "type": ["null", {"type":
"array", "items": "string"}], "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "additionalNumericVariantAnnotations",
        "additionalTextualVariantAnnotations",
        "alternate",
        "calledGenotypes",
        "chromosome",
        "comments",
        "dbSNPid",
        "position",
        "reference",
        "reportEvents",
        "versionControl",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'calledGenotypes': CalledGenotype,
            'reportEvents': ReportEvent,
            'versionControl': VersionControl,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'calledGenotypes': CalledGenotype,
            'reportEvents': ReportEvent,
            'versionControl': VersionControl,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'additionalNumericVariantAnnotations',
        'additionalTextualVariantAnnotations', 'alternate',
        'calledGenotypes', 'chromosome', 'comments', 'dbSNPid',
        'position', 'reference', 'reportEvents', 'versionControl'
    ]

    def __init__(self, **kwargs):
        self.additionalNumericVariantAnnotations = kwargs.get(
            'additionalNumericVariantAnnotations', None)
        self.additionalTextualVariantAnnotations = kwargs.get(
            'additionalTextualVariantAnnotations', None)
        self.alternate = kwargs.get(
            'alternate', None)
        self.calledGenotypes = kwargs.get(
            'calledGenotypes', None)
        self.chromosome = kwargs.get(
            'chromosome', None)
        self.comments = kwargs.get(
            'comments', None)
        self.dbSNPid = kwargs.get(
            'dbSNPid', None)
        self.position = kwargs.get(
            'position', None)
        self.reference = kwargs.get(
            'reference', None)
        self.reportEvents = kwargs.get(
            'reportEvents', None)
        self.versionControl = kwargs.get(
            'versionControl', VersionControl())


class ReportedVariantCancer(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "ReportedVariantCancer", "namespace": "Gel_BioInf_Models", "fields":
[{"name": "chromosome", "type": "string", "doc": ""}, {"name": "position", "type": "int", "doc":
""}, {"name": "reference", "type": "string", "doc": ""}, {"name": "alternate", "type": "string",
"doc": ""}, {"name": "CosmicId", "type": ["null", "string"], "doc": ""}, {"name": "CosmicLink",
"type": ["null", "string"], "doc": ""}, {"name": "dbSNPid", "type": ["null", "string"], "doc": ""},
{"name": "cDNAchange", "type": ["null", "string"], "doc": ""}, {"name": "proteinChange", "type":
["null", "string"], "doc": ""}, {"name": "depthReference", "type": ["null", "int"], "doc": ""},
{"name": "depthAlternate", "type": ["null", "int"], "doc": ""}, {"name": "VAF", "type": ["null",
"int"], "doc": ""}, {"name": "commonAF", "type": ["null", "int"], "doc": ""}, {"name":
"additionalTextualVariantAnnotations", "type": ["null", {"type": "map", "values": "string"}], "doc":
""}, {"name": "additionalNumericVariantAnnotations", "type": ["null", {"type": "map", "values":
"float"}], "doc": ""}, {"name": "comments", "type": ["null", {"type": "array", "items": "string"}],
"doc": ""}, {"name": "reportEvents", "type": {"type": "array", "items": {"type": "record", "name":
"ReportEventCancer", "fields": [{"name": "reportEventId", "type": "string", "doc": ""}, {"name":
"genomicFeatureCancer", "type": {"type": "record", "name": "GenomicFeatureCancer", "fields":
[{"name": "featureType", "type": {"type": "enum", "name": "FeatureTypes", "symbols":
["RegulatoryRegion", "Gene", "Transcript"]}, "doc": ""}, {"name": "ensemblId", "type": "string",
"doc": ""}, {"name": "refSeqTranscriptId", "type": "string", "doc": ""}, {"name": "refSeqProteinId",
"type": "string", "doc": ""}, {"name": "geneName", "type": "string", "doc": ""}, {"name":
"roleInCancer", "type": ["null", "string"], "doc": ""}]}, "doc": ""}, {"name": "soTerm", "type":
"string", "doc": ""}, {"name": "soName", "type": "string", "doc": ""}, {"name": "actions", "type":
{"type": "array", "items": {"type": "record", "name": "Actions", "fields": [{"name": "actionType",
"type": {"type": "enum", "name": "ActionType", "doc": "", "symbols": ["Therapy", "Prognosis",
"Diagnosis"]}}, {"name": "tumorType", "type": "string", "doc": ""}, {"name": "evidence", "type":
["null", {"type": "array", "items": "string"}], "doc": ""}, {"name": "drug", "type": ["null",
"string"], "doc": ""}, {"name": "status", "type": ["null", "string"], "doc": ""}, {"name":
"comments", "type": ["null", {"type": "array", "items": "string"}], "doc": ""}, {"name":
"evidenceType", "type": ["null", "string"], "doc": ""}, {"name": "source", "type": "string", "doc":
""}]}}, "doc": ""}, {"name": "tier", "type": ["null", {"type": "enum", "name": "Tier", "doc": "",
"symbols": ["NONE", "TIER1", "TIER2", "TIER3"]}], "doc": ""}]}}}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "CosmicId",
        "CosmicLink",
        "VAF",
        "additionalNumericVariantAnnotations",
        "additionalTextualVariantAnnotations",
        "alternate",
        "cDNAchange",
        "chromosome",
        "comments",
        "commonAF",
        "dbSNPid",
        "depthAlternate",
        "depthReference",
        "position",
        "proteinChange",
        "reference",
        "reportEvents",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'reportEvents': ReportEventCancer,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'reportEvents': ReportEventCancer,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'CosmicId', 'CosmicLink', 'VAF',
        'additionalNumericVariantAnnotations',
        'additionalTextualVariantAnnotations', 'alternate',
        'cDNAchange', 'chromosome', 'comments', 'commonAF', 'dbSNPid',
        'depthAlternate', 'depthReference', 'position',
        'proteinChange', 'reference', 'reportEvents'
    ]

    def __init__(self, **kwargs):
        self.CosmicId = kwargs.get(
            'CosmicId', None)
        self.CosmicLink = kwargs.get(
            'CosmicLink', None)
        self.VAF = kwargs.get(
            'VAF', None)
        self.additionalNumericVariantAnnotations = kwargs.get(
            'additionalNumericVariantAnnotations', None)
        self.additionalTextualVariantAnnotations = kwargs.get(
            'additionalTextualVariantAnnotations', None)
        self.alternate = kwargs.get(
            'alternate', None)
        self.cDNAchange = kwargs.get(
            'cDNAchange', None)
        self.chromosome = kwargs.get(
            'chromosome', None)
        self.comments = kwargs.get(
            'comments', None)
        self.commonAF = kwargs.get(
            'commonAF', None)
        self.dbSNPid = kwargs.get(
            'dbSNPid', None)
        self.depthAlternate = kwargs.get(
            'depthAlternate', None)
        self.depthReference = kwargs.get(
            'depthReference', None)
        self.position = kwargs.get(
            'position', None)
        self.proteinChange = kwargs.get(
            'proteinChange', None)
        self.reference = kwargs.get(
            'reference', None)
        self.reportEvents = kwargs.get(
            'reportEvents', None)


class SampleType(object):
    """
    No documentation
    """
    germline = "germline"
    tumor = "tumor"

    def __hash__(self):
        return str(self).__hash__()


class SamtoolsScope(object):
    """
    No documentation
    """
    all = "all"
    filtered = "filtered"

    def __hash__(self):
        return str(self).__hash__()


class SamtoolsStats(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "SamtoolsStats", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"samtoolsScope", "type": {"type": "enum", "name": "SamtoolsScope", "symbols": ["all", "filtered"]}},
{"name": "SAMTOOLS_1ST_FRAGMENTS", "type": "double"}, {"name": "SAMTOOLS_AVERAGE_LENGTH", "type":
"double"}, {"name": "SAMTOOLS_AVERAGE_QUALITY", "type": "double"}, {"name":
"SAMTOOLS_BASES_DUPLICATED", "type": "double"}, {"name": "SAMTOOLS_BASES_MAPPED", "type": "double"},
{"name": "SAMTOOLS_BASES_MAPPED_CIGAR", "type": "double"}, {"name": "SAMTOOLS_BASES_TRIMMED",
"type": "double"}, {"name": "SAMTOOLS_ERROR_RATE", "type": "double"}, {"name":
"SAMTOOLS_FILTERED_SEQUENCES", "type": "double"}, {"name": "SAMTOOLS_INSERT_SIZE_AVERAGE", "type":
"double"}, {"name": "SAMTOOLS_INSERT_SIZE_STANDARD_DEVIATION", "type": "double"}, {"name":
"SAMTOOLS_INWARD_ORIENTED_PAIRS", "type": "double"}, {"name": "SAMTOOLS_IS_SORTED", "type":
"double"}, {"name": "SAMTOOLS_LAST_FRAGMENTS", "type": "double"}, {"name":
"SAMTOOLS_MAXIMUM_LENGTH", "type": "double"}, {"name": "SAMTOOLS_MISMATCHES", "type": "double"},
{"name": "SAMTOOLS_NON_PRIMARY_ALIGNMENTS", "type": "double"}, {"name":
"SAMTOOLS_OUTWARD_ORIENTED_PAIRS", "type": "double"}, {"name":
"SAMTOOLS_PAIRS_ON_DIFFERENT_CHROMOSOMES", "type": "double"}, {"name":
"SAMTOOLS_PAIRS_WITH_OTHER_ORIENTATION", "type": "double"}, {"name": "SAMTOOLS_RAW_TOTAL_SEQUENCES",
"type": "double"}, {"name": "SAMTOOLS_READS_DUPLICATED", "type": "double"}, {"name":
"SAMTOOLS_READS_MAPPED", "type": "double"}, {"name": "SAMTOOLS_READS_MAPPED_AND_PAIRED", "type":
"double"}, {"name": "SAMTOOLS_READS_MQ0", "type": "double"}, {"name": "SAMTOOLS_READS_PAIRED",
"type": "double"}, {"name": "SAMTOOLS_READS_PROPERLY_PAIRED", "type": "double"}, {"name":
"SAMTOOLS_READS_QC_FAILED", "type": "double"}, {"name": "SAMTOOLS_READS_UNMAPPED", "type":
"double"}, {"name": "SAMTOOLS_SEQUENCES", "type": "double"}, {"name": "SAMTOOLS_TOTAL_LENGTH",
"type": "double"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "SAMTOOLS_1ST_FRAGMENTS",
        "SAMTOOLS_AVERAGE_LENGTH",
        "SAMTOOLS_AVERAGE_QUALITY",
        "SAMTOOLS_BASES_DUPLICATED",
        "SAMTOOLS_BASES_MAPPED",
        "SAMTOOLS_BASES_MAPPED_CIGAR",
        "SAMTOOLS_BASES_TRIMMED",
        "SAMTOOLS_ERROR_RATE",
        "SAMTOOLS_FILTERED_SEQUENCES",
        "SAMTOOLS_INSERT_SIZE_AVERAGE",
        "SAMTOOLS_INSERT_SIZE_STANDARD_DEVIATION",
        "SAMTOOLS_INWARD_ORIENTED_PAIRS",
        "SAMTOOLS_IS_SORTED",
        "SAMTOOLS_LAST_FRAGMENTS",
        "SAMTOOLS_MAXIMUM_LENGTH",
        "SAMTOOLS_MISMATCHES",
        "SAMTOOLS_NON_PRIMARY_ALIGNMENTS",
        "SAMTOOLS_OUTWARD_ORIENTED_PAIRS",
        "SAMTOOLS_PAIRS_ON_DIFFERENT_CHROMOSOMES",
        "SAMTOOLS_PAIRS_WITH_OTHER_ORIENTATION",
        "SAMTOOLS_RAW_TOTAL_SEQUENCES",
        "SAMTOOLS_READS_DUPLICATED",
        "SAMTOOLS_READS_MAPPED",
        "SAMTOOLS_READS_MAPPED_AND_PAIRED",
        "SAMTOOLS_READS_MQ0",
        "SAMTOOLS_READS_PAIRED",
        "SAMTOOLS_READS_PROPERLY_PAIRED",
        "SAMTOOLS_READS_QC_FAILED",
        "SAMTOOLS_READS_UNMAPPED",
        "SAMTOOLS_SEQUENCES",
        "SAMTOOLS_TOTAL_LENGTH",
        "samtoolsScope",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'SAMTOOLS_1ST_FRAGMENTS', 'SAMTOOLS_AVERAGE_LENGTH',
        'SAMTOOLS_AVERAGE_QUALITY', 'SAMTOOLS_BASES_DUPLICATED',
        'SAMTOOLS_BASES_MAPPED', 'SAMTOOLS_BASES_MAPPED_CIGAR',
        'SAMTOOLS_BASES_TRIMMED', 'SAMTOOLS_ERROR_RATE',
        'SAMTOOLS_FILTERED_SEQUENCES', 'SAMTOOLS_INSERT_SIZE_AVERAGE',
        'SAMTOOLS_INSERT_SIZE_STANDARD_DEVIATION',
        'SAMTOOLS_INWARD_ORIENTED_PAIRS', 'SAMTOOLS_IS_SORTED',
        'SAMTOOLS_LAST_FRAGMENTS', 'SAMTOOLS_MAXIMUM_LENGTH',
        'SAMTOOLS_MISMATCHES', 'SAMTOOLS_NON_PRIMARY_ALIGNMENTS',
        'SAMTOOLS_OUTWARD_ORIENTED_PAIRS',
        'SAMTOOLS_PAIRS_ON_DIFFERENT_CHROMOSOMES',
        'SAMTOOLS_PAIRS_WITH_OTHER_ORIENTATION',
        'SAMTOOLS_RAW_TOTAL_SEQUENCES', 'SAMTOOLS_READS_DUPLICATED',
        'SAMTOOLS_READS_MAPPED', 'SAMTOOLS_READS_MAPPED_AND_PAIRED',
        'SAMTOOLS_READS_MQ0', 'SAMTOOLS_READS_PAIRED',
        'SAMTOOLS_READS_PROPERLY_PAIRED', 'SAMTOOLS_READS_QC_FAILED',
        'SAMTOOLS_READS_UNMAPPED', 'SAMTOOLS_SEQUENCES',
        'SAMTOOLS_TOTAL_LENGTH', 'samtoolsScope'
    ]

    def __init__(self, **kwargs):
        self.SAMTOOLS_1ST_FRAGMENTS = kwargs.get(
            'SAMTOOLS_1ST_FRAGMENTS', None)
        self.SAMTOOLS_AVERAGE_LENGTH = kwargs.get(
            'SAMTOOLS_AVERAGE_LENGTH', None)
        self.SAMTOOLS_AVERAGE_QUALITY = kwargs.get(
            'SAMTOOLS_AVERAGE_QUALITY', None)
        self.SAMTOOLS_BASES_DUPLICATED = kwargs.get(
            'SAMTOOLS_BASES_DUPLICATED', None)
        self.SAMTOOLS_BASES_MAPPED = kwargs.get(
            'SAMTOOLS_BASES_MAPPED', None)
        self.SAMTOOLS_BASES_MAPPED_CIGAR = kwargs.get(
            'SAMTOOLS_BASES_MAPPED_CIGAR', None)
        self.SAMTOOLS_BASES_TRIMMED = kwargs.get(
            'SAMTOOLS_BASES_TRIMMED', None)
        self.SAMTOOLS_ERROR_RATE = kwargs.get(
            'SAMTOOLS_ERROR_RATE', None)
        self.SAMTOOLS_FILTERED_SEQUENCES = kwargs.get(
            'SAMTOOLS_FILTERED_SEQUENCES', None)
        self.SAMTOOLS_INSERT_SIZE_AVERAGE = kwargs.get(
            'SAMTOOLS_INSERT_SIZE_AVERAGE', None)
        self.SAMTOOLS_INSERT_SIZE_STANDARD_DEVIATION = kwargs.get(
            'SAMTOOLS_INSERT_SIZE_STANDARD_DEVIATION', None)
        self.SAMTOOLS_INWARD_ORIENTED_PAIRS = kwargs.get(
            'SAMTOOLS_INWARD_ORIENTED_PAIRS', None)
        self.SAMTOOLS_IS_SORTED = kwargs.get(
            'SAMTOOLS_IS_SORTED', None)
        self.SAMTOOLS_LAST_FRAGMENTS = kwargs.get(
            'SAMTOOLS_LAST_FRAGMENTS', None)
        self.SAMTOOLS_MAXIMUM_LENGTH = kwargs.get(
            'SAMTOOLS_MAXIMUM_LENGTH', None)
        self.SAMTOOLS_MISMATCHES = kwargs.get(
            'SAMTOOLS_MISMATCHES', None)
        self.SAMTOOLS_NON_PRIMARY_ALIGNMENTS = kwargs.get(
            'SAMTOOLS_NON_PRIMARY_ALIGNMENTS', None)
        self.SAMTOOLS_OUTWARD_ORIENTED_PAIRS = kwargs.get(
            'SAMTOOLS_OUTWARD_ORIENTED_PAIRS', None)
        self.SAMTOOLS_PAIRS_ON_DIFFERENT_CHROMOSOMES = kwargs.get(
            'SAMTOOLS_PAIRS_ON_DIFFERENT_CHROMOSOMES', None)
        self.SAMTOOLS_PAIRS_WITH_OTHER_ORIENTATION = kwargs.get(
            'SAMTOOLS_PAIRS_WITH_OTHER_ORIENTATION', None)
        self.SAMTOOLS_RAW_TOTAL_SEQUENCES = kwargs.get(
            'SAMTOOLS_RAW_TOTAL_SEQUENCES', None)
        self.SAMTOOLS_READS_DUPLICATED = kwargs.get(
            'SAMTOOLS_READS_DUPLICATED', None)
        self.SAMTOOLS_READS_MAPPED = kwargs.get(
            'SAMTOOLS_READS_MAPPED', None)
        self.SAMTOOLS_READS_MAPPED_AND_PAIRED = kwargs.get(
            'SAMTOOLS_READS_MAPPED_AND_PAIRED', None)
        self.SAMTOOLS_READS_MQ0 = kwargs.get(
            'SAMTOOLS_READS_MQ0', None)
        self.SAMTOOLS_READS_PAIRED = kwargs.get(
            'SAMTOOLS_READS_PAIRED', None)
        self.SAMTOOLS_READS_PROPERLY_PAIRED = kwargs.get(
            'SAMTOOLS_READS_PROPERLY_PAIRED', None)
        self.SAMTOOLS_READS_QC_FAILED = kwargs.get(
            'SAMTOOLS_READS_QC_FAILED', None)
        self.SAMTOOLS_READS_UNMAPPED = kwargs.get(
            'SAMTOOLS_READS_UNMAPPED', None)
        self.SAMTOOLS_SEQUENCES = kwargs.get(
            'SAMTOOLS_SEQUENCES', None)
        self.SAMTOOLS_TOTAL_LENGTH = kwargs.get(
            'SAMTOOLS_TOTAL_LENGTH', None)
        self.samtoolsScope = kwargs.get(
            'samtoolsScope', None)


class SensitiveInformation(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "SensitiveInformation", "namespace": "Gel_BioInf_Models", "fields":
[{"name": "versionControl", "type": {"type": "record", "name": "VersionControl", "fields": [{"name":
"GitVersionControl", "type": "string", "doc": "", "default": "2.1.0"}]}, "doc": ""}, {"name":
"gelID", "type": "string"}, {"name": "externalIds", "type": ["null", {"type": "array", "items":
"string"}]}, {"name": "genomicMedicineCenter", "type": ["null", "string"]}, {"name":
"fullNameOfResponsibleConsultant", "type": ["null", "string"]}, {"name": "contactNumber", "type":
["null", "string"]}, {"name": "hospitalOfResponsibleConsultant", "type": ["null", "string"]},
{"name": "centerSampleId", "type": ["null", "string"]}, {"name": "originatingCenter", "type":
["null", "string"]}, {"name": "centerPatientId", "type": ["null", "string"]}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "centerPatientId",
        "centerSampleId",
        "contactNumber",
        "externalIds",
        "fullNameOfResponsibleConsultant",
        "gelID",
        "genomicMedicineCenter",
        "hospitalOfResponsibleConsultant",
        "originatingCenter",
        "versionControl",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'versionControl': VersionControl,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'versionControl': VersionControl,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'centerPatientId', 'centerSampleId', 'contactNumber',
        'externalIds', 'fullNameOfResponsibleConsultant', 'gelID',
        'genomicMedicineCenter', 'hospitalOfResponsibleConsultant',
        'originatingCenter', 'versionControl'
    ]

    def __init__(self, **kwargs):
        self.centerPatientId = kwargs.get(
            'centerPatientId', None)
        self.centerSampleId = kwargs.get(
            'centerSampleId', None)
        self.contactNumber = kwargs.get(
            'contactNumber', None)
        self.externalIds = kwargs.get(
            'externalIds', None)
        self.fullNameOfResponsibleConsultant = kwargs.get(
            'fullNameOfResponsibleConsultant', None)
        self.gelID = kwargs.get(
            'gelID', None)
        self.genomicMedicineCenter = kwargs.get(
            'genomicMedicineCenter', None)
        self.hospitalOfResponsibleConsultant = kwargs.get(
            'hospitalOfResponsibleConsultant', None)
        self.originatingCenter = kwargs.get(
            'originatingCenter', None)
        self.versionControl = kwargs.get(
            'versionControl', VersionControl())


class Sex(object):
    """
    Sex
    """
    male = "male"
    female = "female"
    unknown = "unknown"
    undetermined = "undetermined"

    def __hash__(self):
        return str(self).__hash__()


class SomaticOrGermline(object):
    """
    this defines whether the variant reported is considered to be a
    somatic or germline finding.
    """
    unknown = "unknown"
    somatic = "somatic"
    germline = "germline"

    def __hash__(self):
        return str(self).__hash__()


class State(object):
    """
    This is the master state for this sample, for example
    caution,quality could be used to say that a sample under this
    individual has quality issues.  ready: sample is ready to be used
    pending: sample is in the process of being analysed hold: sample
    is on hold pending investigation fail: sample has failed a QC
    check caution: sample is ready but should be used with caution
    """
    ready = "ready"
    pending = "pending"
    hold = "hold"
    fail = "fail"
    caution = "caution"

    def __hash__(self):
        return str(self).__hash__()


class TernaryOption(object):
    """
    This the define a yes/no/unknown case
    """
    yes = "yes"
    no = "no"
    unknown = "unknown"

    def __hash__(self):
        return str(self).__hash__()


class Tier(object):
    """
    Possible tiers as defined by Genomics England
    """
    NONE = "NONE"
    TIER1 = "TIER1"
    TIER2 = "TIER2"
    TIER3 = "TIER3"

    def __hash__(self):
        return str(self).__hash__()


class TumorChecks(ProtocolElement):
    """
    Tumor contamination check results  We ask whether somatic calls
    are common in 1000G project.  Common is defined as 0.01.  The
    percentage of these variants gives an indication of the
    contamination level of the sample.  If another sample contaminates
    or the germline is swapped then the percentage of somatic variants
    in 1000G will be high.  * `well_id`: Individual sample ID *
    `num_variants_gt_filter`: The number of somatic variants that are
    common in 1000G * `num_variants_lt_filter`: The number of somatic
    variants rare or non-existant in 1000G *
    `perc_variants_gt_filter`: The percentage of somatic variants that
    are common in 1000G
    """
    _schemaSource = """
{"type": "record", "name": "TumorChecks", "namespace": "Gel_BioInf_Models", "doc": "", "fields":
[{"name": "well_id", "type": "string"}, {"name": "num_variants_gt_filter", "type": "double"},
{"name": "num_variants_lt_filter", "type": "double"}, {"name": "perc_variants_gt_filter", "type":
"double"}, {"name": "TumorContaminationContEst", "type": "double"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "TumorContaminationContEst",
        "num_variants_gt_filter",
        "num_variants_lt_filter",
        "perc_variants_gt_filter",
        "well_id",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'TumorContaminationContEst', 'num_variants_gt_filter',
        'num_variants_lt_filter', 'perc_variants_gt_filter', 'well_id'
    ]

    def __init__(self, **kwargs):
        self.TumorContaminationContEst = kwargs.get(
            'TumorContaminationContEst', None)
        self.num_variants_gt_filter = kwargs.get(
            'num_variants_gt_filter', None)
        self.num_variants_lt_filter = kwargs.get(
            'num_variants_lt_filter', None)
        self.perc_variants_gt_filter = kwargs.get(
            'perc_variants_gt_filter', None)
        self.well_id = kwargs.get(
            'well_id', None)


class VariantClassification(object):
    """
    This is the classification of the variant according to standard
    practice guidelines (e.g. ACMG)
    """
    BENIGN = "BENIGN"
    LIKELY_BENIGN = "LIKELY_BENIGN"
    VUS = "VUS"
    LIKELY_PATHOGENIC = "LIKELY_PATHOGENIC"
    PATHOGENIC = "PATHOGENIC"

    def __hash__(self):
        return str(self).__hash__()


class VcfMetrics(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "VcfMetrics", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"NUMBER_OF_SAMPLES", "type": "double"}, {"name": "NUMBER_OF_INDELS", "type": "double"}, {"name":
"NUMBER_OF_MNPS", "type": "double"}, {"name": "NUMBER_OF_MULTIALLELIC_SNP_SITES", "type": "double"},
{"name": "NUMBER_OF_SNPS", "type": "double"}, {"name": "NUMBER_OF_RECORDS", "type": "double"},
{"name": "NUMBER_OF_OTHERS", "type": "double"}, {"name": "NUMBER_OF_MULTIALLELIC_SITES", "type":
"double"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "NUMBER_OF_INDELS",
        "NUMBER_OF_MNPS",
        "NUMBER_OF_MULTIALLELIC_SITES",
        "NUMBER_OF_MULTIALLELIC_SNP_SITES",
        "NUMBER_OF_OTHERS",
        "NUMBER_OF_RECORDS",
        "NUMBER_OF_SAMPLES",
        "NUMBER_OF_SNPS",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'NUMBER_OF_INDELS', 'NUMBER_OF_MNPS',
        'NUMBER_OF_MULTIALLELIC_SITES',
        'NUMBER_OF_MULTIALLELIC_SNP_SITES', 'NUMBER_OF_OTHERS',
        'NUMBER_OF_RECORDS', 'NUMBER_OF_SAMPLES', 'NUMBER_OF_SNPS'
    ]

    def __init__(self, **kwargs):
        self.NUMBER_OF_INDELS = kwargs.get(
            'NUMBER_OF_INDELS', None)
        self.NUMBER_OF_MNPS = kwargs.get(
            'NUMBER_OF_MNPS', None)
        self.NUMBER_OF_MULTIALLELIC_SITES = kwargs.get(
            'NUMBER_OF_MULTIALLELIC_SITES', None)
        self.NUMBER_OF_MULTIALLELIC_SNP_SITES = kwargs.get(
            'NUMBER_OF_MULTIALLELIC_SNP_SITES', None)
        self.NUMBER_OF_OTHERS = kwargs.get(
            'NUMBER_OF_OTHERS', None)
        self.NUMBER_OF_RECORDS = kwargs.get(
            'NUMBER_OF_RECORDS', None)
        self.NUMBER_OF_SAMPLES = kwargs.get(
            'NUMBER_OF_SAMPLES', None)
        self.NUMBER_OF_SNPS = kwargs.get(
            'NUMBER_OF_SNPS', None)


class VcfTSTV(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "VcfTSTV", "namespace": "Gel_BioInf_Models", "fields": [{"name": "TS_1",
"type": "double"}, {"name": "TV", "type": "double"}, {"name": "TS", "type": "double"}, {"name":
"TS_TV", "type": "double"}, {"name": "TV_1", "type": "double"}, {"name": "TS_TV_1", "type":
"double"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "TS",
        "TS_1",
        "TS_TV",
        "TS_TV_1",
        "TV",
        "TV_1",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'TS', 'TS_1', 'TS_TV', 'TS_TV_1', 'TV', 'TV_1'
    ]

    def __init__(self, **kwargs):
        self.TS = kwargs.get(
            'TS', None)
        self.TS_1 = kwargs.get(
            'TS_1', None)
        self.TS_TV = kwargs.get(
            'TS_TV', None)
        self.TS_TV_1 = kwargs.get(
            'TS_TV_1', None)
        self.TV = kwargs.get(
            'TV', None)
        self.TV_1 = kwargs.get(
            'TV_1', None)


class VerifyBamId(ProtocolElement):
    """
    Verify Bam ID  This is only run on germline samples  * `SEQ_ID`: *
    `CHIP_ID`: * `SNPS`: * `READS`: * `AVG_DP`: * `FREEMIX`: *
    `FREELK1`: * `FREELK0`: * `FREE_RH`: * `FREE_RA`: * `CHIPMIX`: *
    `CHIPLK1`: * `CHIPLK0`: * `CHIP_RA`: * `DPREF`: * `RDPHET`: *
    `RDPALT`:
    """
    _schemaSource = """
{"type": "record", "name": "VerifyBamId", "namespace": "Gel_BioInf_Models", "doc": "", "fields":
[{"name": "SEQ_ID", "type": "string"}, {"name": "CHIP_ID", "type": "string"}, {"name": "SNPS",
"type": "double"}, {"name": "READS", "type": "double"}, {"name": "AVG_DP", "type": "double"},
{"name": "FREEMIX", "type": "double"}, {"name": "FREELK1", "type": "double"}, {"name": "FREELK0",
"type": "double"}, {"name": "FREE_RH", "type": "string"}, {"name": "FREE_RA", "type": "string"},
{"name": "CHIPMIX", "type": "string"}, {"name": "CHIPLK1", "type": "string"}, {"name": "CHIPLK0",
"type": "string"}, {"name": "CHIP_RA", "type": "string"}, {"name": "DPREF", "type": "string"},
{"name": "RDPHET", "type": "string"}, {"name": "RDPALT", "type": "string"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "AVG_DP",
        "CHIPLK0",
        "CHIPLK1",
        "CHIPMIX",
        "CHIP_ID",
        "CHIP_RA",
        "DPREF",
        "FREELK0",
        "FREELK1",
        "FREEMIX",
        "FREE_RA",
        "FREE_RH",
        "RDPALT",
        "RDPHET",
        "READS",
        "SEQ_ID",
        "SNPS",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'AVG_DP', 'CHIPLK0', 'CHIPLK1', 'CHIPMIX', 'CHIP_ID',
        'CHIP_RA', 'DPREF', 'FREELK0', 'FREELK1', 'FREEMIX',
        'FREE_RA', 'FREE_RH', 'RDPALT', 'RDPHET', 'READS', 'SEQ_ID',
        'SNPS'
    ]

    def __init__(self, **kwargs):
        self.AVG_DP = kwargs.get(
            'AVG_DP', None)
        self.CHIPLK0 = kwargs.get(
            'CHIPLK0', None)
        self.CHIPLK1 = kwargs.get(
            'CHIPLK1', None)
        self.CHIPMIX = kwargs.get(
            'CHIPMIX', None)
        self.CHIP_ID = kwargs.get(
            'CHIP_ID', None)
        self.CHIP_RA = kwargs.get(
            'CHIP_RA', None)
        self.DPREF = kwargs.get(
            'DPREF', None)
        self.FREELK0 = kwargs.get(
            'FREELK0', None)
        self.FREELK1 = kwargs.get(
            'FREELK1', None)
        self.FREEMIX = kwargs.get(
            'FREEMIX', None)
        self.FREE_RA = kwargs.get(
            'FREE_RA', None)
        self.FREE_RH = kwargs.get(
            'FREE_RH', None)
        self.RDPALT = kwargs.get(
            'RDPALT', None)
        self.RDPHET = kwargs.get(
            'RDPHET', None)
        self.READS = kwargs.get(
            'READS', None)
        self.SEQ_ID = kwargs.get(
            'SEQ_ID', None)
        self.SNPS = kwargs.get(
            'SNPS', None)


class VersionControl(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "VersionControl", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"GitVersionControl", "type": "string", "doc": "", "default": "2.1.0"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {}

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'GitVersionControl'
    ]

    def __init__(self, **kwargs):
        self.GitVersionControl = kwargs.get(
            'GitVersionControl', '2.1.0')


class VirtualPanel(ProtocolElement):
    """
    This record defines a gene panel.
    """
    _schemaSource = """
{"type": "record", "name": "VirtualPanel", "namespace": "Gel_BioInf_Models", "doc": "", "fields":
[{"name": "panelName", "type": "string", "doc": ""}, {"name": "panelVersion", "type": "string",
"doc": ""}, {"name": "ensemblVersion", "type": ["null", "string"], "doc": ""}, {"name":
"dataModelCatalogueVersion", "type": "string", "doc": "", "default": "4.2"}, {"name": "geneIds",
"type": {"type": "array", "items": "string"}, "doc": ""}, {"name": "Transcripts", "type": ["null",
{"type": "array", "items": "string"}], "doc": ""}, {"name": "relevantRegions", "type": ["null",
{"type": "record", "name": "File", "fields": [{"name": "SampleId", "type": ["null", "string",
{"type": "array", "items": "string"}], "doc": ""}, {"name": "URIFile", "type": "string", "doc": ""},
{"name": "fileType", "type": {"type": "enum", "name": "FileType", "doc": "", "symbols": ["BAM",
"gVCF", "VCF_small", "VCF_CNV", "VCF_SV", "VCF_SV_CNV", "SVG", "TIER", "ANN", "BigWig"]}}]}], "doc":
""}, {"name": "clinicalRelevantVariants", "type": ["null", "File"], "doc": ""}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "Transcripts",
        "clinicalRelevantVariants",
        "ensemblVersion",
        "geneIds",
        "panelName",
        "panelVersion",
        "relevantRegions",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'clinicalRelevantVariants': File,
            'relevantRegions': File,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'clinicalRelevantVariants': File,
            'relevantRegions': File,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'Transcripts', 'clinicalRelevantVariants',
        'dataModelCatalogueVersion', 'ensemblVersion', 'geneIds',
        'panelName', 'panelVersion', 'relevantRegions'
    ]

    def __init__(self, **kwargs):
        self.Transcripts = kwargs.get(
            'Transcripts', None)
        self.clinicalRelevantVariants = kwargs.get(
            'clinicalRelevantVariants', None)
        self.dataModelCatalogueVersion = kwargs.get(
            'dataModelCatalogueVersion', '4.2')
        self.ensemblVersion = kwargs.get(
            'ensemblVersion', None)
        self.geneIds = kwargs.get(
            'geneIds', None)
        self.panelName = kwargs.get(
            'panelName', None)
        self.panelVersion = kwargs.get(
            'panelVersion', None)
        self.relevantRegions = kwargs.get(
            'relevantRegions', None)


class WholeGenomeCoverage(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "WholeGenomeCoverage", "namespace": "Gel_BioInf_Models", "fields":
[{"name": "coverageSummary", "type": {"type": "array", "items": {"type": "record", "name":
"CoverageSummary", "fields": [{"name": "wellId", "type": "string"}, {"name": "n", "type": "double"},
{"name": "mean", "type": "double"}, {"name": "sd", "type": "double"}, {"name": "pct25", "type":
"double"}, {"name": "median", "type": "double"}, {"name": "pct75", "type": "double"}, {"name":
"scope", "type": "string"}, {"name": "localRMSD", "type": "double"}, {"name": "COSMIC30Xcov",
"type": "double"}]}}}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "coverageSummary",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'coverageSummary': CoverageSummary,
        }
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {
            'coverageSummary': CoverageSummary,
        }

        return embeddedTypes[fieldName]

    __slots__ = [
        'coverageSummary'
    ]

    def __init__(self, **kwargs):
        self.coverageSummary = kwargs.get(
            'coverageSummary', None)


class individualState(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "individualState", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"state", "type": {"type": "enum", "name": "State", "doc": "", "symbols": ["ready", "pending",
"hold", "fail", "caution", "blocked"]}}, {"name": "reason", "type": ["null", {"type": "enum",
"name": "Reason", "symbols": ["duplicate", "consent", "pedigree", "contamination", "quality",
"plinksex", "inbreedingcoefficient", "in_qc"]}]}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "reason",
        "state",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'reason', 'state'
    ]

    def __init__(self, **kwargs):
        self.reason = kwargs.get(
            'reason', None)
        self.state = kwargs.get(
            'state', None)


class individualTests(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "individualTests", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"plinksex", "type": "boolean"}, {"name": "inbreedingcoefficient", "type": "boolean"}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "inbreedingcoefficient",
        "plinksex",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'inbreedingcoefficient', 'plinksex'
    ]

    def __init__(self, **kwargs):
        self.inbreedingcoefficient = kwargs.get(
            'inbreedingcoefficient', None)
        self.plinksex = kwargs.get(
            'plinksex', None)


class sampleState(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "sampleState", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"state", "type": ["null", {"type": "enum", "name": "State", "doc": "", "symbols": ["ready",
"pending", "hold", "fail", "caution"]}]}, {"name": "reason", "type": {"type": "array", "items":
{"type": "enum", "name": "Reason", "symbols": ["median_coverage", "in_analysis", "duplicate",
"pedigree_mendelian_errors", "pedigree_ibd_sharing", "contamination", "quality", "sex_query",
"perc_bases_ge_15x_mapQ_ge11", "GbQ30NoDupsNoClip", "arrayconcordance", "high_cnv", "in_qc",
"pass_qc", "other"]}}}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "reason",
        "state",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'reason', 'state'
    ]

    def __init__(self, **kwargs):
        self.reason = kwargs.get(
            'reason', None)
        self.state = kwargs.get(
            'state', None)


class sampleTests(ProtocolElement):
    """
    No documentation
    """
    _schemaSource = """
{"type": "record", "name": "sampleTests", "namespace": "Gel_BioInf_Models", "fields": [{"name":
"verifybamid", "type": ["null", "boolean"]}, {"name": "arrayconcordance", "type": ["null",
"boolean"]}, {"name": "contamination", "type": ["null", "boolean"]}]}
"""
    schema = avro_parse(_schemaSource)
    requiredFields = {
        "arrayconcordance",
        "contamination",
        "verifybamid",
    }

    @classmethod
    def isEmbeddedType(cls, fieldName):
        embeddedTypes = {}
        return fieldName in embeddedTypes

    @classmethod
    def getEmbeddedType(cls, fieldName):
        embeddedTypes = {}

        return embeddedTypes[fieldName]

    __slots__ = [
        'arrayconcordance', 'contamination', 'verifybamid'
    ]

    def __init__(self, **kwargs):
        self.arrayconcordance = kwargs.get(
            'arrayconcordance', None)
        self.contamination = kwargs.get(
            'contamination', None)
        self.verifybamid = kwargs.get(
            'verifybamid', None)
